<!DOCTYPE html>
<html lang="en" dir="auto">

<head><script src="/livereload.js?mindelay=10&amp;v=2&amp;port=1313&amp;path=livereload" data-no-instant defer></script><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">
<title>scRNAçš„ç»†èƒè´¨æ§ | Jeason&#39;s Blog</title>
<meta name="keywords" content="SingleCell">
<meta name="description" content="åŠ¨æœº scRNA-seqæ•°æ®çš„ä½è´¨é‡æ–‡åº“å¯èƒ½æ¥è‡ªäºï¼šç»†èƒåˆ†é€‰ç¯èŠ‚çš„ç ´åã€æ–‡åº“åˆ¶å¤‡å¤±è¯¯ï¼ˆæ²¡æœ‰è¶³å¤Ÿçš„åè½¬å½•æˆ–PCRæ¬¡æ•°ï¼‰â€¦ è¡¨ç°åœ¨ï¼šç»†èƒæ€»è¡¨è¾¾é‡ä½ã€åŸºæœ¬æ²¡æœ‰è¡¨è¾¾çš„åŸºå› ã€é«˜çº¿ç²’ä½“æˆ–spike-inå æ¯”ã€‚ è¿™äº›ä½è´¨é‡çš„åº“æ˜¯æœ‰é—®é¢˜çš„ï¼Œå› ä¸ºå®ƒä»¬å¯èƒ½åœ¨ä¸‹æ¸¸åˆ†æä¸­å¯¼è‡´è¯¯å¯¼çš„ç»“æœ ã€‚ èšç±»é—®é¢˜ï¼š ä½è´¨é‡çš„ç»†èƒ">
<meta name="author" content="ä½œè€…:Jeason">
<link rel="canonical" href="http://localhost:1313/en/posts/tech/sc_qc/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.e029fa55ed0a02bb684da89d2a53a4e1ba9e26e3719fa48460be52776460b523.css" integrity="sha256-4Cn6Ve0KArtoTaidKlOk4bqeJuNxn6SEYL5Sd2RgtSM=" rel="preload stylesheet" as="style">
<link rel="icon" href="http://localhost:1313/img/avatar.jpeg">
<link rel="icon" type="image/png" sizes="16x16" href="http://localhost:1313/img/avatar.jpeg">
<link rel="icon" type="image/png" sizes="32x32" href="http://localhost:1313/img/avatar.jpeg">
<link rel="apple-touch-icon" href="http://localhost:1313/img/avatar.jpeg">
<link rel="mask-icon" href="http://localhost:1313/img/avatar.jpeg">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="http://localhost:1313/en/posts/tech/sc_qc/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --code-block-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><script type="text/javascript"
        async
        src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
  tex2jax: {
    inlineMath: [['$','$'], ['\\(','\\)']],
    displayMath: [['$$','$$'], ['\[\[','\]\]']],
    processEscapes: true,
    processEnvironments: true,
    skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
    TeX: { equationNumbers: { autoNumber: "AMS" },
         extensions: ["AMSmath.js", "AMSsymbols.js"] }
  }
});

MathJax.Hub.Queue(function() {
    
    
    
    var all = MathJax.Hub.getAllJax(), i;
    for(i = 0; i < all.length; i += 1) {
        all[i].SourceElement().parentNode.className += ' has-jax';
    }
});
</script>

<style>
code.has-jax {
    font: inherit;
    font-size: 100%;
    background: inherit;
    border: inherit;
    color: #515151;
}
</style>


<script>
var _hmt = _hmt || [];
(function() {
  var hm = document.createElement("script");
  hm.src = ""; 
  var s = document.getElementsByTagName("script")[0]; 
  s.parentNode.insertBefore(hm, s);
})();
</script>



<meta property="og:title" content="scRNAçš„ç»†èƒè´¨æ§" />
<meta property="og:description" content="åŠ¨æœº scRNA-seqæ•°æ®çš„ä½è´¨é‡æ–‡åº“å¯èƒ½æ¥è‡ªäºï¼šç»†èƒåˆ†é€‰ç¯èŠ‚çš„ç ´åã€æ–‡åº“åˆ¶å¤‡å¤±è¯¯ï¼ˆæ²¡æœ‰è¶³å¤Ÿçš„åè½¬å½•æˆ–PCRæ¬¡æ•°ï¼‰â€¦ è¡¨ç°åœ¨ï¼šç»†èƒæ€»è¡¨è¾¾é‡ä½ã€åŸºæœ¬æ²¡æœ‰è¡¨è¾¾çš„åŸºå› ã€é«˜çº¿ç²’ä½“æˆ–spike-inå æ¯”ã€‚ è¿™äº›ä½è´¨é‡çš„åº“æ˜¯æœ‰é—®é¢˜çš„ï¼Œå› ä¸ºå®ƒä»¬å¯èƒ½åœ¨ä¸‹æ¸¸åˆ†æä¸­å¯¼è‡´è¯¯å¯¼çš„ç»“æœ ã€‚ èšç±»é—®é¢˜ï¼š ä½è´¨é‡çš„ç»†èƒ" />
<meta property="og:type" content="article" />
<meta property="og:url" content="http://localhost:1313/en/posts/tech/sc_qc/" />
<meta property="og:image" content="https://source.unsplash.com/random/400x200?code" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2020-08-12T16:50:44+08:00" />
<meta property="article:modified_time" content="2024-02-29T16:50:44+08:00" />

<meta name="twitter:card" content="summary_large_image" />
<meta name="twitter:image" content="https://source.unsplash.com/random/400x200?code" />
<meta name="twitter:title" content="scRNAçš„ç»†èƒè´¨æ§"/>
<meta name="twitter:description" content="åŠ¨æœº scRNA-seqæ•°æ®çš„ä½è´¨é‡æ–‡åº“å¯èƒ½æ¥è‡ªäºï¼šç»†èƒåˆ†é€‰ç¯èŠ‚çš„ç ´åã€æ–‡åº“åˆ¶å¤‡å¤±è¯¯ï¼ˆæ²¡æœ‰è¶³å¤Ÿçš„åè½¬å½•æˆ–PCRæ¬¡æ•°ï¼‰â€¦ è¡¨ç°åœ¨ï¼šç»†èƒæ€»è¡¨è¾¾é‡ä½ã€åŸºæœ¬æ²¡æœ‰è¡¨è¾¾çš„åŸºå› ã€é«˜çº¿ç²’ä½“æˆ–spike-inå æ¯”ã€‚ è¿™äº›ä½è´¨é‡çš„åº“æ˜¯æœ‰é—®é¢˜çš„ï¼Œå› ä¸ºå®ƒä»¬å¯èƒ½åœ¨ä¸‹æ¸¸åˆ†æä¸­å¯¼è‡´è¯¯å¯¼çš„ç»“æœ ã€‚ èšç±»é—®é¢˜ï¼š ä½è´¨é‡çš„ç»†èƒ"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "ğŸ“š æ–‡ç« ",
      "item": "http://localhost:1313/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "ğŸ‘¨ğŸ»â€ğŸ’» æŠ€æœ¯",
      "item": "http://localhost:1313/en/posts/tech/"
    }, 
    {
      "@type": "ListItem",
      "position":  3 ,
      "name": "scRNAçš„ç»†èƒè´¨æ§",
      "item": "http://localhost:1313/en/posts/tech/sc_qc/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "scRNAçš„ç»†èƒè´¨æ§",
  "name": "scRNAçš„ç»†èƒè´¨æ§",
  "description": "åŠ¨æœº scRNA-seqæ•°æ®çš„ä½è´¨é‡æ–‡åº“å¯èƒ½æ¥è‡ªäºï¼šç»†èƒåˆ†é€‰ç¯èŠ‚çš„ç ´åã€æ–‡åº“åˆ¶å¤‡å¤±è¯¯ï¼ˆæ²¡æœ‰è¶³å¤Ÿçš„åè½¬å½•æˆ–PCRæ¬¡æ•°ï¼‰â€¦ è¡¨ç°åœ¨ï¼šç»†èƒæ€»è¡¨è¾¾é‡ä½ã€åŸºæœ¬æ²¡æœ‰è¡¨è¾¾çš„åŸºå› ã€é«˜çº¿ç²’ä½“æˆ–spike-inå æ¯”ã€‚ è¿™äº›ä½è´¨é‡çš„åº“æ˜¯æœ‰é—®é¢˜çš„ï¼Œå› ä¸ºå®ƒä»¬å¯èƒ½åœ¨ä¸‹æ¸¸åˆ†æä¸­å¯¼è‡´è¯¯å¯¼çš„ç»“æœ ã€‚ èšç±»é—®é¢˜ï¼š ä½è´¨é‡çš„ç»†èƒ",
  "keywords": [
    "SingleCell"
  ],
  "articleBody": "åŠ¨æœº scRNA-seqæ•°æ®çš„ä½è´¨é‡æ–‡åº“å¯èƒ½æ¥è‡ªäºï¼šç»†èƒåˆ†é€‰ç¯èŠ‚çš„ç ´åã€æ–‡åº“åˆ¶å¤‡å¤±è¯¯ï¼ˆæ²¡æœ‰è¶³å¤Ÿçš„åè½¬å½•æˆ–PCRæ¬¡æ•°ï¼‰â€¦ è¡¨ç°åœ¨ï¼šç»†èƒæ€»è¡¨è¾¾é‡ä½ã€åŸºæœ¬æ²¡æœ‰è¡¨è¾¾çš„åŸºå› ã€é«˜çº¿ç²’ä½“æˆ–spike-inå æ¯”ã€‚\nè¿™äº›ä½è´¨é‡çš„åº“æ˜¯æœ‰é—®é¢˜çš„ï¼Œå› ä¸ºå®ƒä»¬å¯èƒ½åœ¨ä¸‹æ¸¸åˆ†æä¸­å¯¼è‡´è¯¯å¯¼çš„ç»“æœ ã€‚\nèšç±»é—®é¢˜ï¼š ä½è´¨é‡çš„ç»†èƒä¼šèšé›†æˆä¸€ç¾¤ï¼Œå¯¹ç»“æœçš„è§£é‡Šé€ æˆå¹²æ‰°ï¼Œå› ä¸ºä»è¿™ç¾¤ç»†èƒä¸­å¾—ä¸åˆ°ä»€ä¹ˆæœ‰ç”¨çš„ä¿¡æ¯ï¼Œä½†æ˜¯å®ƒçš„ç¡®ä¹Ÿæ˜¯ä¸€ç¾¤ã€‚è¿™ç§ç°è±¡äº§ç”Ÿçš„åŸå› æœ‰å¯èƒ½æ˜¯ï¼šç»†èƒç ´åä»¥åï¼Œçº¿ç²’ä½“æˆ–æ ¸RNAså æ¯”å‡é«˜ã€‚ æœ€å·®çš„æƒ…å†µå°±æ˜¯ï¼šä¸åŒç±»å‹çš„ä½è´¨é‡ç»†èƒï¼Œä¹Ÿèƒ½èšåœ¨ä¸€èµ·ï¼Œå› ä¸ºç›¸æ¯”äºå›ºæœ‰çš„ç”Ÿç‰©å·®å¼‚ï¼Œæ›´ç›¸ä¼¼çš„ä½è´¨é‡è®©å®ƒä»¬ç›¸ä¾ç›¸åã€‚é™¤æ­¤ä»¥å¤–ï¼Œæœ¬æ¥éå¸¸å°çš„ç»†èƒæ–‡åº“ä¹Ÿèƒ½èšæˆä¸€ç¾¤ï¼Œå› ä¸ºlogè½¬æ¢åå®ƒä»¬çš„å¹³å‡è¡¨è¾¾é‡ä¼šå‘ç”Ÿå¾ˆå¤§çš„å˜åŒ–(A. Lun 2018). æ–¹å·®ä¼°è®¡æˆ–ä¸»æˆåˆ†åˆ†æ:é¦–å…ˆåœ¨PCAåˆ†ææ—¶ï¼Œä½è´¨é‡å’Œé«˜è´¨é‡ä¹‹é—´çš„å·®å¼‚ç›¸æ¯”äºç”Ÿç‰©å­¦å·®å¼‚ä¼šä½“ç°æ›´æ˜æ˜¾ï¼Œå æ®ä¸»è¦çš„æˆåˆ†ï¼Œå‡å°‘é™ç»´ç»“æœçš„å¯é æ€§ã€‚å¦å¤–ï¼ŒæŸä¸ªåŸºå› å¯èƒ½åœ¨ä¸¤ä¸ªç»†èƒä¹‹é—´æ²¡ä»€ä¹ˆè¡¨è¾¾å·®å¼‚ï¼Œä½†æ˜¯å¦‚æœæ‰€å¤„ç¯å¢ƒå·®å¼‚å¾ˆå¤§ï¼ˆä¸€ä¸ªç»†èƒè´¨é‡å¾ˆä½ï¼Œå¦ä¸€ä¸ªç»†èƒè´¨é‡æ­£å¸¸ï¼‰ï¼Œé‚£ä¹ˆåœ¨å·®å¼‚ä¼°ç®—è¿‡ç¨‹ä¸­ï¼Œå°±ä¼šæŠŠè¿™ä¸ªå·®å¼‚ä¹Ÿä¼šè¢«å½“æˆå·®å¼‚è¡¨è¾¾åŸºå› ã€‚ä¾‹å¦‚ï¼šä¸€ä¸ªä½è´¨é‡ç»†èƒæ–‡åº“çš„æ€»è¡¨è¾¾é‡éå¸¸ä½ï¼ˆæ¥è¿‘0ï¼‰ï¼Œä½†æ°å·§è¿˜å­˜åœ¨ä¸€ä¸ªåŸºå› æœ‰è¡¨è¾¾é‡ï¼Œé‚£ä¹ˆè¿™ä¸ªåŸºå› çš„è¡¨è¾¾é‡åœ¨åç»­çš„æ–‡åº“å½’ä¸€åŒ–è¿‡ç¨‹ä¸­å°±ä¼šå°¤ä¸ºçªå‡º å¥‡æ€ªçš„è½¬å½•æœ¬ä¸Šè°ƒï¼šå®éªŒéš¾å…ä¼šæ··å…¥å¤–æºçš„æ±¡æŸ“è½¬å½•æœ¬ï¼Œä½†è¿™ä¸ªé‡å¾ˆå°‘å¹¶ä¸”åœ¨æ‰€æœ‰ç»†èƒä¸­éƒ½åº”è¯¥æ˜¯å·®ä¸å¤šæ°´å¹³çš„ã€‚ä½†å¦‚æœæŸä¸ªç»†èƒè´¨é‡ä½ï¼Œæ–‡åº“å°ï¼Œé‚£ä¹ˆåœ¨æ ¡æ­£æ–‡åº“å·®å¼‚è¿‡ç¨‹ä¸­ï¼Œå…¶ä¸­çš„æ±¡æŸ“è½¬å½•æœ¬è¡¨è¾¾é‡å°±ä¼šâ€œçªé£çŒ›è¿›â€ï¼Œçœ‹èµ·æ¥æ˜¯ä¸€äº›æ˜æ˜¾ä¸Šè°ƒçš„â€œå¥‡æ€ªåŸºå› â€ã€‚å®é™…ä¸Šï¼Œè¿™äº›å¥‡æ€ªçš„åŸºå› ä¾ç„¶åœ¨å…¶ä»–ç»†èƒä¸­å­˜åœ¨ï¼Œå¹¶ä¸”çœŸå®çš„è¡¨è¾¾é‡å·®ä¸å¤šï¼Œå¹¶ä¸”æ˜¯ä¸åº”è¯¥å æ®ä¸»ä½“åœ°ä½çš„ã€‚ ä¸ºäº†æœ€å¤§ç¨‹åº¦é¿å…ä¸Šé¢ä¸€ç§æˆ–å¤šç§æƒ…å†µçš„å‘ç”Ÿï¼Œåº”è¯¥åœ¨å½’ä¸€åŒ–ä¹‹å‰å»æ‰è¿™äº›ä½è´¨é‡çš„ç»†èƒï¼Œè¿™ä¸ªè¿‡ç¨‹å°±æ˜¯**ç»†èƒçš„è´¨æ§ **\nä½¿ç”¨A. T. L. Lun et al. (2017)çš„å°å‹scRNAæ•°æ®ï¼ˆæœªè¿›è¡ŒQCï¼‰è¿›è¡Œæµ‹è¯•\n#--- loading ---# library(scRNAseq) sce.416b \u003c- LunSpikeInData(which=\"416b\") table(sce.416b$block) 20160113 20160325 96 96 sce.416b$block \u003c- factor(sce.416b$block) sce.416b ## class: SingleCellExperiment ## dim: 46604 192 ## metadata(0): ## assays(1): counts ## rownames(46604): ENSMUSG00000102693 ENSMUSG00000064842 ... ## ENSMUSG00000095742 CBFB-MYH11-mcherry ## rowData names(1): Length ## colnames(192): SLX-9555.N701_S502.C89V9ANXX.s_1.r_1 ## SLX-9555.N701_S503.C89V9ANXX.s_1.r_1 ... ## SLX-11312.N712_S508.H5H5YBBXX.s_8.r_1 ## SLX-11312.N712_S517.H5H5YBBXX.s_8.r_1 ## colData names(9): Source Name cell line ... spike-in addition block ## reducedDimNames(0): ## altExpNames(2): ERCC SIRV QCæŒ‡æ ‡çš„é€‰æ‹© é‰´å®šç»†èƒæ˜¯å¦æ˜¯ä½è´¨é‡çš„ï¼Œéœ€è¦ç”¨åˆ°å‡ ä¸ªæŒ‡æ ‡ã€‚è™½ç„¶ä¸‹é¢è¿™äº›æŒ‡æ ‡æ˜¯ä½¿ç”¨SMART-seq2æ•°æ®è¿›è¡Œå±•ç¤ºçš„ï¼Œä½†ä¾ç„¶é€‚ç”¨äºUMIæ•°æ®ï¼ˆæ¯”å¦‚MARS-seqã€droplet-basedæŠ€æœ¯ï¼‰\næ–‡åº“å¤§å°ï¼ˆlibrary sizeï¼‰ ï¼šæŒ‡çš„æ˜¯æ¯ä¸ªç»†èƒä¸­æ‰€æœ‰åŸºå› çš„è¡¨è¾¾é‡ä¹‹å’Œã€‚ç»†èƒçš„æ–‡åº“å¦‚æœå¾ˆå°ï¼Œè¯´æ˜åœ¨æ–‡åº“åˆ¶å¤‡è¿‡ç¨‹ä¸­å­˜åœ¨RNAçš„æŸå¤±ï¼Œè¦ä¹ˆæ˜¯ç”±äºç»†èƒè£‚è§£ï¼Œè¦ä¹ˆæ˜¯cDNAæ•è·ä¸è¶³å¯¼è‡´åç»­çš„æ‰©å¢é‡å°‘ æ¯ä¸ªç»†èƒä¸­è¡¨è¾¾åŸºå› çš„æ•°ç›®ï¼ˆnumber of expressed features in each cellï¼‰ï¼šæŒ‡çš„æ˜¯ç»†èƒä¸­é0è¡¨è¾¾é‡çš„åŸºå› æ•°é‡ã€‚å¦‚æœç»†èƒä¸­åŸºæœ¬æ²¡æœ‰åŸºå› è¡¨è¾¾ï¼Œå¾ˆå¯èƒ½æ˜¯è½¬å½•æœ¬æ•è·å¤±è´¥ æ¯”å¯¹åˆ°spike-inè½¬å½•æœ¬çš„readsæ¯”ä¾‹ï¼ˆproportion of reads mapped to spike-in transcriptsï¼‰ï¼šè®¡ç®—æ–¹æ³•æ˜¯ï¼šspike-in counts / all features (including spike-ins) for each cellã€‚æ¯ä¸ªç»†èƒéƒ½åº”è¯¥åŠ å…¥ç­‰é‡çš„å¤–æºè½¬å½•æœ¬ï¼ˆspike-inï¼‰ï¼Œå¦‚æœå“ªä¸ªç»†èƒçš„spike-inæ¯”ä¾‹æé«˜äº†ï¼Œè¯´æ˜å®ƒçš„å†…æºRNAå«é‡å‡å°‘ï¼ˆæ¯”å¦‚åœ¨ç»†èƒåˆ†é€‰é˜¶æ®µå‡ºç°çš„ç»†èƒè£‚è§£æˆ–è€…RNAé™è§£ï¼‰ æ¯”å¯¹åˆ°çº¿ç²’ä½“åŸºå› ç»„çš„readsæ¯”ä¾‹ï¼ˆproportion of reads mapped to genes in the mitochondrial genomeï¼‰ ï¼šå¦‚æœæ²¡æœ‰spike-inï¼Œé‚£ä¹ˆä½¿ç”¨çº¿ç²’ä½“æŒ‡æ ‡ä¹Ÿæ˜¯èƒ½è¯´æ˜é—®é¢˜çš„(Islam et al. 2014; Ilicic et al. 2016)ã€‚æ¯”å¯¹åˆ°çº¿ç²’ä½“çš„readså¢å¤šï¼Œè¯´æ˜ç»†èƒè´¨ä¸­çš„RNAå‡å°‘ï¼Œå¯èƒ½å­˜åœ¨ç»†èƒç©¿å­”çš„æƒ…å†µï¼Œè€Œè¿™ä¸ªå­”çš„å¤§å°ï¼Œå¯èƒ½åªæ˜¯å°†ç»†èƒè´¨ä¸­å­˜åœ¨çš„mRNAæµå‡ºå»ï¼Œä½†çº¿ç²’ä½“çš„ä½“ç§¯è¶…è¿‡äº†å­”çš„å¤§å°ï¼Œæ‰€ä»¥è¿˜ç•™åœ¨äº†ç»†èƒä¸­ï¼Œé€ æˆä¸€å®šç¨‹åº¦çš„å¯Œé›†ï¼Œå¯¼è‡´çº¿ç²’ä½“RNAå æ¯”å‡é«˜ã€‚ å¯¹äºæ¯ä¸ªç»†èƒï¼Œå¯ä»¥ç”¨scateråŒ…çš„perCellQCMetrics()å‡½æ•°è¿›è¡Œè®¡ç®—ï¼Œå…¶ä¸­sumè¿™ä¸€åˆ—è¡¨ç¤ºæ¯ä¸ªç»†èƒçš„æ–‡åº“å¤§å°ï¼›detectedè¿™ä¸€åˆ—è¡¨ç¤ºæ£€æµ‹åˆ°çš„åŸºå› æ•°é‡ï¼›subsets_Mito_percentè¿™ä¸€åˆ—è¡¨ç¤ºæ¯”å¯¹åˆ°çº¿ç²’ä½“åŸºå› ç»„çš„readså æ¯”ï¼›altexps_ERCC_percentè¡¨ç¤ºæ¯”å¯¹åˆ°ERCC spike-inçš„readså æ¯”\n# Retrieving the mitochondrial transcripts using genomic locations included in # the row-level annotation for the SingleCellExperiment. location \u003c- rowRanges(sce.416b) is.mito \u003c- any(seqnames(location)==\"MT\") # ALTERNATIVELY: using resources in AnnotationHub to retrieve chromosomal # locations given the Ensembl IDs; this should yield the same result. library(AnnotationHub) ens.mm.v97 \u003c- AnnotationHub()[[\"AH73905\"]] chr.loc \u003c- mapIds(ens.mm.v97, keys=rownames(sce.416b), keytype=\"GENEID\", column=\"SEQNAME\") is.mito.alt \u003c- which(chr.loc==\"MT\") library(scater) df \u003c- perCellQCMetrics(sce.416b, subsets=list(Mito=is.mito)) df ## DataFrame with 192 rows and 16 columns ## sum detected percent_top_50 percent_top_100 percent_top_200 ## ## 1 865936 7618 26.7218 32.2773 39.7208 ## 2 1076277 7521 29.4043 35.0354 42.2581 ## 3 1180138 8306 27.3454 32.4770 39.3296 ## 4 1342593 8143 35.8092 40.2666 46.2460 ## 5 1668311 7154 34.1198 39.0901 45.6660 ## ... ... ... ... ... ... ## 188 776622 8174 45.9362 49.7010 54.6101 ## 189 1299950 8956 38.0829 42.8930 49.0622 ## 190 1800696 9530 30.6675 35.5839 41.8550 ## 191 46731 6649 32.2998 37.9149 44.5999 ## 192 1866692 10964 26.6632 31.2584 37.5608 ## percent_top_500 subsets_Mito_sum subsets_Mito_detected subsets_Mito_percent ## ## 1 52.9038 78790 20 9.09882 ## 2 55.7454 98613 20 9.16242 ## 3 51.9337 100341 19 8.50248 ## 4 57.1210 104882 20 7.81190 ## 5 58.2004 129559 22 7.76588 ## ... ... ... ... ... ## 188 64.4249 48126 20 6.19684 ## 189 60.6675 112225 25 8.63302 ## 190 53.6781 135693 23 7.53559 ## 191 56.5235 3505 16 7.50037 ## 192 48.9489 150375 29 8.05569 ## altexps_ERCC_sum altexps_ERCC_detected altexps_ERCC_percent ## ## 1 65278 39 6.80658 ## 2 74748 40 6.28030 ## 3 60878 42 4.78949 ## 4 60073 42 4.18567 ## 5 136810 44 7.28887 ## ... ... ... ... ## 188 61575 39 7.17620 ## 189 94982 41 6.65764 ## 190 113707 40 5.81467 ## 191 7580 44 13.48898 ## 192 48664 39 2.51930 ## altexps_SIRV_sum altexps_SIRV_detected altexps_SIRV_percent total ## ## 1 27828 7 2.90165 959042 ## 2 39173 7 3.29130 1190198 ## 3 30058 7 2.36477 1271074 ## 4 32542 7 2.26741 1435208 ## 5 71850 7 3.82798 1876971 ## ... ... ... ... ... ## 188 19848 7 2.313165 858045 ## 189 31729 7 2.224004 1426661 ## 190 41116 7 2.102562 1955519 ## 191 1883 7 3.350892 56194 ## 192 16289 7 0.843271 1931645 å¦å¤–ï¼Œè¿˜å¯ä»¥ä½¿ç”¨addPerCellQC()ï¼Œå®ƒä¼šæŠŠæ¯ä¸ªç»†èƒçš„QCæŒ‡æ ‡åŠ åˆ°SingleCellExperimentå¯¹è±¡çš„colDataä¸­ï¼Œæ–¹ä¾¿åé¢è°ƒå–\nsce.416b \u003c- addPerCellQC(sce.416b, subsets=list(Mito=is.mito)) colnames(colData(sce.416b)) ## [1] \"Source Name\" \"cell line\" ## [3] \"cell type\" \"single cell well quality\" ## [5] \"genotype\" \"phenotype\" ## [7] \"strain\" \"spike-in addition\" ## [9] \"block\" \"sum\" ## [11] \"detected\" \"percent_top_50\" ## [13] \"percent_top_100\" \"percent_top_200\" ## [15] \"percent_top_500\" \"subsets_Mito_sum\" ## [17] \"subsets_Mito_detected\" \"subsets_Mito_percent\" ## [19] \"altexps_ERCC_sum\" \"altexps_ERCC_detected\" ## [21] \"altexps_ERCC_percent\" \"altexps_SIRV_sum\" ## [23] \"altexps_SIRV_detected\" \"altexps_SIRV_percent\" ## [25] \"total\" å½“ç„¶ï¼Œè¿™é‡ŒåšQCç»Ÿè®¡çš„å‰æå‡è®¾æ˜¯ï¼šæ¯ä¸ªç»†èƒçš„qcæŒ‡æ ‡éƒ½æ˜¯ç‹¬ç«‹äºç”Ÿç‰©å­¦çŠ¶æ€çš„ã€‚ä¹Ÿå°±æ˜¯è¯´ï¼ŒqcæŒ‡æ ‡ä¸ä¼šé‚£ä¹ˆæ™ºèƒ½åœ°è¯†åˆ«ç»†èƒç±»å‹ç„¶åè¿›è¡Œåˆ¤æ–­ã€‚å®ƒä¼šè®¤ä¸ºï¼ˆå¦‚æ–‡åº“å¤ªå°ã€é«˜çº¿ç²’ä½“å æ¯”ï¼‰éƒ½æ˜¯ç”±æŠ€æœ¯è¯¯å·®å¼•èµ·çš„ï¼Œè€Œéç”Ÿç‰©å› ç´ ã€‚ä½†æ˜¯æœ‰ä¸€ä¸ªé—®é¢˜ï¼Œå¦‚æœæŸäº›ç»†èƒç±»å‹æœ¬èº«çš„RNAå«é‡å°±å¾ˆä½ï¼Œæˆ–è€…å®ƒä»¬æœ¬æ¥å°±å«æœ‰å¾ˆå¤šçš„çº¿ç²’ä½“è½¬å½•æœ¬å‘¢ï¼Ÿå†æ ¹æ®è¿™ä¸ªæŒ‡æ ‡è¿›è¡Œè¿‡æ»¤ï¼Œä¼šä¸ä¼šæŸå¤±ä¸€äº›ç»†èƒç±»å‹å‘¢ï¼Ÿ\nè¯†åˆ«ä½è´¨é‡ç»†èƒ 1. ä½¿ç”¨å›ºå®šçš„é˜ˆå€¼ è¯†åˆ«ä½è´¨é‡ç»†èƒæœ€ç®€å•æ–¹æ³•æ˜¯åœ¨QCåº¦é‡ä¸Šåº”ç”¨é˜ˆå€¼ã€‚ä¾‹å¦‚è®¾å®šæ–‡åº“ä½äº10ä¸‡readsçš„ç»†èƒæ˜¯ä½è´¨é‡çš„ï¼Œæˆ–è€…è¡¨è¾¾åŸºå› æ•°é‡å°‘äº5000ä¸ªï¼Œspike-inæˆ–çº¿ç²’ä½“å æ¯”é«˜äº10%ã€‚\nqc.lib \u003c- df$sum \u003c 1e5 qc.nexprs \u003c- df$detected \u003c 5e3 qc.spike \u003c- df$altexps_ERCC_percent \u003e 10 qc.mito \u003c- df$subsets_Mito_percent \u003e 10 discard \u003c- qc.lib | qc.nexprs | qc.spike | qc.mito # Summarize the number of cells removed for each reason. DataFrame(LibSize=sum(qc.lib), NExprs=sum(qc.nexprs), SpikeProp=sum(qc.spike), MitoProp=sum(qc.mito), Total=sum(discard)) ## DataFrame with 1 row and 5 columns ## LibSize NExprs SpikeProp MitoProp Total ## ## 1 3 0 19 14 33 è™½ç„¶çœ‹èµ·æ¥ç®€å•ï¼Œä½†ä½¿ç”¨è¿™ç§æ–¹æ³•éœ€è¦ä¸°å¯Œçš„ç»éªŒï¼Œäº†è§£å®éªŒè®¾è®¡å’Œç»†èƒçŠ¶æ€ï¼›å¦å¤–å³ä½¿ä½¿ç”¨åŒä¸€ç§æ–‡åº“åˆ¶å¤‡æ–¹æ³•ï¼Œä½†ç”±äºç»†èƒæ•è·æ•ˆç‡å’Œæµ‹åºæ·±åº¦çš„ä¸åŒï¼Œè¿™ä¸ªé˜ˆå€¼ä¾ç„¶éœ€è¦é€‚æ—¶è°ƒæ•´ã€‚å› æ­¤å¯¹äºç ”ç©¶äººå‘˜è¦æ±‚å¾ˆé«˜ã€‚\n2. ä½¿ç”¨ç›¸å¯¹é˜ˆå€¼ é‰´å®šç¦»ç¾¤ç‚¹ ä¸ºäº†è·å¾—ç›¸å¯¹é˜ˆå€¼ï¼Œå…ˆå‡è®¾å¤§éƒ¨åˆ†ç»†èƒéƒ½æ˜¯é«˜è´¨é‡çš„ï¼Œç„¶åå»æ‰¾ç¦»ç¾¤ç‚¹ä½œä¸ºä½è´¨é‡ã€‚é‚£ä¹ˆæŒ‰ç…§ä»€ä¹ˆæ–¹æ³•æ‰¾ç¦»ç¾¤ç‚¹å‘¢ï¼Ÿå¸¸ç”¨çš„ä¸€ä¸ªå‡½æ•°isOutlierä½¿ç”¨çš„æ˜¯MADæŒ‡æ ‡ï¼ˆç»å¯¹ä¸­ä½å·®æ¥ä¼°è®¡æ–¹å·®,å…ˆè®¡ç®—å‡ºæ•°æ®ä¸å®ƒä»¬çš„ä¸­ä½æ•°ä¹‹é—´çš„åå·®ï¼Œç„¶åè¿™äº›åå·®çš„ç»å¯¹å€¼çš„ä¸­ä½æ•°å°±æ˜¯MADï¼Œmedian absolute deviationï¼‰ã€‚å¦‚æœè¶…è¿‡ä¸­ä½æ•°3å€MADçš„å€¼å°±æ˜¯ç¦»ç¾¤å€¼ã€‚\nä½¿ç”¨isOutlieræ—¶ï¼Œå¦‚æœè¦ç›¸å‡ï¼ˆä¾‹å¦‚ï¼šdf$sum - 3* MADï¼‰ï¼Œå°±ç”¨type=\"lower\"ï¼Œæ­¤æ—¶ä¸€èˆ¬è¿˜è¦åšä¸ªlogè½¬æ¢log=TRUE ï¼Œä¿è¯å¾—åˆ°çš„ç»“æœä¸æ˜¯è´Ÿæ•°\nqc.lib2 \u003c- isOutlier(df$sum, log=TRUE, type=\"lower\") qc.nexprs2 \u003c- isOutlier(df$detected, log=TRUE, type=\"lower\") qc.spike2 \u003c- isOutlier(df$altexps_ERCC_percent, type=\"higher\") qc.mito2 \u003c- isOutlier(df$subsets_Mito_percent, type=\"higher\") attr(qc.lib2, \"thresholds\") lower higher 434082.9 Inf attr(qc.nexprs2, \"thresholds\") lower higher 5231.468 Inf attr(qc.spike2, \"thresholds\") lower higher -Inf 14.15371 attr(qc.mito2, \"thresholds\") lower higher -Inf 11.91734 ç”¨ç›¸å¯¹é˜ˆå€¼è¿‡æ»¤çš„ç»†èƒæ•°é‡ç»Ÿè®¡ï¼š\ndiscard2 \u003c- qc.lib2 | qc.nexprs2 | qc.spike2 | qc.mito2 # Summarize the number of cells removed for each reason. DataFrame(LibSize=sum(qc.lib2), NExprs=sum(qc.nexprs2), SpikeProp=sum(qc.spike2), MitoProp=sum(qc.mito2), Total=sum(discard2)) ## DataFrame with 1 row and 5 columns ## LibSize NExprs SpikeProp MitoProp Total ## ## 1 4 0 1 2 6 é™¤æ­¤ä»¥å¤–ï¼Œè¿˜æœ‰ä¸€ç§æ›´å¿«çš„è®¡ç®—æ–¹æ³•ï¼Œä¸€æ­¥æ•´åˆäº†ä¸Šé¢çš„æ“ä½œï¼š\nreasons \u003c- quickPerCellQC(df, percent_subsets=c(\"subsets_Mito_percent\", \"altexps_ERCC_percent\")) colSums(as.matrix(reasons)) ## low_lib_size low_n_features high_subsets_Mito_percent ## 4 0 2 ## high_altexps_ERCC_percent discard ## 1 6 ä½¿ç”¨â€ç›¸å¯¹â€œçš„é˜ˆå€¼ä¸€ä¸ªå¥½å¤„å°±æ˜¯å¯ä»¥æ ¹æ®æµ‹åºæ·±åº¦ã€cDNAæ•è·æ•ˆç‡ã€çº¿ç²’ä½“å«é‡ç­‰ç­‰è¿›è¡Œé˜ˆå€¼çš„è°ƒæ•´ï¼Œåœ¨ç»éªŒä¸æ˜¯è¶³å¤Ÿä¸°å¯Œçš„æ—¶å€™ï¼Œå¯ä»¥è¾…åŠ©åˆ¤æ–­ã€‚ä½†ä»éœ€è¦æ³¨æ„çš„æ˜¯ï¼Œä½¿ç”¨ç›¸å¯¹é˜ˆå€¼æ˜¯æœ‰å‰æçš„ï¼Œé‚£å°±æ˜¯ï¼šè®¤ä¸ºå¤§éƒ¨åˆ†ç»†èƒéƒ½æ˜¯é«˜è´¨é‡çš„\nç¦»ç¾¤ç‚¹æ£€æµ‹çš„å‡è®¾ ç¦»ç¾¤ç‚¹çš„æ£€æµ‹çš„å‡è®¾æ˜¯å¤§éƒ¨åˆ†ç»†èƒçš„è´¨é‡éƒ½ä¸é”™ã€‚è¿™ä¸€ç‚¹å‡è®¾å¯ä»¥é€šè¿‡å®éªŒå»éªŒè¯ï¼ˆæ¯”å¦‚è‚‰çœ¼æ£€æŸ¥ç»†èƒå®Œæ•´æ€§ï¼‰ã€‚ä½†å½“å¤§éƒ¨åˆ†ç»†èƒè´¨é‡éƒ½å¾ˆä½ï¼Œä½¿ç”¨ç›¸å¯¹é˜ˆå€¼ç»“æœå°±å¯¹å¤§é‡çš„ä½è´¨é‡ç»†èƒæ— è®¡å¯æ–½ã€‚å› ä¸ºå®ƒä½¿ç”¨äº†MADå€¼ï¼Œå’Œä¸­ä½æ•°æœ‰å…³ç³»ï¼Œé‚£ä¹ˆå¯ä»¥è¯•æƒ³ï¼šå¦‚æœä¸€å †æ•°éƒ½ä¸åˆæ ¼ï¼Œé‚£ä¹ˆå®ƒä»¬çš„ä¸­ä½æ•°ä¹Ÿä¸åˆæ ¼ï¼Œå› æ­¤åŸæ¥æ­£ç¡®çš„å€¼ï¼Œå…¶å®åœ¨è¿™ç¾¤ä¸åˆæ ¼çš„æ•°å€¼ä¸­ï¼Œå°±æ˜¯â€œç¦»ç¾¤â€çš„ã€‚ å¦ä¸€ä¸ªå‡è®¾å°±æ˜¯ï¼šæ¯ä¸ªç»†èƒçš„qcæŒ‡æ ‡éƒ½æ˜¯ç‹¬ç«‹äºç”Ÿç‰©å­¦çŠ¶æ€çš„ã€‚ä¹Ÿå°±æ˜¯è¯´ï¼ŒqcæŒ‡æ ‡ä¸ä¼šé‚£ä¹ˆæ™ºèƒ½åœ°è¯†åˆ«ç»†èƒç±»å‹ç„¶åè¿›è¡Œåˆ¤æ–­ã€‚åœ¨å¼‚è´¨æ€§å¾ˆé«˜çš„ç»„ç»‡ä¸­ï¼Œ å°±æ˜¯å­˜åœ¨å†…æºRNAå«é‡ä½ï¼Œè€Œçº¿ç²’ä½“åŸºå› å æ¯”é«˜çš„ç»†èƒã€‚å¦‚æœä½¿ç”¨â€ä¸€åˆ€åˆ‡â€œçš„å›ºå®šé˜ˆå€¼ï¼Œå®ƒä»¬å°±å¾ˆå¯èƒ½ä¼šè¢«å½“æˆç¦»ç¾¤ç‚¹è¢«è¿‡æ»¤ã€‚è€Œæ˜¯ç”¨MADè®¡ç®—æ–¹æ³•æ£€æµ‹çš„ç»“æœå¯èƒ½å°±æ˜¯ï¼šè™½ç„¶ä¸€å †ç»†èƒçš„æŸä¸ªqcæŒ‡æ ‡å·®å¼‚å¾ˆå¤§ï¼Œä½†ä¸­ä½æ•°ä¹Ÿåœ¨å˜ï¼Œéšä¹‹å˜åŒ–çš„è¿˜æœ‰MADå€¼ï¼Œè¿™æ ·æœ€åçš„ç»“æœä¸è‡³äºå’ŒçœŸå®ç”Ÿç‰©å­¦æƒ…å†µå·®å¤ªå¤š è€ƒè™‘å®éªŒçš„å› ç´  å¾ˆå¤šå¤§å‹çš„å®éªŒéƒ½åŒ…å«å¤šä¸ªç»†èƒç³»ï¼Œè€Œä¸”å¯èƒ½é‡‡ç”¨çš„å®éªŒæ–¹æ³•ä¸åŒï¼ˆæ¯”å¦‚é€‰ç”¨ä¸åŒçš„æµ‹åºæ·±åº¦ï¼‰ï¼Œè¿™å°±äº§ç”Ÿäº†å®éªŒçš„ä¸åŒæ‰¹æ¬¡ã€‚è¿™ç§æƒ…å†µä¸‹ï¼Œ åº”è¯¥å¯¹æ¯ä¸ªæ‰¹æ¬¡åˆ†åˆ«è¿›è¡Œæ£€æµ‹ã€‚\nå¦‚æœæ¯ä¸ªæ‰¹æ¬¡éƒ½æ˜¯ä¸€ä¸ªSingleCellExperimentå¯¹è±¡ï¼Œé‚£ä¹ˆisOutlier()å‡½æ•°å¯ä»¥æŒ‰ä¸Šé¢çš„æ–¹æ³•ç›´æ¥ä½¿ç”¨ï¼›ä½†æ˜¯å¦‚æœä¸åŒæ‰¹æ¬¡çš„ç»†èƒå·²ç»æ··åˆæˆä¸€æ•´ä¸ªSingleCellExperimentå¯¹è±¡ï¼Œé‚£ä¹ˆbatch=è¿™ä¸ªå‚æ•°å°±æ´¾ä¸Šç”¨åœºäº†ã€‚\nåŒæ ·ä»¥è¿™ä¸ª416Bæ•°æ®é›†ä¸ºä¾‹ï¼Œä»–åŒ…å«äº†ä¸¤ç§ä¸åŒçš„å®éªŒç±»å‹ã€‚ç„¶åæˆ‘ä»¬å°±å¯ä»¥ä½¿ç”¨batch=å‚æ•°å»è¿›è¡Œè´¨æ§ã€‚\nbatch \u003c- paste0(sce.416b$phenotype, \"-\", sce.416b$Plate) batch.reasons \u003c- quickPerCellQC(df, percent_subsets=c(\"subsets_Mito_percent\", \"altexps_ERCC_percent\"), batch=batch) colSums(as.matrix(batch.reasons)) ## low_lib_size low_n_features high_subsets_Mito_percent ## 4 2 2 ## high_altexps_ERCC_percent discard ## 4 7 ä½†æ˜¯ï¼Œbatchå‚æ•°ä¸æ˜¯ä¸‡èƒ½çš„ï¼Œä¹‹å‰ä¹Ÿè¯´è¿‡ï¼Œè¿™ç§ç›¸å¯¹é˜ˆå€¼éœ€è¦ä¸€ä¸ªå‡è®¾ï¼šæ¯ä¸ªæ‰¹æ¬¡çš„å¤§éƒ¨åˆ†ç»†èƒéƒ½æ˜¯é«˜è´¨é‡çš„ã€‚å½“æŸä¸ªæ‰¹æ¬¡çš„ç»†èƒæ•´ä½“è´¨é‡åä½ï¼Œç¦»ç¾¤ç‚¹æ£€æµ‹å¾ˆæœ‰å¯èƒ½å¤±è´¥\nä¾‹å¦‚ï¼Œåœ¨Grun et al. (2016)çš„æ•°æ®é›†ä¸­æœ‰ä¸¤ä¸ªdonorçš„å®éªŒæ˜¯å¤±è´¥çš„ã€‚å®ƒä»¬çš„ERCCå«é‡ç›¸å¯¹å…¶ä»–æ‰¹æ¬¡é«˜ï¼Œå¢åŠ äº†ä¸­ä½æ•°å’ŒMADå€¼ï¼Œå¯¼è‡´è¿‡æ»¤ä½è´¨é‡ç»†èƒå¤±è´¥ã€‚å› æ­¤è¿™ç§æƒ…å†µä¸‹ï¼Œå¯ä»¥å…ˆç®—å…¶ä»–å‡ ä¸ªæ‰¹æ¬¡çš„ä¸­ä½æ•°å’ŒMADå€¼ï¼Œç„¶åå‚è€ƒè¿™äº›å€¼å»å¯¹æœ‰é—®é¢˜çš„æ‰¹æ¬¡è¿›è¡Œè¿‡æ»¤ã€‚\nlibrary(scRNAseq) sce.grun \u003c- GrunPancreasData() sce.grun \u003c- addPerCellQC(sce.grun) # First attempt with batch-specific thresholds. discard.ercc \u003c- isOutlier(sce.grun$altexps_ERCC_percent, type=\"higher\", batch=sce.grun$donor) with.blocking \u003c- plotColData(sce.grun, x=\"donor\", y=\"altexps_ERCC_percent\", colour_by=I(discard.ercc)) # Second attempt, sharing information across batches # to avoid dramatically different thresholds for unusual batches. discard.ercc2 \u003c- isOutlier(sce.grun$altexps_ERCC_percent, type=\"higher\", batch=sce.grun$donor, subset=sce.grun$donor %in% c(\"D17\", \"D2\", \"D7\")) without.blocking \u003c- plotColData(sce.grun, x=\"donor\", y=\"altexps_ERCC_percent\", colour_by=I(discard.ercc2)) gridExtra::grid.arrange(with.blocking, without.blocking, ncol=2) æ³¨ï¼šå¯ä»¥çœ‹åˆ°ï¼Œå·¦å›¾æ˜¯æŒ‰ç…§æ¯ä¸ªæ‰¹æ¬¡åˆ†åˆ«é‰´å®šçš„ç¦»ç¾¤ç‚¹ï¼›å³å›¾æ˜¯ç”¨è´¨é‡å¥½çš„æ‰¹æ¬¡è®¡ç®—çš„é˜ˆå€¼ï¼Œç„¶åè¿ç”¨åˆ°å·®çš„æ‰¹æ¬¡ä¸Šçš„ç»“æœ\nä¸ºäº†é‰´åˆ«æœ‰é—®é¢˜çš„æ‰¹æ¬¡ï¼Œå¯ä»¥å…ˆå°†æ¯ä¸ªæ‰¹æ¬¡åˆ†åˆ«è®¡ç®—ç»“æœï¼Œç„¶åæ¯”è¾ƒå®ƒä»¬çš„é˜ˆå€¼ï¼Œå¦‚æœæ¯”åŒç±»æ‰¹æ¬¡è¶…å‡ºå¤ªå¤šï¼Œå°±è¦è­¦è§‰ã€‚\nercc.thresholds \u003c- attr(discard.ercc, \"thresholds\")[\"higher\",] ercc.thresholds ## D10 D17 D2 D3 D7 ## 73.610696 7.599947 6.010975 113.105828 15.216956 names(ercc.thresholds)[isOutlier(ercc.thresholds, type=\"higher\")] ## [1] \"D10\" \"D3\" å¯ä»¥çœ‹åˆ°D10ã€D3çš„é˜ˆå€¼å°±è¶…è¿‡å…¶ä»–æ‰¹æ¬¡å¾ˆå¤š\nä½†æ˜¯è¿™ä¹ˆåšçš„å‰æéƒ½æ˜¯ï¼šæˆ‘ä»¬è®¤ä¸ºæ‰¹æ¬¡ä¸­çš„ç»†èƒè´¨é‡æ•´ä½“è¿˜ä¸é”™ã€‚å¦‚æœæˆ‘ä»¬ä¿è¯ä¸äº†ç»†èƒè´¨é‡ï¼Œé‚£ä¹ˆè¿™ç§è®¡ç®—ç›¸å¯¹é˜ˆå€¼çš„æ–¹æ³•å°±ä¸æˆç«‹ï¼Œè¿˜æ˜¯è¦ä½¿ç”¨ç»å¯¹é˜ˆå€¼ï¼Œæ‰‹åŠ¨å»é™¤ã€‚\n3. å…¶ä»–æ–¹æ³• å¦ä¸€ä¸ªç­–ç•¥æ˜¯æ ¹æ®æ¯ä¸ªç»†èƒçš„QCæŒ‡æ ‡æ¥åœ¨é«˜ç»´ç©ºé—´ä¸­è¯†åˆ«å¼‚å¸¸å€¼ã€‚åˆ©ç”¨robustbase åŒ…ï¼Œå°†ç»†èƒæ”¾åœ¨é«˜ç»´ç©ºé—´ï¼Œæ ¹æ®ä»–ä»¬çš„QCæŒ‡æ ‡ï¼ˆæ–‡åº“å¤§å°ã€è¡¨è¾¾åŸºå› æ•°ã€spike-inæ¯”ä¾‹ã€çº¿ç²’ä½“æ¯”ä¾‹ï¼‰ï¼Œç„¶åä½¿ç”¨isOutlier()æ¥è¯†åˆ«è¡¨ç°å‡ºå¼‚å¸¸é«˜outlylieræ°´å¹³çš„ä½è´¨é‡ç»†èƒ\nstats \u003c- cbind(log10(df$sum), log10(df$detected), df$subsets_Mito_percent, df$altexps_ERCC_percent) library(robustbase) outlying \u003c- adjOutlyingness(stats, only.outlyingness = TRUE) multi.outlier \u003c- isOutlier(outlying, type = \"higher\") summary(multi.outlier) ## Mode FALSE TRUE ## logical 182 10 é™¤æ­¤ä»¥å¤–ï¼Œæœ‰æ—¶è¿˜å¯ä»¥æ ¹æ®åŸºå› è¡¨è¾¾é‡è¿›è¡Œè¿‡æ»¤ï¼Œä¸è¿‡åœ¨bulkè½¬å½•ç»„ä¸­ç”¨çš„æ¯”è¾ƒå¤šï¼Œä½†æ˜¯åœ¨scRNAä¸­è¿™æ ·æ“ä½œå¯èƒ½ä¼šæŸå¤±ä¸€äº›æœ¬èº«æ•°é‡å°±æ¯”è¾ƒå°‘çš„é«˜è´¨é‡ç»†èƒç¾¤ä½“ï¼ˆæ¯”å¦‚ä¸€äº›ç½•è§ç»†èƒï¼Œæœ¬èº«åŸºå› è¡¨è¾¾é‡å°±ä¸æ˜¯å¾ˆé«˜ï¼‰\nç”»å›¾æ£€æŸ¥ æ£€æŸ¥QCåº¦é‡çš„åˆ†å¸ƒä»¥ç¡®å®šå¯èƒ½çš„é—®é¢˜æ˜¯ä¸€ä¸ªå¾ˆå¥½çš„å®è·µã€‚åœ¨æœ€ç†æƒ³çš„æƒ…å†µä¸‹ï¼Œæˆ‘ä»¬ä¼šçœ‹åˆ°æ­£æ€åˆ†å¸ƒï¼Œå¯ä»¥è¯æ˜åœ¨ç¦»ç¾¤å€¼æ£€æµ‹ä¸­ä½¿ç”¨çš„3 MADé˜ˆå€¼æ˜¯åˆç†çš„ã€‚å¦ä¸€ç§æ¨¡å¼ä¸‹çš„å¤§é‡ç»†èƒè¡¨æ˜QCæŒ‡æ ‡å¯èƒ½ä¸æŸäº›ç”Ÿç‰©çŠ¶æ€ç›¸å…³ï¼Œå¯èƒ½å¯¼è‡´è¿‡æ»¤è¿‡ç¨‹ä¸­ä¸åŒç»†èƒç±»å‹çš„ä¸¢å¤±;æˆ–è€…æœ‰ä¸ä¸€è‡´çš„åº“å‡†å¤‡ä¸ºä¸€ä¸ªå­é›†çš„ç»†èƒï¼Œä¸€ä¸ªéç½•è§çš„ç°è±¡åœ¨æ¿çš„åè®®ã€‚\n# æŠŠQCæŒ‡æ ‡å’ŒåŸæ¥çš„sce.416bç»†èƒä¿¡æ¯æ•´åˆèµ·æ¥ colData(sce.416b) \u003c- cbind(colData(sce.416b), df) # ä¿®æ”¹ä¸€ä¸‹æ•´åˆåçš„ä¿¡æ¯ sce.416b$block \u003c- factor(sce.416b$block) sce.416b$phenotype \u003c- ifelse(grepl(\"induced\", sce.416b$phenotype), \"induced\", \"wild type\") sce.416b$discard \u003c- reasons$discard # ç»˜å›¾ gridExtra::grid.arrange( plotColData(sce.416b, x=\"block\", y=\"sum\", colour_by=\"discard\", other_fields=\"phenotype\") + facet_wrap(~phenotype) + scale_y_log10() + ggtitle(\"Total count\"), plotColData(sce.416b, x=\"block\", y=\"detected\", colour_by=\"discard\", other_fields=\"phenotype\") + facet_wrap(~phenotype) + scale_y_log10() + ggtitle(\"Detected features\"), plotColData(sce.416b, x=\"block\", y=\"subsets_Mito_percent\", colour_by=\"discard\", other_fields=\"phenotype\") + facet_wrap(~phenotype) + ggtitle(\"Mito percent\"), plotColData(sce.416b, x=\"block\", y=\"altexps_ERCC_percent\", colour_by=\"discard\", other_fields=\"phenotype\") + facet_wrap(~phenotype) + ggtitle(\"ERCC percent\"), ncol=1 ) å±•ç¤ºçš„æ˜¯ä¸åŒæ‰¹æ¬¡çš„QCæŒ‡æ ‡\nå¦ä¸€ç§æœ‰ç”¨çš„è¯Šæ–­æ–¹æ³•æ˜¯ç»˜åˆ¶ç›¸å¯¹äºå…¶ä»–QCæŒ‡æ ‡çš„çº¿ç²’ä½“è®¡æ•°æ¯”ä¾‹å›¾ã€‚\nä¸ºäº†ç¡®ä¿ä¸å­˜åœ¨è¿™æ ·çš„ç»†èƒï¼šè™½ç„¶ç»†èƒæ–‡åº“å¤§ï¼Œä½†å®ƒçš„çº¿ç²’ä½“å æ¯”ä¹Ÿé«˜ã€‚å¦å¤–ä¹Ÿæ˜¯ä¸ºäº†é¿å…æ„å¤–è¿‡æ»¤æ‰æœ¬æ¥æ˜¯é«˜è´¨é‡ä½†åŒæ—¶å…·æœ‰é«˜ä»£è°¢æ´»æ€§ï¼ˆå³é«˜çº¿ç²’ä½“å æ¯”ï¼‰çš„ç»†èƒï¼ˆå¦‚è‚è„ç»†èƒï¼‰\n======æœªå®Œ\né’ˆå¯¹Dropletæ•°æ®çš„ç»†èƒåˆ¤æ–­ èƒŒæ™¯ ç”±äºè¿™ç§å»ºåº“æ–¹æ³•çš„ç‹¬ç‰¹æ€§ï¼Œæˆ‘ä»¬æ²¡åŠæ³•äº‹å…ˆçŸ¥é“æŸä¸€ä¸ªç»†èƒæ–‡åº“ï¼ˆæ¯”å¦‚ä¸€ä¸ªcell barcodeï¼‰æ˜¯çœŸæ­£åŒ…å«ä¸€ä¸ªç»†èƒè¿˜æ˜¯åªæ˜¯ä¸€ä¸ªç©ºçš„æ¶²æ»´ï¼ˆdropletï¼‰ã€‚å› æ­¤ï¼Œç¬¬ä¸€æ­¥æ˜¯éœ€è¦æ ¹æ®å¾—åˆ°çš„è¡¨è¾¾é‡ä¿¡æ¯ï¼Œæ¥æ¨æ–­æ¶²æ»´æ˜¯ä¸æ˜¯ç©ºçš„ã€‚ä½†ä»…ä»…æ ¹æ®è¡¨è¾¾é‡åˆ¤æ–­è¿˜æ˜¯ä¸å¤ªé è°±ï¼Œå› ä¸ºè¿˜æœ‰å¯èƒ½åœ¨ç©ºçš„æ¶²æ»´ä¸­ä¾ç„¶åŒ…å«å¤–æºRNAï¼Œæœ€åçš„ç»†èƒæ–‡åº“ä¾æ—§ä¸ä¸º0ï¼Œä½†ç¡®å®ä¸åŒ…å«ç»†èƒã€‚\nè¿™é‡Œä¸ºäº†è¯´æ˜è¿™ä¸ªé—®é¢˜ï¼Œä½¿ç”¨äº†ä¸€ä¸ªæœªè¿‡æ»¤çš„10X PBMCæ•°æ®\n# æ•°æ®ä¸‹è½½ library(BiocFileCache) bfc \u003c- BiocFileCache(\"raw_data\", ask = FALSE) raw.path \u003c- bfcrpath(bfc, file.path(\"http://cf.10xgenomics.com/samples\", \"cell-exp/2.1.0/pbmc4k/pbmc4k_raw_gene_bc_matrices.tar.gz\")) # è§£å‹æ•°æ® untar(raw.path, exdir=file.path(tempdir(), \"pbmc4k\")) library(DropletUtils) library(Matrix) fname \u003c- file.path(tempdir(), \"pbmc4k/raw_gene_bc_matrices/GRCh38\") sce.pbmc \u003c- read10xCounts(fname, col.names=TRUE) sce.pbmc ## class: SingleCellExperiment ## dim: 33694 737280 ## metadata(1): Samples ## assays(1): counts ## rownames(33694): ENSG00000243485 ENSG00000237613 ... ENSG00000277475 ## ENSG00000268674 ## rowData names(2): ID Symbol ## colnames(737280): AAACCTGAGAAACCAT-1 AAACCTGAGAAACCGC-1 ... ## TTTGTCATCTTTAGTC-1 TTTGTCATCTTTCCTC-1 ## colData names(2): Sample Barcode ## reducedDimNames(0): ## altExpNames(0): æ•´ä½“è§‚å¯Ÿä¸åŒçš„barcodesï¼ˆä¸ä¸€å®šéƒ½æ˜¯çœŸçš„ç»†èƒï¼‰çš„æ–‡åº“å¤§å°åˆ†å¸ƒï¼š\nbcrank \u003c- barcodeRanks(counts(sce.pbmc)) # Only showing unique points for plotting speed. uniq \u003c- !duplicated(bcrank$rank) plot(bcrank$rank[uniq], bcrank$total[uniq], log=\"xy\", xlab=\"Rank\", ylab=\"Total UMI count\", cex.lab=1.2) abline(h=metadata(bcrank)$inflection, col=\"darkgreen\", lty=2) abline(h=metadata(bcrank)$knee, col=\"dodgerblue\", lty=2) legend(\"bottomleft\", legend=c(\"Inflection\", \"Knee\"), col=c(\"darkgreen\", \"dodgerblue\"), lty=2, cex=1.2) çœ‹åˆ°å„ä¸ªbarcodesçš„æ–‡åº“å¤§å°æœ‰é«˜æœ‰ä½ï¼Œå¹¶ä¸”ç›¸å·®è¾ƒå¤§ï¼Œå› æ­¤å¯èƒ½å¯¹åº”ç€çœŸå®å­˜åœ¨çš„ç»†èƒå’Œç©ºæ¶²æ»´ã€‚å½“ç„¶æœ€ç®€å•çš„åŠæ³•è¿˜æ˜¯â€ä¸€åˆ€åˆ‡â€œï¼Œç•™ä¸‹é‚£äº›æ–‡åº“è¾ƒå¤§çš„ç»†èƒï¼Œä¸è¿‡è¿˜æ˜¯æœ‰æŸå¤±çœŸå®ç»†èƒç±»å‹çš„é£é™©\næ£€æµ‹ç©ºçš„æ¶²æ»´ æˆ‘ä»¬ä½¿ç”¨emptyDrops()å‡½æ•°ï¼Œæ£€æŸ¥æ¯ä¸ªbarcodeçš„è¡¨è¾¾é‡æ˜¯ä¸æ˜¯å’Œå¤–æºRNAçš„è¡¨è¾¾é‡æœ‰æ˜¾è‘—å·®å¼‚(Lun et al. 2019)ã€‚å¦‚æœå­˜åœ¨æ˜¾è‘—å·®å¼‚ï¼Œå°±è¯´æ˜barcodeä¸­æ›´æœ‰å¯èƒ½æ˜¯ä¸€ä¸ªç»†èƒã€‚è¿™ç§æ–¹æ³•å¯ä»¥å¸®åŠ©åŒºåˆ†ï¼šæµ‹åºè´¨é‡å¥½çš„ç©ºæ¶²æ»´ å’Œ åŒ…å«ç»†èƒä½†RNAå«é‡è¾ƒå°‘çš„æ¶²æ»´ã€‚å°½ç®¡å®ƒä»¬æ€»ä½“çš„è¡¨è¾¾é‡å¯èƒ½å¾ˆç›¸ä¼¼ï¼Œä½†æœ¬è´¨ä¸åŒï¼Œè¿˜æ˜¯è¦åŒºåˆ†çš„ã€‚\nemptyDrops() ä½¿ç”¨Monte Carloç»Ÿè®¡æ¨¡æ‹Ÿè®¡ç®—på€¼ï¼Œå¦‚æœè¦é‡å¤ç»“æœï¼Œéœ€è¦è®¾ç½®éšæœºç§å­ã€‚å¦å¤–è®¾ç½® false discovery rate (FDR)ä¸º0.1%ï¼Œæ„å‘³ç€ä¸è¶…è¿‡0.1%çš„barcodesæ˜¯ç©ºçš„ã€‚\n",
  "wordCount" : "5677",
  "inLanguage": "en",
  "image":"https://source.unsplash.com/random/400x200?code","datePublished": "2020-08-12T16:50:44+08:00",
  "dateModified": "2024-02-29T16:50:44+08:00",
  "author":{
    "@type": "Person",
    "name": "Jeason"
  },
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "http://localhost:1313/en/posts/tech/sc_qc/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Jeason's Blog",
    "logo": {
      "@type": "ImageObject",
      "url": "http://localhost:1313/img/avatar.jpeg"
    }
  }
}
</script>
</head>

<body class="" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="http://localhost:1313/en/" accesskey="h" title="Jeason&#39;s Blog (Alt + H)">
                <img src="http://localhost:1313/img/avatar.jpeg" alt="" aria-label="logo"
                    height="35">Jeason&#39;s Blog</a>
            <div class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="18" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="18" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                
            </div>
        </div>
        <ul id="menu">
            <li class="menu-item">
                <a href="http://localhost:1313/en/search" title="ğŸ” æœç´¢ (Alt &#43; /)" accesskey=/>
                    <span>ğŸ” æœç´¢</span>
                    
                </a>
            
            </li>
            <li class="menu-item">
                <a href="http://localhost:1313/en/" title="ğŸ¡ ä¸»é¡µ">
                    <span>ğŸ¡ ä¸»é¡µ</span>
                    
                </a>
            
            </li>
            <li class="menu-item">
                <a href="http://localhost:1313/en/posts" title="ğŸ“š æ–‡ç« ">
                    <span>ğŸ“š æ–‡ç« </span>
                    
                </a>
            
            </li>
            <li class="menu-item">
                <a href="http://localhost:1313/en/archives" title="ğŸ“ˆ å½’æ¡£">
                    <span>ğŸ“ˆ å½’æ¡£</span>
                    
                </a>
            
            </li>
        </ul>
    </nav>
</header>
<main class="main page">
<article class="post-single">
    <div id="single-content">
        <header class="post-header">
            <div class="breadcrumbs"><a href="http://localhost:1313/en/">ä¸»é¡µ</a>&nbsp;Â»&nbsp;<a href="http://localhost:1313/en/posts/">ğŸ“š æ–‡ç« </a>&nbsp;Â»&nbsp;<a href="http://localhost:1313/en/posts/tech/">ğŸ‘¨ğŸ»â€ğŸ’» æŠ€æœ¯</a></div>
            <h1 class="post-title">
                scRNAçš„ç»†èƒè´¨æ§
            </h1>
            <div class="post-meta">åˆ›å»º: 2020-08-12 | æ›´æ–°: 2024-02-29 | ä½œè€…:Jeason




                
                <span id="busuanzi_container_page_pv">
                    &nbsp;|&nbsp;è®¿é—®:&nbsp;<span id="busuanzi_value_page_pv"></span>
                </span>

</div>
        </header> <aside id="toc-container" class="toc-container wide">
    <div class="toc">
        <details  open>
        <summary accesskey="c" title="(Alt + C)">
            <span class="details">æ–‡ç« ç›®å½•</span>
        </summary>

        <div class="inner"><ul>
                <li>
                    <a href="#%e5%8a%a8%e6%9c%ba" aria-label="åŠ¨æœº">åŠ¨æœº</a></li>
                <li>
                    <a href="#qc%e6%8c%87%e6%a0%87%e7%9a%84%e9%80%89%e6%8b%a9" aria-label="QCæŒ‡æ ‡çš„é€‰æ‹©">QCæŒ‡æ ‡çš„é€‰æ‹©</a></li>
                <li>
                    <a href="#%e8%af%86%e5%88%ab%e4%bd%8e%e8%b4%a8%e9%87%8f%e7%bb%86%e8%83%9e" aria-label="è¯†åˆ«ä½è´¨é‡ç»†èƒ">è¯†åˆ«ä½è´¨é‡ç»†èƒ</a><ul>
                        
                <li>
                    <a href="#1-%e4%bd%bf%e7%94%a8%e5%9b%ba%e5%ae%9a%e7%9a%84%e9%98%88%e5%80%bc" aria-label="1. ä½¿ç”¨å›ºå®šçš„é˜ˆå€¼">1. ä½¿ç”¨å›ºå®šçš„é˜ˆå€¼</a></li>
                <li>
                    <a href="#2-%e4%bd%bf%e7%94%a8%e7%9b%b8%e5%af%b9%e9%98%88%e5%80%bc" aria-label="2. ä½¿ç”¨ç›¸å¯¹é˜ˆå€¼">2. ä½¿ç”¨ç›¸å¯¹é˜ˆå€¼</a><ul>
                        
                <li>
                    <a href="#%e9%89%b4%e5%ae%9a%e7%a6%bb%e7%be%a4%e7%82%b9" aria-label="é‰´å®šç¦»ç¾¤ç‚¹">é‰´å®šç¦»ç¾¤ç‚¹</a></li></ul>
                </li>
                <li>
                    <a href="#%e7%a6%bb%e7%be%a4%e7%82%b9%e6%a3%80%e6%b5%8b%e7%9a%84%e5%81%87%e8%ae%be" aria-label="ç¦»ç¾¤ç‚¹æ£€æµ‹çš„å‡è®¾">ç¦»ç¾¤ç‚¹æ£€æµ‹çš„å‡è®¾</a></li>
                <li>
                    <a href="#%e8%80%83%e8%99%91%e5%ae%9e%e9%aa%8c%e7%9a%84%e5%9b%a0%e7%b4%a0" aria-label="è€ƒè™‘å®éªŒçš„å› ç´ ">è€ƒè™‘å®éªŒçš„å› ç´ </a></li>
                <li>
                    <a href="#3-%e5%85%b6%e4%bb%96%e6%96%b9%e6%b3%95" aria-label="3. å…¶ä»–æ–¹æ³•">3. å…¶ä»–æ–¹æ³•</a></li></ul>
                </li>
                <li>
                    <a href="#%e7%94%bb%e5%9b%be%e6%a3%80%e6%9f%a5" aria-label="ç”»å›¾æ£€æŸ¥">ç”»å›¾æ£€æŸ¥</a></li>
                <li>
                    <a href="#%e9%92%88%e5%af%b9droplet%e6%95%b0%e6%8d%ae%e7%9a%84%e7%bb%86%e8%83%9e%e5%88%a4%e6%96%ad" aria-label="é’ˆå¯¹Dropletæ•°æ®çš„ç»†èƒåˆ¤æ–­">é’ˆå¯¹Dropletæ•°æ®çš„ç»†èƒåˆ¤æ–­</a><ul>
                        
                <li>
                    <a href="#%e8%83%8c%e6%99%af" aria-label="èƒŒæ™¯">èƒŒæ™¯</a></li>
                <li>
                    <a href="#%e6%a3%80%e6%b5%8b%e7%a9%ba%e7%9a%84%e6%b6%b2%e6%bb%b4" aria-label="æ£€æµ‹ç©ºçš„æ¶²æ»´">æ£€æµ‹ç©ºçš„æ¶²æ»´</a>
                </li>
            </ul>
            </li>
            </ul>
        </div>
        </details>
    </div>
</aside>
<script>
    let activeElement;
    let elements;
    window.addEventListener('DOMContentLoaded', function (event) {
        checkTocPosition();

        elements = document.querySelectorAll('h1[id],h2[id],h3[id],h4[id],h5[id],h6[id]');
        
        activeElement = elements[0];
        const id = encodeURI(activeElement.getAttribute('id')).toLowerCase();
        document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
    }, false);

    window.addEventListener('resize', function(event) {
        checkTocPosition();
    }, false);

    window.addEventListener('scroll', () => {
        
        activeElement = Array.from(elements).find((element) => {
            if ((getOffsetTop(element) - window.pageYOffset) > 0 &&
                (getOffsetTop(element) - window.pageYOffset) < window.innerHeight/2) {
                return element;
            }
        }) || activeElement

        elements.forEach(element => {
            const id = encodeURI(element.getAttribute('id')).toLowerCase();
            if (element === activeElement){
                document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
            } else {
                document.querySelector(`.inner ul li a[href="#${id}"]`).classList.remove('active');
            }
        })
    }, false);

    const main = parseInt(getComputedStyle(document.body).getPropertyValue('--article-width'), 10);
    const toc = parseInt(getComputedStyle(document.body).getPropertyValue('--toc-width'), 10);
    const gap = parseInt(getComputedStyle(document.body).getPropertyValue('--gap'), 10);

    function checkTocPosition() {
        const width = document.body.scrollWidth;
        if (width - main - (toc * 2) - (gap * 4) > 0) {
            document.getElementById("toc-container").classList.add("wide");
        } else {
            document.getElementById("toc-container").classList.remove("wide");
        }
    }

    function getOffsetTop(element) {
        if (!element.getClientRects().length) {
            return 0;
        }
        let rect = element.getBoundingClientRect();
        let win = element.ownerDocument.defaultView;
        return rect.top + win.pageYOffset;
    }
</script>
        <div class="post-content"><h2 id="åŠ¨æœº">åŠ¨æœº<a hidden class="anchor" aria-hidden="true" href="#åŠ¨æœº">#</a></h2>
<p>scRNA-seqæ•°æ®çš„ä½è´¨é‡æ–‡åº“å¯èƒ½æ¥è‡ªäºï¼šç»†èƒåˆ†é€‰ç¯èŠ‚çš„ç ´åã€æ–‡åº“åˆ¶å¤‡å¤±è¯¯ï¼ˆæ²¡æœ‰è¶³å¤Ÿçš„åè½¬å½•æˆ–PCRæ¬¡æ•°ï¼‰â€¦ è¡¨ç°åœ¨ï¼šç»†èƒæ€»è¡¨è¾¾é‡ä½ã€åŸºæœ¬æ²¡æœ‰è¡¨è¾¾çš„åŸºå› ã€é«˜çº¿ç²’ä½“æˆ–spike-inå æ¯”ã€‚</p>
<p>è¿™äº›ä½è´¨é‡çš„åº“æ˜¯æœ‰é—®é¢˜çš„ï¼Œå› ä¸ºå®ƒä»¬å¯èƒ½åœ¨ä¸‹æ¸¸åˆ†æä¸­å¯¼è‡´è¯¯å¯¼çš„ç»“æœ ã€‚</p>
<ul>
<li><strong>èšç±»é—®é¢˜ï¼š</strong> ä½è´¨é‡çš„ç»†èƒä¼šèšé›†æˆä¸€ç¾¤ï¼Œå¯¹ç»“æœçš„è§£é‡Šé€ æˆå¹²æ‰°ï¼Œå› ä¸ºä»è¿™ç¾¤ç»†èƒä¸­å¾—ä¸åˆ°ä»€ä¹ˆæœ‰ç”¨çš„ä¿¡æ¯ï¼Œä½†æ˜¯å®ƒçš„ç¡®ä¹Ÿæ˜¯ä¸€ç¾¤ã€‚è¿™ç§ç°è±¡äº§ç”Ÿçš„åŸå› æœ‰å¯èƒ½æ˜¯ï¼šç»†èƒç ´åä»¥åï¼Œçº¿ç²’ä½“æˆ–æ ¸RNAså æ¯”å‡é«˜ã€‚  æœ€å·®çš„æƒ…å†µå°±æ˜¯ï¼šä¸åŒç±»å‹çš„ä½è´¨é‡ç»†èƒï¼Œä¹Ÿèƒ½èšåœ¨ä¸€èµ·ï¼Œå› ä¸ºç›¸æ¯”äºå›ºæœ‰çš„ç”Ÿç‰©å·®å¼‚ï¼Œæ›´ç›¸ä¼¼çš„ä½è´¨é‡è®©å®ƒä»¬ç›¸ä¾ç›¸åã€‚é™¤æ­¤ä»¥å¤–ï¼Œæœ¬æ¥éå¸¸å°çš„ç»†èƒæ–‡åº“ä¹Ÿèƒ½èšæˆä¸€ç¾¤ï¼Œå› ä¸ºlogè½¬æ¢åå®ƒä»¬çš„å¹³å‡è¡¨è¾¾é‡ä¼šå‘ç”Ÿå¾ˆå¤§çš„å˜åŒ–(A. Lun <a href="https://osca.bioconductor.org/quality-control.html#ref-lun2018overcoming" target="_blank" rel="noopener" style="color:#42b983";>2018</a>).</li>
<li><strong>æ–¹å·®ä¼°è®¡æˆ–ä¸»æˆåˆ†åˆ†æ</strong>:é¦–å…ˆåœ¨PCAåˆ†ææ—¶ï¼Œä½è´¨é‡å’Œé«˜è´¨é‡ä¹‹é—´çš„å·®å¼‚ç›¸æ¯”äºç”Ÿç‰©å­¦å·®å¼‚ä¼šä½“ç°æ›´æ˜æ˜¾ï¼Œå æ®ä¸»è¦çš„æˆåˆ†ï¼Œå‡å°‘é™ç»´ç»“æœçš„å¯é æ€§ã€‚å¦å¤–ï¼ŒæŸä¸ªåŸºå› å¯èƒ½åœ¨ä¸¤ä¸ªç»†èƒä¹‹é—´æ²¡ä»€ä¹ˆè¡¨è¾¾å·®å¼‚ï¼Œä½†æ˜¯å¦‚æœæ‰€å¤„ç¯å¢ƒå·®å¼‚å¾ˆå¤§ï¼ˆä¸€ä¸ªç»†èƒè´¨é‡å¾ˆä½ï¼Œå¦ä¸€ä¸ªç»†èƒè´¨é‡æ­£å¸¸ï¼‰ï¼Œé‚£ä¹ˆåœ¨å·®å¼‚ä¼°ç®—è¿‡ç¨‹ä¸­ï¼Œå°±ä¼šæŠŠè¿™ä¸ªå·®å¼‚ä¹Ÿä¼šè¢«å½“æˆå·®å¼‚è¡¨è¾¾åŸºå› ã€‚ä¾‹å¦‚ï¼šä¸€ä¸ªä½è´¨é‡ç»†èƒæ–‡åº“çš„æ€»è¡¨è¾¾é‡éå¸¸ä½ï¼ˆæ¥è¿‘0ï¼‰ï¼Œä½†æ°å·§è¿˜å­˜åœ¨ä¸€ä¸ªåŸºå› æœ‰è¡¨è¾¾é‡ï¼Œé‚£ä¹ˆè¿™ä¸ªåŸºå› çš„è¡¨è¾¾é‡åœ¨åç»­çš„æ–‡åº“å½’ä¸€åŒ–è¿‡ç¨‹ä¸­å°±ä¼šå°¤ä¸ºçªå‡º</li>
<li><strong>å¥‡æ€ªçš„è½¬å½•æœ¬ä¸Šè°ƒ</strong>ï¼šå®éªŒéš¾å…ä¼šæ··å…¥å¤–æºçš„æ±¡æŸ“è½¬å½•æœ¬ï¼Œä½†è¿™ä¸ªé‡å¾ˆå°‘å¹¶ä¸”åœ¨æ‰€æœ‰ç»†èƒä¸­éƒ½åº”è¯¥æ˜¯å·®ä¸å¤šæ°´å¹³çš„ã€‚ä½†å¦‚æœæŸä¸ªç»†èƒè´¨é‡ä½ï¼Œæ–‡åº“å°ï¼Œé‚£ä¹ˆåœ¨æ ¡æ­£æ–‡åº“å·®å¼‚è¿‡ç¨‹ä¸­ï¼Œå…¶ä¸­çš„æ±¡æŸ“è½¬å½•æœ¬è¡¨è¾¾é‡å°±ä¼šâ€œçªé£çŒ›è¿›â€ï¼Œçœ‹èµ·æ¥æ˜¯ä¸€äº›æ˜æ˜¾ä¸Šè°ƒçš„â€œå¥‡æ€ªåŸºå› â€ã€‚å®é™…ä¸Šï¼Œè¿™äº›å¥‡æ€ªçš„åŸºå› ä¾ç„¶åœ¨å…¶ä»–ç»†èƒä¸­å­˜åœ¨ï¼Œå¹¶ä¸”çœŸå®çš„è¡¨è¾¾é‡å·®ä¸å¤šï¼Œå¹¶ä¸”æ˜¯ä¸åº”è¯¥å æ®ä¸»ä½“åœ°ä½çš„ã€‚</li>
</ul>
<p>ä¸ºäº†æœ€å¤§ç¨‹åº¦é¿å…ä¸Šé¢ä¸€ç§æˆ–å¤šç§æƒ…å†µçš„å‘ç”Ÿï¼Œåº”è¯¥åœ¨å½’ä¸€åŒ–ä¹‹å‰å»æ‰è¿™äº›ä½è´¨é‡çš„ç»†èƒï¼Œè¿™ä¸ªè¿‡ç¨‹å°±æ˜¯**ç»†èƒçš„è´¨æ§ **</p>
<p>ä½¿ç”¨A. T. L. Lun et al. (2017)çš„å°å‹scRNAæ•°æ®ï¼ˆæœªè¿›è¡ŒQCï¼‰è¿›è¡Œæµ‹è¯•</p>
<div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#776e71">#--- loading ---#</span>
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">library</span>(scRNAseq)
</span></span><span style="display:flex;"><span>sce.416b <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">LunSpikeInData</span>(which<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;416b&#34;</span>) 
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">table</span>(sce.416b<span style="color:#5bc4bf">$</span>block)
</span></span><span style="display:flex;"><span><span style="color:#f99b15">20160113</span> <span style="color:#f99b15">20160325</span> 
</span></span><span style="display:flex;"><span>      <span style="color:#f99b15">96</span>       <span style="color:#f99b15">96</span> 
</span></span><span style="display:flex;"><span>sce.416b<span style="color:#5bc4bf">$</span>block <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">factor</span>(sce.416b<span style="color:#5bc4bf">$</span>block)
</span></span><span style="display:flex;"><span>sce.416b
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#776e71">## class: SingleCellExperiment </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## dim: 46604 192 </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## metadata(0):</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## assays(1): counts</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## rownames(46604): ENSMUSG00000102693 ENSMUSG00000064842 ...</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##   ENSMUSG00000095742 CBFB-MYH11-mcherry</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## rowData names(1): Length</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## colnames(192): SLX-9555.N701_S502.C89V9ANXX.s_1.r_1</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##   SLX-9555.N701_S503.C89V9ANXX.s_1.r_1 ...</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##   SLX-11312.N712_S508.H5H5YBBXX.s_8.r_1</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##   SLX-11312.N712_S517.H5H5YBBXX.s_8.r_1</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## colData names(9): Source Name cell line ... spike-in addition block</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## reducedDimNames(0):</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## altExpNames(2): ERCC SIRV</span>
</span></span></code></pre></div><h2 id="qcæŒ‡æ ‡çš„é€‰æ‹©">QCæŒ‡æ ‡çš„é€‰æ‹©<a hidden class="anchor" aria-hidden="true" href="#qcæŒ‡æ ‡çš„é€‰æ‹©">#</a></h2>
<p>é‰´å®šç»†èƒæ˜¯å¦æ˜¯ä½è´¨é‡çš„ï¼Œéœ€è¦ç”¨åˆ°å‡ ä¸ªæŒ‡æ ‡ã€‚è™½ç„¶ä¸‹é¢è¿™äº›æŒ‡æ ‡æ˜¯ä½¿ç”¨SMART-seq2æ•°æ®è¿›è¡Œå±•ç¤ºçš„ï¼Œä½†ä¾ç„¶é€‚ç”¨äºUMIæ•°æ®ï¼ˆæ¯”å¦‚MARS-seqã€droplet-basedæŠ€æœ¯ï¼‰</p>
<ul>
<li><strong>æ–‡åº“å¤§å°ï¼ˆlibrary sizeï¼‰</strong> ï¼šæŒ‡çš„æ˜¯æ¯ä¸ªç»†èƒä¸­æ‰€æœ‰åŸºå› çš„è¡¨è¾¾é‡ä¹‹å’Œã€‚ç»†èƒçš„æ–‡åº“å¦‚æœå¾ˆå°ï¼Œè¯´æ˜åœ¨æ–‡åº“åˆ¶å¤‡è¿‡ç¨‹ä¸­å­˜åœ¨RNAçš„æŸå¤±ï¼Œè¦ä¹ˆæ˜¯ç”±äºç»†èƒè£‚è§£ï¼Œè¦ä¹ˆæ˜¯cDNAæ•è·ä¸è¶³å¯¼è‡´åç»­çš„æ‰©å¢é‡å°‘</li>
<li><strong>æ¯ä¸ªç»†èƒä¸­è¡¨è¾¾åŸºå› çš„æ•°ç›®ï¼ˆnumber of expressed features in each cellï¼‰</strong>ï¼šæŒ‡çš„æ˜¯ç»†èƒä¸­é0è¡¨è¾¾é‡çš„åŸºå› æ•°é‡ã€‚å¦‚æœç»†èƒä¸­åŸºæœ¬æ²¡æœ‰åŸºå› è¡¨è¾¾ï¼Œå¾ˆå¯èƒ½æ˜¯è½¬å½•æœ¬æ•è·å¤±è´¥</li>
<li><strong>æ¯”å¯¹åˆ°spike-inè½¬å½•æœ¬çš„readsæ¯”ä¾‹ï¼ˆproportion of reads mapped to spike-in transcriptsï¼‰</strong>ï¼šè®¡ç®—æ–¹æ³•æ˜¯ï¼š<code>spike-in counts / all features (including spike-ins) for each cell</code>ã€‚æ¯ä¸ªç»†èƒéƒ½åº”è¯¥åŠ å…¥ç­‰é‡çš„å¤–æºè½¬å½•æœ¬ï¼ˆspike-inï¼‰ï¼Œå¦‚æœå“ªä¸ªç»†èƒçš„spike-inæ¯”ä¾‹æé«˜äº†ï¼Œè¯´æ˜å®ƒçš„å†…æºRNAå«é‡å‡å°‘ï¼ˆæ¯”å¦‚åœ¨ç»†èƒåˆ†é€‰é˜¶æ®µå‡ºç°çš„ç»†èƒè£‚è§£æˆ–è€…RNAé™è§£ï¼‰</li>
<li><strong>æ¯”å¯¹åˆ°çº¿ç²’ä½“åŸºå› ç»„çš„readsæ¯”ä¾‹ï¼ˆproportion of reads mapped to genes in the mitochondrial genomeï¼‰</strong> ï¼šå¦‚æœæ²¡æœ‰spike-inï¼Œé‚£ä¹ˆä½¿ç”¨çº¿ç²’ä½“æŒ‡æ ‡ä¹Ÿæ˜¯èƒ½è¯´æ˜é—®é¢˜çš„(Islam et al. 2014; Ilicic et al. 2016)ã€‚æ¯”å¯¹åˆ°çº¿ç²’ä½“çš„readså¢å¤šï¼Œè¯´æ˜ç»†èƒè´¨ä¸­çš„RNAå‡å°‘ï¼Œå¯èƒ½å­˜åœ¨ç»†èƒç©¿å­”çš„æƒ…å†µï¼Œè€Œè¿™ä¸ªå­”çš„å¤§å°ï¼Œå¯èƒ½åªæ˜¯å°†ç»†èƒè´¨ä¸­å­˜åœ¨çš„mRNAæµå‡ºå»ï¼Œä½†çº¿ç²’ä½“çš„ä½“ç§¯è¶…è¿‡äº†å­”çš„å¤§å°ï¼Œæ‰€ä»¥è¿˜ç•™åœ¨äº†ç»†èƒä¸­ï¼Œé€ æˆä¸€å®šç¨‹åº¦çš„å¯Œé›†ï¼Œå¯¼è‡´çº¿ç²’ä½“RNAå æ¯”å‡é«˜ã€‚</li>
</ul>
<p>å¯¹äºæ¯ä¸ªç»†èƒï¼Œå¯ä»¥ç”¨scateråŒ…çš„<code>perCellQCMetrics()</code>å‡½æ•°è¿›è¡Œè®¡ç®—ï¼Œå…¶ä¸­<code>sum</code>è¿™ä¸€åˆ—è¡¨ç¤ºæ¯ä¸ªç»†èƒçš„æ–‡åº“å¤§å°ï¼›<code>detected</code>è¿™ä¸€åˆ—è¡¨ç¤ºæ£€æµ‹åˆ°çš„åŸºå› æ•°é‡ï¼›<code>subsets_Mito_percent</code>è¿™ä¸€åˆ—è¡¨ç¤ºæ¯”å¯¹åˆ°çº¿ç²’ä½“åŸºå› ç»„çš„readså æ¯”ï¼›<code>altexps_ERCC_percent</code>è¡¨ç¤ºæ¯”å¯¹åˆ°ERCC spike-inçš„readså æ¯”</p>
<div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#776e71"># Retrieving the mitochondrial transcripts using genomic locations included in</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71"># the row-level annotation for the SingleCellExperiment.</span>
</span></span><span style="display:flex;"><span>location <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">rowRanges</span>(sce.416b)
</span></span><span style="display:flex;"><span>is.mito <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">any</span>(<span style="color:#06b6ef">seqnames</span>(location)<span style="color:#5bc4bf">==</span><span style="color:#48b685">&#34;MT&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#776e71"># ALTERNATIVELY: using resources in AnnotationHub to retrieve chromosomal</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71"># locations given the Ensembl IDs; this should yield the same result.</span>
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">library</span>(AnnotationHub)
</span></span><span style="display:flex;"><span>ens.mm.v97 <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">AnnotationHub</span>()[[<span style="color:#48b685">&#34;AH73905&#34;</span>]]
</span></span><span style="display:flex;"><span>chr.loc <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">mapIds</span>(ens.mm.v97, keys<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">rownames</span>(sce.416b),
</span></span><span style="display:flex;"><span>    keytype<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;GENEID&#34;</span>, column<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;SEQNAME&#34;</span>)
</span></span><span style="display:flex;"><span>is.mito.alt <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">which</span>(chr.loc<span style="color:#5bc4bf">==</span><span style="color:#48b685">&#34;MT&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">library</span>(scater)
</span></span><span style="display:flex;"><span>df <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">perCellQCMetrics</span>(sce.416b, subsets<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">list</span>(Mito<span style="color:#5bc4bf">=</span>is.mito))
</span></span><span style="display:flex;"><span>df
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#776e71">## DataFrame with 192 rows and 16 columns</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##           sum  detected percent_top_50 percent_top_100 percent_top_200</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##     &lt;integer&gt; &lt;integer&gt;      &lt;numeric&gt;       &lt;numeric&gt;       &lt;numeric&gt;</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 1      865936      7618        26.7218         32.2773         39.7208</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 2     1076277      7521        29.4043         35.0354         42.2581</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 3     1180138      8306        27.3454         32.4770         39.3296</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 4     1342593      8143        35.8092         40.2666         46.2460</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 5     1668311      7154        34.1198         39.0901         45.6660</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## ...       ...       ...            ...             ...             ...</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 188    776622      8174        45.9362         49.7010         54.6101</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 189   1299950      8956        38.0829         42.8930         49.0622</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 190   1800696      9530        30.6675         35.5839         41.8550</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 191     46731      6649        32.2998         37.9149         44.5999</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 192   1866692     10964        26.6632         31.2584         37.5608</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##     percent_top_500 subsets_Mito_sum subsets_Mito_detected subsets_Mito_percent</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##           &lt;numeric&gt;        &lt;integer&gt;             &lt;integer&gt;            &lt;numeric&gt;</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 1           52.9038            78790                    20              9.09882</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 2           55.7454            98613                    20              9.16242</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 3           51.9337           100341                    19              8.50248</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 4           57.1210           104882                    20              7.81190</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 5           58.2004           129559                    22              7.76588</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## ...             ...              ...                   ...                  ...</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 188         64.4249            48126                    20              6.19684</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 189         60.6675           112225                    25              8.63302</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 190         53.6781           135693                    23              7.53559</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 191         56.5235             3505                    16              7.50037</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 192         48.9489           150375                    29              8.05569</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##     altexps_ERCC_sum altexps_ERCC_detected altexps_ERCC_percent</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##            &lt;integer&gt;             &lt;integer&gt;            &lt;numeric&gt;</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 1              65278                    39              6.80658</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 2              74748                    40              6.28030</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 3              60878                    42              4.78949</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 4              60073                    42              4.18567</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 5             136810                    44              7.28887</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## ...              ...                   ...                  ...</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 188            61575                    39              7.17620</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 189            94982                    41              6.65764</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 190           113707                    40              5.81467</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 191             7580                    44             13.48898</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 192            48664                    39              2.51930</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##     altexps_SIRV_sum altexps_SIRV_detected altexps_SIRV_percent     total</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##            &lt;integer&gt;             &lt;integer&gt;            &lt;numeric&gt; &lt;integer&gt;</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 1              27828                     7              2.90165    959042</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 2              39173                     7              3.29130   1190198</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 3              30058                     7              2.36477   1271074</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 4              32542                     7              2.26741   1435208</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 5              71850                     7              3.82798   1876971</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## ...              ...                   ...                  ...       ...</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 188            19848                     7             2.313165    858045</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 189            31729                     7             2.224004   1426661</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 190            41116                     7             2.102562   1955519</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 191             1883                     7             3.350892     56194</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 192            16289                     7             0.843271   1931645</span>
</span></span></code></pre></div><p>å¦å¤–ï¼Œè¿˜å¯ä»¥ä½¿ç”¨<code>addPerCellQC()</code>ï¼Œå®ƒä¼šæŠŠæ¯ä¸ªç»†èƒçš„QCæŒ‡æ ‡åŠ åˆ°<code>SingleCellExperiment</code>å¯¹è±¡çš„<code>colData</code>ä¸­ï¼Œæ–¹ä¾¿åé¢è°ƒå–</p>
<div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>sce.416b <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">addPerCellQC</span>(sce.416b, subsets<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">list</span>(Mito<span style="color:#5bc4bf">=</span>is.mito))
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">colnames</span>(<span style="color:#06b6ef">colData</span>(sce.416b))
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#776e71">##  [1] &#34;Source Name&#34;              &#34;cell line&#34;               </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##  [3] &#34;cell type&#34;                &#34;single cell well quality&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##  [5] &#34;genotype&#34;                 &#34;phenotype&#34;               </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##  [7] &#34;strain&#34;                   &#34;spike-in addition&#34;       </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##  [9] &#34;block&#34;                    &#34;sum&#34;                     </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## [11] &#34;detected&#34;                 &#34;percent_top_50&#34;          </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## [13] &#34;percent_top_100&#34;          &#34;percent_top_200&#34;         </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## [15] &#34;percent_top_500&#34;          &#34;subsets_Mito_sum&#34;        </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## [17] &#34;subsets_Mito_detected&#34;    &#34;subsets_Mito_percent&#34;    </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## [19] &#34;altexps_ERCC_sum&#34;         &#34;altexps_ERCC_detected&#34;   </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## [21] &#34;altexps_ERCC_percent&#34;     &#34;altexps_SIRV_sum&#34;        </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## [23] &#34;altexps_SIRV_detected&#34;    &#34;altexps_SIRV_percent&#34;    </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## [25] &#34;total&#34;</span>
</span></span></code></pre></div><p>å½“ç„¶ï¼Œè¿™é‡ŒåšQCç»Ÿè®¡çš„å‰æå‡è®¾æ˜¯ï¼šæ¯ä¸ªç»†èƒçš„qcæŒ‡æ ‡éƒ½æ˜¯ç‹¬ç«‹äºç”Ÿç‰©å­¦çŠ¶æ€çš„ã€‚ä¹Ÿå°±æ˜¯è¯´ï¼ŒqcæŒ‡æ ‡ä¸ä¼šé‚£ä¹ˆæ™ºèƒ½åœ°è¯†åˆ«ç»†èƒç±»å‹ç„¶åè¿›è¡Œåˆ¤æ–­ã€‚å®ƒä¼šè®¤ä¸ºï¼ˆå¦‚æ–‡åº“å¤ªå°ã€é«˜çº¿ç²’ä½“å æ¯”ï¼‰éƒ½æ˜¯ç”±æŠ€æœ¯è¯¯å·®å¼•èµ·çš„ï¼Œè€Œéç”Ÿç‰©å› ç´ ã€‚ä½†æ˜¯æœ‰ä¸€ä¸ªé—®é¢˜ï¼Œ<strong>å¦‚æœæŸäº›ç»†èƒç±»å‹æœ¬èº«çš„RNAå«é‡å°±å¾ˆä½ï¼Œæˆ–è€…å®ƒä»¬æœ¬æ¥å°±å«æœ‰å¾ˆå¤šçš„çº¿ç²’ä½“è½¬å½•æœ¬å‘¢ï¼Ÿå†æ ¹æ®è¿™ä¸ªæŒ‡æ ‡è¿›è¡Œè¿‡æ»¤ï¼Œä¼šä¸ä¼šæŸå¤±ä¸€äº›ç»†èƒç±»å‹å‘¢ï¼Ÿ</strong></p>
<h2 id="è¯†åˆ«ä½è´¨é‡ç»†èƒ">è¯†åˆ«ä½è´¨é‡ç»†èƒ<a hidden class="anchor" aria-hidden="true" href="#è¯†åˆ«ä½è´¨é‡ç»†èƒ">#</a></h2>
<h3 id="1-ä½¿ç”¨å›ºå®šçš„é˜ˆå€¼">1. ä½¿ç”¨å›ºå®šçš„é˜ˆå€¼<a hidden class="anchor" aria-hidden="true" href="#1-ä½¿ç”¨å›ºå®šçš„é˜ˆå€¼">#</a></h3>
<p>è¯†åˆ«ä½è´¨é‡ç»†èƒæœ€ç®€å•æ–¹æ³•æ˜¯åœ¨QCåº¦é‡ä¸Šåº”ç”¨é˜ˆå€¼ã€‚ä¾‹å¦‚è®¾å®šæ–‡åº“ä½äº10ä¸‡readsçš„ç»†èƒæ˜¯ä½è´¨é‡çš„ï¼Œæˆ–è€…è¡¨è¾¾åŸºå› æ•°é‡å°‘äº5000ä¸ªï¼Œspike-inæˆ–çº¿ç²’ä½“å æ¯”é«˜äº10%ã€‚</p>
<div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>qc.lib <span style="color:#5bc4bf">&lt;-</span> df<span style="color:#5bc4bf">$</span>sum <span style="color:#5bc4bf">&lt;</span> <span style="color:#f99b15">1e5</span>
</span></span><span style="display:flex;"><span>qc.nexprs <span style="color:#5bc4bf">&lt;-</span> df<span style="color:#5bc4bf">$</span>detected <span style="color:#5bc4bf">&lt;</span> <span style="color:#f99b15">5e3</span>
</span></span><span style="display:flex;"><span>qc.spike <span style="color:#5bc4bf">&lt;-</span> df<span style="color:#5bc4bf">$</span>altexps_ERCC_percent <span style="color:#5bc4bf">&gt;</span> <span style="color:#f99b15">10</span>
</span></span><span style="display:flex;"><span>qc.mito <span style="color:#5bc4bf">&lt;-</span> df<span style="color:#5bc4bf">$</span>subsets_Mito_percent <span style="color:#5bc4bf">&gt;</span> <span style="color:#f99b15">10</span>
</span></span><span style="display:flex;"><span>discard <span style="color:#5bc4bf">&lt;-</span> qc.lib <span style="color:#5bc4bf">|</span> qc.nexprs <span style="color:#5bc4bf">|</span> qc.spike <span style="color:#5bc4bf">|</span> qc.mito
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#776e71"># Summarize the number of cells removed for each reason.</span>
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">DataFrame</span>(LibSize<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">sum</span>(qc.lib), NExprs<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">sum</span>(qc.nexprs),
</span></span><span style="display:flex;"><span>    SpikeProp<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">sum</span>(qc.spike), MitoProp<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">sum</span>(qc.mito), Total<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">sum</span>(discard))
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#776e71">## DataFrame with 1 row and 5 columns</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##     LibSize    NExprs SpikeProp  MitoProp     Total</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##   &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 1         3         0        19        14        33</span>
</span></span></code></pre></div><p>è™½ç„¶çœ‹èµ·æ¥ç®€å•ï¼Œä½†ä½¿ç”¨è¿™ç§æ–¹æ³•éœ€è¦ä¸°å¯Œçš„ç»éªŒï¼Œäº†è§£å®éªŒè®¾è®¡å’Œç»†èƒçŠ¶æ€ï¼›å¦å¤–å³ä½¿ä½¿ç”¨åŒä¸€ç§æ–‡åº“åˆ¶å¤‡æ–¹æ³•ï¼Œä½†ç”±äºç»†èƒæ•è·æ•ˆç‡å’Œæµ‹åºæ·±åº¦çš„ä¸åŒï¼Œè¿™ä¸ªé˜ˆå€¼ä¾ç„¶éœ€è¦é€‚æ—¶è°ƒæ•´ã€‚å› æ­¤å¯¹äºç ”ç©¶äººå‘˜è¦æ±‚å¾ˆé«˜ã€‚</p>
<h3 id="2-ä½¿ç”¨ç›¸å¯¹é˜ˆå€¼">2. ä½¿ç”¨ç›¸å¯¹é˜ˆå€¼<a hidden class="anchor" aria-hidden="true" href="#2-ä½¿ç”¨ç›¸å¯¹é˜ˆå€¼">#</a></h3>
<h4 id="é‰´å®šç¦»ç¾¤ç‚¹">é‰´å®šç¦»ç¾¤ç‚¹<a hidden class="anchor" aria-hidden="true" href="#é‰´å®šç¦»ç¾¤ç‚¹">#</a></h4>
<p>ä¸ºäº†è·å¾—ç›¸å¯¹é˜ˆå€¼ï¼Œå…ˆå‡è®¾å¤§éƒ¨åˆ†ç»†èƒéƒ½æ˜¯é«˜è´¨é‡çš„ï¼Œç„¶åå»æ‰¾ç¦»ç¾¤ç‚¹ä½œä¸ºä½è´¨é‡ã€‚é‚£ä¹ˆæŒ‰ç…§ä»€ä¹ˆæ–¹æ³•æ‰¾ç¦»ç¾¤ç‚¹å‘¢ï¼Ÿå¸¸ç”¨çš„ä¸€ä¸ªå‡½æ•°<code>isOutlier</code>ä½¿ç”¨çš„æ˜¯MADæŒ‡æ ‡ï¼ˆç»å¯¹ä¸­ä½å·®æ¥ä¼°è®¡æ–¹å·®,å…ˆè®¡ç®—å‡ºæ•°æ®ä¸å®ƒä»¬çš„ä¸­ä½æ•°ä¹‹é—´çš„åå·®ï¼Œç„¶åè¿™äº›åå·®çš„ç»å¯¹å€¼çš„ä¸­ä½æ•°å°±æ˜¯MADï¼Œmedian absolute deviationï¼‰ã€‚å¦‚æœ<strong>è¶…è¿‡ä¸­ä½æ•°3å€MADçš„å€¼å°±æ˜¯ç¦»ç¾¤å€¼</strong>ã€‚</p>
<p>ä½¿ç”¨<code>isOutlier</code>æ—¶ï¼Œå¦‚æœè¦ç›¸å‡ï¼ˆä¾‹å¦‚ï¼š<code>df$sum - 3* MAD</code>ï¼‰ï¼Œå°±ç”¨<code>type=&quot;lower&quot;</code>ï¼Œæ­¤æ—¶ä¸€èˆ¬è¿˜è¦åšä¸ªlogè½¬æ¢<code>log=TRUE</code> ï¼Œä¿è¯å¾—åˆ°çš„ç»“æœä¸æ˜¯è´Ÿæ•°</p>
<div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>qc.lib2 <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">isOutlier</span>(df<span style="color:#5bc4bf">$</span>sum, log<span style="color:#5bc4bf">=</span><span style="color:#815ba4">TRUE</span>, type<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;lower&#34;</span>)
</span></span><span style="display:flex;"><span>qc.nexprs2 <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">isOutlier</span>(df<span style="color:#5bc4bf">$</span>detected, log<span style="color:#5bc4bf">=</span><span style="color:#815ba4">TRUE</span>, type<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;lower&#34;</span>)
</span></span><span style="display:flex;"><span>qc.spike2 <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">isOutlier</span>(df<span style="color:#5bc4bf">$</span>altexps_ERCC_percent, type<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;higher&#34;</span>)
</span></span><span style="display:flex;"><span>qc.mito2 <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">isOutlier</span>(df<span style="color:#5bc4bf">$</span>subsets_Mito_percent, type<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;higher&#34;</span>)
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#06b6ef">attr</span>(qc.lib2, <span style="color:#48b685">&#34;thresholds&#34;</span>)
</span></span><span style="display:flex;"><span>   lower   higher 
</span></span><span style="display:flex;"><span><span style="color:#f99b15">434082.9</span>      <span style="color:#815ba4">Inf</span> 
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">attr</span>(qc.nexprs2, <span style="color:#48b685">&#34;thresholds&#34;</span>)
</span></span><span style="display:flex;"><span>   lower   higher 
</span></span><span style="display:flex;"><span><span style="color:#f99b15">5231.468</span>      <span style="color:#815ba4">Inf</span> 
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">attr</span>(qc.spike2, <span style="color:#48b685">&#34;thresholds&#34;</span>)
</span></span><span style="display:flex;"><span>   lower   higher 
</span></span><span style="display:flex;"><span>    <span style="color:#5bc4bf">-</span><span style="color:#815ba4">Inf</span> <span style="color:#f99b15">14.15371</span> 
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">attr</span>(qc.mito2, <span style="color:#48b685">&#34;thresholds&#34;</span>)
</span></span><span style="display:flex;"><span>   lower   higher 
</span></span><span style="display:flex;"><span>    <span style="color:#5bc4bf">-</span><span style="color:#815ba4">Inf</span> <span style="color:#f99b15">11.91734</span> 
</span></span></code></pre></div><p>ç”¨ç›¸å¯¹é˜ˆå€¼è¿‡æ»¤çš„ç»†èƒæ•°é‡ç»Ÿè®¡ï¼š</p>
<div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>discard2 <span style="color:#5bc4bf">&lt;-</span> qc.lib2 <span style="color:#5bc4bf">|</span> qc.nexprs2 <span style="color:#5bc4bf">|</span> qc.spike2 <span style="color:#5bc4bf">|</span> qc.mito2
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#776e71"># Summarize the number of cells removed for each reason.</span>
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">DataFrame</span>(LibSize<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">sum</span>(qc.lib2), NExprs<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">sum</span>(qc.nexprs2),
</span></span><span style="display:flex;"><span>    SpikeProp<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">sum</span>(qc.spike2), MitoProp<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">sum</span>(qc.mito2), Total<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">sum</span>(discard2))
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#776e71">## DataFrame with 1 row and 5 columns</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##     LibSize    NExprs SpikeProp  MitoProp     Total</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##   &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 1         4         0         1         2         6</span>
</span></span></code></pre></div><p><strong>é™¤æ­¤ä»¥å¤–ï¼Œè¿˜æœ‰ä¸€ç§æ›´å¿«çš„è®¡ç®—æ–¹æ³•ï¼Œä¸€æ­¥æ•´åˆäº†ä¸Šé¢çš„æ“ä½œï¼š</strong></p>
<div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>reasons <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">quickPerCellQC</span>(df, percent_subsets<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">c</span>(<span style="color:#48b685">&#34;subsets_Mito_percent&#34;</span>,
</span></span><span style="display:flex;"><span>    <span style="color:#48b685">&#34;altexps_ERCC_percent&#34;</span>))
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">colSums</span>(<span style="color:#06b6ef">as.matrix</span>(reasons))
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#776e71">##              low_lib_size            low_n_features high_subsets_Mito_percent </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##                         4                         0                         2 </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## high_altexps_ERCC_percent                   discard </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##                         1                         6</span>
</span></span></code></pre></div><p>ä½¿ç”¨â€ç›¸å¯¹â€œçš„é˜ˆå€¼ä¸€ä¸ªå¥½å¤„å°±æ˜¯å¯ä»¥æ ¹æ®æµ‹åºæ·±åº¦ã€cDNAæ•è·æ•ˆç‡ã€çº¿ç²’ä½“å«é‡ç­‰ç­‰è¿›è¡Œé˜ˆå€¼çš„è°ƒæ•´ï¼Œåœ¨ç»éªŒä¸æ˜¯è¶³å¤Ÿä¸°å¯Œçš„æ—¶å€™ï¼Œå¯ä»¥è¾…åŠ©åˆ¤æ–­ã€‚ä½†ä»éœ€è¦æ³¨æ„çš„æ˜¯ï¼Œ<strong>ä½¿ç”¨ç›¸å¯¹é˜ˆå€¼æ˜¯æœ‰å‰æçš„ï¼Œé‚£å°±æ˜¯ï¼šè®¤ä¸ºå¤§éƒ¨åˆ†ç»†èƒéƒ½æ˜¯é«˜è´¨é‡çš„</strong></p>
<h3 id="ç¦»ç¾¤ç‚¹æ£€æµ‹çš„å‡è®¾">ç¦»ç¾¤ç‚¹æ£€æµ‹çš„å‡è®¾<a hidden class="anchor" aria-hidden="true" href="#ç¦»ç¾¤ç‚¹æ£€æµ‹çš„å‡è®¾">#</a></h3>
<ul>
<li>ç¦»ç¾¤ç‚¹çš„æ£€æµ‹çš„å‡è®¾æ˜¯<strong>å¤§éƒ¨åˆ†ç»†èƒçš„è´¨é‡éƒ½ä¸é”™</strong>ã€‚è¿™ä¸€ç‚¹å‡è®¾å¯ä»¥é€šè¿‡å®éªŒå»éªŒè¯ï¼ˆæ¯”å¦‚è‚‰çœ¼æ£€æŸ¥ç»†èƒå®Œæ•´æ€§ï¼‰ã€‚<strong>ä½†å½“å¤§éƒ¨åˆ†ç»†èƒè´¨é‡éƒ½å¾ˆä½ï¼Œä½¿ç”¨ç›¸å¯¹é˜ˆå€¼ç»“æœå°±å¯¹å¤§é‡çš„ä½è´¨é‡ç»†èƒæ— è®¡å¯æ–½</strong>ã€‚å› ä¸ºå®ƒä½¿ç”¨äº†MADå€¼ï¼Œå’Œä¸­ä½æ•°æœ‰å…³ç³»ï¼Œé‚£ä¹ˆå¯ä»¥è¯•æƒ³ï¼šå¦‚æœä¸€å †æ•°éƒ½ä¸åˆæ ¼ï¼Œé‚£ä¹ˆå®ƒä»¬çš„ä¸­ä½æ•°ä¹Ÿä¸åˆæ ¼ï¼Œå› æ­¤åŸæ¥æ­£ç¡®çš„å€¼ï¼Œå…¶å®åœ¨è¿™ç¾¤ä¸åˆæ ¼çš„æ•°å€¼ä¸­ï¼Œå°±æ˜¯â€œç¦»ç¾¤â€çš„ã€‚</li>
<li>å¦ä¸€ä¸ªå‡è®¾å°±æ˜¯ï¼š<strong>æ¯ä¸ªç»†èƒçš„qcæŒ‡æ ‡éƒ½æ˜¯ç‹¬ç«‹äºç”Ÿç‰©å­¦çŠ¶æ€çš„</strong>ã€‚ä¹Ÿå°±æ˜¯è¯´ï¼ŒqcæŒ‡æ ‡ä¸ä¼šé‚£ä¹ˆæ™ºèƒ½åœ°è¯†åˆ«ç»†èƒç±»å‹ç„¶åè¿›è¡Œåˆ¤æ–­ã€‚åœ¨å¼‚è´¨æ€§å¾ˆé«˜çš„ç»„ç»‡ä¸­ï¼Œ å°±æ˜¯å­˜åœ¨å†…æºRNAå«é‡ä½ï¼Œè€Œçº¿ç²’ä½“åŸºå› å æ¯”é«˜çš„ç»†èƒã€‚å¦‚æœä½¿ç”¨â€ä¸€åˆ€åˆ‡â€œçš„å›ºå®šé˜ˆå€¼ï¼Œå®ƒä»¬å°±å¾ˆå¯èƒ½ä¼šè¢«å½“æˆç¦»ç¾¤ç‚¹è¢«è¿‡æ»¤ã€‚è€Œæ˜¯ç”¨MADè®¡ç®—æ–¹æ³•æ£€æµ‹çš„ç»“æœå¯èƒ½å°±æ˜¯ï¼šè™½ç„¶ä¸€å †ç»†èƒçš„æŸä¸ªqcæŒ‡æ ‡å·®å¼‚å¾ˆå¤§ï¼Œä½†ä¸­ä½æ•°ä¹Ÿåœ¨å˜ï¼Œéšä¹‹å˜åŒ–çš„è¿˜æœ‰MADå€¼ï¼Œè¿™æ ·æœ€åçš„ç»“æœä¸è‡³äºå’ŒçœŸå®ç”Ÿç‰©å­¦æƒ…å†µå·®å¤ªå¤š</li>
</ul>
<h3 id="è€ƒè™‘å®éªŒçš„å› ç´ ">è€ƒè™‘å®éªŒçš„å› ç´ <a hidden class="anchor" aria-hidden="true" href="#è€ƒè™‘å®éªŒçš„å› ç´ ">#</a></h3>
<p>å¾ˆå¤šå¤§å‹çš„å®éªŒéƒ½åŒ…å«å¤šä¸ªç»†èƒç³»ï¼Œè€Œä¸”å¯èƒ½é‡‡ç”¨çš„å®éªŒæ–¹æ³•ä¸åŒï¼ˆæ¯”å¦‚é€‰ç”¨ä¸åŒçš„æµ‹åºæ·±åº¦ï¼‰ï¼Œè¿™å°±äº§ç”Ÿäº†å®éªŒçš„ä¸åŒæ‰¹æ¬¡ã€‚<strong>è¿™ç§æƒ…å†µä¸‹ï¼Œ åº”è¯¥å¯¹æ¯ä¸ªæ‰¹æ¬¡åˆ†åˆ«è¿›è¡Œæ£€æµ‹ã€‚</strong></p>
<p>å¦‚æœæ¯ä¸ªæ‰¹æ¬¡éƒ½æ˜¯ä¸€ä¸ª<code>SingleCellExperiment</code>å¯¹è±¡ï¼Œé‚£ä¹ˆ<code>isOutlier()</code>å‡½æ•°å¯ä»¥æŒ‰ä¸Šé¢çš„æ–¹æ³•ç›´æ¥ä½¿ç”¨ï¼›ä½†æ˜¯å¦‚æœä¸åŒæ‰¹æ¬¡çš„ç»†èƒå·²ç»æ··åˆæˆä¸€æ•´ä¸ª<code>SingleCellExperiment</code>å¯¹è±¡ï¼Œé‚£ä¹ˆ<code>batch=</code>è¿™ä¸ªå‚æ•°å°±æ´¾ä¸Šç”¨åœºäº†ã€‚</p>
<p>åŒæ ·ä»¥è¿™ä¸ª416Bæ•°æ®é›†ä¸ºä¾‹ï¼Œä»–åŒ…å«äº†ä¸¤ç§ä¸åŒçš„å®éªŒç±»å‹ã€‚ç„¶åæˆ‘ä»¬å°±å¯ä»¥ä½¿ç”¨<code>batch=</code>å‚æ•°å»è¿›è¡Œè´¨æ§ã€‚</p>
<div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>batch <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">paste0</span>(sce.416b<span style="color:#5bc4bf">$</span>phenotype, <span style="color:#48b685">&#34;-&#34;</span>, sce.416b<span style="color:#5bc4bf">$</span>Plate)
</span></span><span style="display:flex;"><span>batch.reasons <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">quickPerCellQC</span>(df, percent_subsets<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">c</span>(<span style="color:#48b685">&#34;subsets_Mito_percent&#34;</span>,
</span></span><span style="display:flex;"><span>    <span style="color:#48b685">&#34;altexps_ERCC_percent&#34;</span>), batch<span style="color:#5bc4bf">=</span>batch)
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">colSums</span>(<span style="color:#06b6ef">as.matrix</span>(batch.reasons))
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#776e71">##              low_lib_size            low_n_features high_subsets_Mito_percent </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##                         4                         2                         2 </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## high_altexps_ERCC_percent                   discard </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##                         4                         7</span>
</span></span></code></pre></div><p><strong>ä½†æ˜¯ï¼Œ<code>batch</code>å‚æ•°ä¸æ˜¯ä¸‡èƒ½çš„</strong>ï¼Œä¹‹å‰ä¹Ÿè¯´è¿‡ï¼Œè¿™ç§ç›¸å¯¹é˜ˆå€¼éœ€è¦ä¸€ä¸ª<strong>å‡è®¾ï¼šæ¯ä¸ªæ‰¹æ¬¡çš„å¤§éƒ¨åˆ†ç»†èƒéƒ½æ˜¯é«˜è´¨é‡çš„</strong>ã€‚å½“æŸä¸ªæ‰¹æ¬¡çš„ç»†èƒæ•´ä½“è´¨é‡åä½ï¼Œç¦»ç¾¤ç‚¹æ£€æµ‹å¾ˆæœ‰å¯èƒ½å¤±è´¥</p>
<p>ä¾‹å¦‚ï¼Œåœ¨Grun et al. (2016)çš„æ•°æ®é›†ä¸­æœ‰ä¸¤ä¸ªdonorçš„å®éªŒæ˜¯å¤±è´¥çš„ã€‚å®ƒä»¬çš„ERCCå«é‡ç›¸å¯¹å…¶ä»–æ‰¹æ¬¡é«˜ï¼Œå¢åŠ äº†ä¸­ä½æ•°å’ŒMADå€¼ï¼Œå¯¼è‡´è¿‡æ»¤ä½è´¨é‡ç»†èƒå¤±è´¥ã€‚å› æ­¤è¿™ç§æƒ…å†µä¸‹ï¼Œå¯ä»¥å…ˆç®—å…¶ä»–å‡ ä¸ªæ‰¹æ¬¡çš„ä¸­ä½æ•°å’ŒMADå€¼ï¼Œç„¶åå‚è€ƒè¿™äº›å€¼å»å¯¹æœ‰é—®é¢˜çš„æ‰¹æ¬¡è¿›è¡Œè¿‡æ»¤ã€‚</p>
<div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#06b6ef">library</span>(scRNAseq)
</span></span><span style="display:flex;"><span>sce.grun <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">GrunPancreasData</span>()
</span></span><span style="display:flex;"><span>sce.grun <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">addPerCellQC</span>(sce.grun)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#776e71"># First attempt with batch-specific thresholds.</span>
</span></span><span style="display:flex;"><span>discard.ercc <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">isOutlier</span>(sce.grun<span style="color:#5bc4bf">$</span>altexps_ERCC_percent,
</span></span><span style="display:flex;"><span>    type<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;higher&#34;</span>, batch<span style="color:#5bc4bf">=</span>sce.grun<span style="color:#5bc4bf">$</span>donor)
</span></span><span style="display:flex;"><span>with.blocking <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">plotColData</span>(sce.grun, x<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;donor&#34;</span>, y<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;altexps_ERCC_percent&#34;</span>,
</span></span><span style="display:flex;"><span>    colour_by<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">I</span>(discard.ercc))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#776e71"># Second attempt, sharing information across batches</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71"># to avoid dramatically different thresholds for unusual batches.</span>
</span></span><span style="display:flex;"><span>discard.ercc2 <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">isOutlier</span>(sce.grun<span style="color:#5bc4bf">$</span>altexps_ERCC_percent,
</span></span><span style="display:flex;"><span>    type<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;higher&#34;</span>, batch<span style="color:#5bc4bf">=</span>sce.grun<span style="color:#5bc4bf">$</span>donor,
</span></span><span style="display:flex;"><span>    subset<span style="color:#5bc4bf">=</span>sce.grun<span style="color:#5bc4bf">$</span>donor <span style="color:#5bc4bf">%in%</span> <span style="color:#06b6ef">c</span>(<span style="color:#48b685">&#34;D17&#34;</span>, <span style="color:#48b685">&#34;D2&#34;</span>, <span style="color:#48b685">&#34;D7&#34;</span>))
</span></span><span style="display:flex;"><span>without.blocking <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">plotColData</span>(sce.grun, x<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;donor&#34;</span>, y<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;altexps_ERCC_percent&#34;</span>,
</span></span><span style="display:flex;"><span>    colour_by<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">I</span>(discard.ercc2))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>gridExtra<span style="color:#5bc4bf">::</span><span style="color:#06b6ef">grid.arrange</span>(with.blocking, without.blocking, ncol<span style="color:#5bc4bf">=</span><span style="color:#f99b15">2</span>) 
</span></span></code></pre></div><p><img loading="lazy" src="https://cdn.jsdelivr.net/gh/Moonerss/CDN/paper/quality-control/qc-plot-pancreas-1.png" alt=""  />
</p>
<blockquote>
<p>æ³¨ï¼šå¯ä»¥çœ‹åˆ°ï¼Œå·¦å›¾æ˜¯æŒ‰ç…§æ¯ä¸ªæ‰¹æ¬¡åˆ†åˆ«é‰´å®šçš„ç¦»ç¾¤ç‚¹ï¼›å³å›¾æ˜¯ç”¨è´¨é‡å¥½çš„æ‰¹æ¬¡è®¡ç®—çš„é˜ˆå€¼ï¼Œç„¶åè¿ç”¨åˆ°å·®çš„æ‰¹æ¬¡ä¸Šçš„ç»“æœ</p>
</blockquote>
<p>ä¸ºäº†é‰´åˆ«æœ‰é—®é¢˜çš„æ‰¹æ¬¡ï¼Œå¯ä»¥å…ˆå°†æ¯ä¸ªæ‰¹æ¬¡åˆ†åˆ«è®¡ç®—ç»“æœï¼Œç„¶åæ¯”è¾ƒå®ƒä»¬çš„é˜ˆå€¼ï¼Œå¦‚æœæ¯”åŒç±»æ‰¹æ¬¡è¶…å‡ºå¤ªå¤šï¼Œå°±è¦è­¦è§‰ã€‚</p>
<div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>ercc.thresholds <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">attr</span>(discard.ercc, <span style="color:#48b685">&#34;thresholds&#34;</span>)[<span style="color:#48b685">&#34;higher&#34;</span>,]
</span></span><span style="display:flex;"><span>ercc.thresholds
</span></span><span style="display:flex;"><span><span style="color:#776e71">##        D10        D17         D2         D3         D7 </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##  73.610696   7.599947   6.010975 113.105828  15.216956</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">names</span>(ercc.thresholds)<span style="color:#06b6ef">[isOutlier</span>(ercc.thresholds, type<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;higher&#34;</span>)]
</span></span><span style="display:flex;"><span><span style="color:#776e71">## [1] &#34;D10&#34; &#34;D3&#34;</span>
</span></span></code></pre></div><blockquote>
<p>å¯ä»¥çœ‹åˆ°D10ã€D3çš„é˜ˆå€¼å°±è¶…è¿‡å…¶ä»–æ‰¹æ¬¡å¾ˆå¤š</p>
</blockquote>
<p>ä½†æ˜¯è¿™ä¹ˆåšçš„å‰æéƒ½æ˜¯ï¼šæˆ‘ä»¬è®¤ä¸ºæ‰¹æ¬¡ä¸­çš„ç»†èƒè´¨é‡æ•´ä½“è¿˜ä¸é”™ã€‚å¦‚æœæˆ‘ä»¬ä¿è¯ä¸äº†ç»†èƒè´¨é‡ï¼Œé‚£ä¹ˆè¿™ç§è®¡ç®—ç›¸å¯¹é˜ˆå€¼çš„æ–¹æ³•å°±ä¸æˆç«‹ï¼Œè¿˜æ˜¯è¦ä½¿ç”¨ç»å¯¹é˜ˆå€¼ï¼Œæ‰‹åŠ¨å»é™¤ã€‚</p>
<h3 id="3-å…¶ä»–æ–¹æ³•">3. å…¶ä»–æ–¹æ³•<a hidden class="anchor" aria-hidden="true" href="#3-å…¶ä»–æ–¹æ³•">#</a></h3>
<p>å¦ä¸€ä¸ªç­–ç•¥æ˜¯æ ¹æ®æ¯ä¸ªç»†èƒçš„QCæŒ‡æ ‡æ¥åœ¨é«˜ç»´ç©ºé—´ä¸­è¯†åˆ«å¼‚å¸¸å€¼ã€‚åˆ©ç”¨<code>robustbase</code> åŒ…ï¼Œå°†ç»†èƒæ”¾åœ¨é«˜ç»´ç©ºé—´ï¼Œæ ¹æ®ä»–ä»¬çš„QCæŒ‡æ ‡ï¼ˆæ–‡åº“å¤§å°ã€è¡¨è¾¾åŸºå› æ•°ã€spike-inæ¯”ä¾‹ã€çº¿ç²’ä½“æ¯”ä¾‹ï¼‰ï¼Œç„¶åä½¿ç”¨<code>isOutlier()</code>æ¥è¯†åˆ«è¡¨ç°å‡ºå¼‚å¸¸é«˜outlylieræ°´å¹³çš„ä½è´¨é‡ç»†èƒ</p>
<div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>stats <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">cbind</span>(<span style="color:#06b6ef">log10</span>(df<span style="color:#5bc4bf">$</span>sum), <span style="color:#06b6ef">log10</span>(df<span style="color:#5bc4bf">$</span>detected),
</span></span><span style="display:flex;"><span>    df<span style="color:#5bc4bf">$</span>subsets_Mito_percent, df<span style="color:#5bc4bf">$</span>altexps_ERCC_percent)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">library</span>(robustbase)
</span></span><span style="display:flex;"><span>outlying <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">adjOutlyingness</span>(stats, only.outlyingness <span style="color:#5bc4bf">=</span> <span style="color:#815ba4">TRUE</span>)
</span></span><span style="display:flex;"><span>multi.outlier <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">isOutlier</span>(outlying, type <span style="color:#5bc4bf">=</span> <span style="color:#48b685">&#34;higher&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">summary</span>(multi.outlier)
</span></span><span style="display:flex;"><span><span style="color:#776e71">##    Mode   FALSE    TRUE </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## logical     182      10</span>
</span></span></code></pre></div><p>é™¤æ­¤ä»¥å¤–ï¼Œæœ‰æ—¶<strong>è¿˜å¯ä»¥æ ¹æ®åŸºå› è¡¨è¾¾é‡è¿›è¡Œè¿‡æ»¤</strong>ï¼Œä¸è¿‡åœ¨bulkè½¬å½•ç»„ä¸­ç”¨çš„æ¯”è¾ƒå¤šï¼Œä½†æ˜¯åœ¨scRNAä¸­è¿™æ ·æ“ä½œå¯èƒ½ä¼šæŸå¤±ä¸€äº›æœ¬èº«æ•°é‡å°±æ¯”è¾ƒå°‘çš„é«˜è´¨é‡ç»†èƒç¾¤ä½“ï¼ˆæ¯”å¦‚ä¸€äº›ç½•è§ç»†èƒï¼Œæœ¬èº«åŸºå› è¡¨è¾¾é‡å°±ä¸æ˜¯å¾ˆé«˜ï¼‰</p>
<h2 id="ç”»å›¾æ£€æŸ¥">ç”»å›¾æ£€æŸ¥<a hidden class="anchor" aria-hidden="true" href="#ç”»å›¾æ£€æŸ¥">#</a></h2>
<p>æ£€æŸ¥QCåº¦é‡çš„åˆ†å¸ƒä»¥ç¡®å®šå¯èƒ½çš„é—®é¢˜æ˜¯ä¸€ä¸ªå¾ˆå¥½çš„å®è·µã€‚åœ¨æœ€ç†æƒ³çš„æƒ…å†µä¸‹ï¼Œæˆ‘ä»¬ä¼šçœ‹åˆ°æ­£æ€åˆ†å¸ƒï¼Œå¯ä»¥è¯æ˜åœ¨ç¦»ç¾¤å€¼æ£€æµ‹ä¸­ä½¿ç”¨çš„<code>3 MAD</code>é˜ˆå€¼æ˜¯åˆç†çš„ã€‚å¦ä¸€ç§æ¨¡å¼ä¸‹çš„å¤§é‡ç»†èƒè¡¨æ˜QCæŒ‡æ ‡å¯èƒ½ä¸æŸäº›ç”Ÿç‰©çŠ¶æ€ç›¸å…³ï¼Œå¯èƒ½å¯¼è‡´è¿‡æ»¤è¿‡ç¨‹ä¸­ä¸åŒç»†èƒç±»å‹çš„ä¸¢å¤±;æˆ–è€…æœ‰ä¸ä¸€è‡´çš„åº“å‡†å¤‡ä¸ºä¸€ä¸ªå­é›†çš„ç»†èƒï¼Œä¸€ä¸ªéç½•è§çš„ç°è±¡åœ¨æ¿çš„åè®®ã€‚</p>
<div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#776e71"># æŠŠQCæŒ‡æ ‡å’ŒåŸæ¥çš„sce.416bç»†èƒä¿¡æ¯æ•´åˆèµ·æ¥</span>
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">colData</span>(sce.416b) <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">cbind</span>(<span style="color:#06b6ef">colData</span>(sce.416b), df)
</span></span><span style="display:flex;"><span><span style="color:#776e71"># ä¿®æ”¹ä¸€ä¸‹æ•´åˆåçš„ä¿¡æ¯</span>
</span></span><span style="display:flex;"><span>sce.416b<span style="color:#5bc4bf">$</span>block <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">factor</span>(sce.416b<span style="color:#5bc4bf">$</span>block)
</span></span><span style="display:flex;"><span>sce.416b<span style="color:#5bc4bf">$</span>phenotype <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">ifelse</span>(<span style="color:#06b6ef">grepl</span>(<span style="color:#48b685">&#34;induced&#34;</span>, sce.416b<span style="color:#5bc4bf">$</span>phenotype),
</span></span><span style="display:flex;"><span>    <span style="color:#48b685">&#34;induced&#34;</span>, <span style="color:#48b685">&#34;wild type&#34;</span>)
</span></span><span style="display:flex;"><span>sce.416b<span style="color:#5bc4bf">$</span>discard <span style="color:#5bc4bf">&lt;-</span> reasons<span style="color:#5bc4bf">$</span>discard
</span></span><span style="display:flex;"><span><span style="color:#776e71"># ç»˜å›¾</span>
</span></span><span style="display:flex;"><span>gridExtra<span style="color:#5bc4bf">::</span><span style="color:#06b6ef">grid.arrange</span>(
</span></span><span style="display:flex;"><span>    <span style="color:#06b6ef">plotColData</span>(sce.416b, x<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;block&#34;</span>, y<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;sum&#34;</span>, colour_by<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;discard&#34;</span>,
</span></span><span style="display:flex;"><span>        other_fields<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;phenotype&#34;</span>) <span style="color:#5bc4bf">+</span> <span style="color:#06b6ef">facet_wrap</span>(<span style="color:#5bc4bf">~</span>phenotype) <span style="color:#5bc4bf">+</span> 
</span></span><span style="display:flex;"><span>        <span style="color:#06b6ef">scale_y_log10</span>() <span style="color:#5bc4bf">+</span> <span style="color:#06b6ef">ggtitle</span>(<span style="color:#48b685">&#34;Total count&#34;</span>),
</span></span><span style="display:flex;"><span>    <span style="color:#06b6ef">plotColData</span>(sce.416b, x<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;block&#34;</span>, y<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;detected&#34;</span>, colour_by<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;discard&#34;</span>, 
</span></span><span style="display:flex;"><span>        other_fields<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;phenotype&#34;</span>) <span style="color:#5bc4bf">+</span> <span style="color:#06b6ef">facet_wrap</span>(<span style="color:#5bc4bf">~</span>phenotype) <span style="color:#5bc4bf">+</span> 
</span></span><span style="display:flex;"><span>        <span style="color:#06b6ef">scale_y_log10</span>() <span style="color:#5bc4bf">+</span> <span style="color:#06b6ef">ggtitle</span>(<span style="color:#48b685">&#34;Detected features&#34;</span>),
</span></span><span style="display:flex;"><span>    <span style="color:#06b6ef">plotColData</span>(sce.416b, x<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;block&#34;</span>, y<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;subsets_Mito_percent&#34;</span>, 
</span></span><span style="display:flex;"><span>        colour_by<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;discard&#34;</span>, other_fields<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;phenotype&#34;</span>) <span style="color:#5bc4bf">+</span> 
</span></span><span style="display:flex;"><span>        <span style="color:#06b6ef">facet_wrap</span>(<span style="color:#5bc4bf">~</span>phenotype) <span style="color:#5bc4bf">+</span> <span style="color:#06b6ef">ggtitle</span>(<span style="color:#48b685">&#34;Mito percent&#34;</span>),
</span></span><span style="display:flex;"><span>    <span style="color:#06b6ef">plotColData</span>(sce.416b, x<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;block&#34;</span>, y<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;altexps_ERCC_percent&#34;</span>, 
</span></span><span style="display:flex;"><span>        colour_by<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;discard&#34;</span>, other_fields<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;phenotype&#34;</span>) <span style="color:#5bc4bf">+</span> 
</span></span><span style="display:flex;"><span>        <span style="color:#06b6ef">facet_wrap</span>(<span style="color:#5bc4bf">~</span>phenotype) <span style="color:#5bc4bf">+</span> <span style="color:#06b6ef">ggtitle</span>(<span style="color:#48b685">&#34;ERCC percent&#34;</span>),
</span></span><span style="display:flex;"><span>    ncol<span style="color:#5bc4bf">=</span><span style="color:#f99b15">1</span>
</span></span><span style="display:flex;"><span>)
</span></span></code></pre></div><p><img loading="lazy" src="https://cdn.jsdelivr.net/gh/Moonerss/CDN/paper/quality-control/qc-dist-416b-1.png" alt=""  />
</p>
<blockquote>
<p>å±•ç¤ºçš„æ˜¯ä¸åŒæ‰¹æ¬¡çš„QCæŒ‡æ ‡</p>
</blockquote>
<p>å¦ä¸€ç§æœ‰ç”¨çš„è¯Šæ–­æ–¹æ³•æ˜¯ç»˜åˆ¶ç›¸å¯¹äºå…¶ä»–QCæŒ‡æ ‡çš„çº¿ç²’ä½“è®¡æ•°æ¯”ä¾‹å›¾ã€‚</p>
<p>ä¸ºäº†ç¡®ä¿ä¸å­˜åœ¨è¿™æ ·çš„ç»†èƒï¼šè™½ç„¶ç»†èƒæ–‡åº“å¤§ï¼Œä½†å®ƒçš„çº¿ç²’ä½“å æ¯”ä¹Ÿé«˜ã€‚å¦å¤–ä¹Ÿæ˜¯ä¸ºäº†é¿å…æ„å¤–è¿‡æ»¤æ‰æœ¬æ¥æ˜¯é«˜è´¨é‡ä½†åŒæ—¶å…·æœ‰é«˜ä»£è°¢æ´»æ€§ï¼ˆå³é«˜çº¿ç²’ä½“å æ¯”ï¼‰çš„ç»†èƒï¼ˆå¦‚è‚è„ç»†èƒï¼‰</p>
<p>======æœªå®Œ</p>
<h2 id="é’ˆå¯¹dropletæ•°æ®çš„ç»†èƒåˆ¤æ–­">é’ˆå¯¹Dropletæ•°æ®çš„ç»†èƒåˆ¤æ–­<a hidden class="anchor" aria-hidden="true" href="#é’ˆå¯¹dropletæ•°æ®çš„ç»†èƒåˆ¤æ–­">#</a></h2>
<h3 id="èƒŒæ™¯">èƒŒæ™¯<a hidden class="anchor" aria-hidden="true" href="#èƒŒæ™¯">#</a></h3>
<p>ç”±äºè¿™ç§å»ºåº“æ–¹æ³•çš„ç‹¬ç‰¹æ€§ï¼Œæˆ‘ä»¬æ²¡åŠæ³•äº‹å…ˆçŸ¥é“æŸä¸€ä¸ªç»†èƒæ–‡åº“ï¼ˆæ¯”å¦‚ä¸€ä¸ªcell barcodeï¼‰æ˜¯çœŸæ­£åŒ…å«ä¸€ä¸ªç»†èƒè¿˜æ˜¯åªæ˜¯ä¸€ä¸ªç©ºçš„æ¶²æ»´ï¼ˆdropletï¼‰ã€‚å› æ­¤ï¼Œ<strong>ç¬¬ä¸€æ­¥æ˜¯éœ€è¦æ ¹æ®å¾—åˆ°çš„è¡¨è¾¾é‡ä¿¡æ¯ï¼Œæ¥æ¨æ–­æ¶²æ»´æ˜¯ä¸æ˜¯ç©ºçš„</strong>ã€‚ä½†ä»…ä»…æ ¹æ®è¡¨è¾¾é‡åˆ¤æ–­è¿˜æ˜¯ä¸å¤ªé è°±ï¼Œå› ä¸ºè¿˜æœ‰å¯èƒ½åœ¨ç©ºçš„æ¶²æ»´ä¸­ä¾ç„¶åŒ…å«å¤–æºRNAï¼Œæœ€åçš„ç»†èƒæ–‡åº“ä¾æ—§ä¸ä¸º0ï¼Œä½†ç¡®å®ä¸åŒ…å«ç»†èƒã€‚</p>
<p>è¿™é‡Œä¸ºäº†è¯´æ˜è¿™ä¸ªé—®é¢˜ï¼Œ<strong>ä½¿ç”¨äº†ä¸€ä¸ªæœªè¿‡æ»¤çš„10X PBMCæ•°æ®</strong></p>
<div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#776e71"># æ•°æ®ä¸‹è½½</span>
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">library</span>(BiocFileCache)
</span></span><span style="display:flex;"><span>bfc <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">BiocFileCache</span>(<span style="color:#48b685">&#34;raw_data&#34;</span>, ask <span style="color:#5bc4bf">=</span> <span style="color:#815ba4">FALSE</span>)
</span></span><span style="display:flex;"><span>raw.path <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">bfcrpath</span>(bfc, <span style="color:#06b6ef">file.path</span>(<span style="color:#48b685">&#34;http://cf.10xgenomics.com/samples&#34;</span>,
</span></span><span style="display:flex;"><span>    <span style="color:#48b685">&#34;cell-exp/2.1.0/pbmc4k/pbmc4k_raw_gene_bc_matrices.tar.gz&#34;</span>))
</span></span><span style="display:flex;"><span><span style="color:#776e71"># è§£å‹æ•°æ®</span>
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">untar</span>(raw.path, exdir<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">file.path</span>(<span style="color:#06b6ef">tempdir</span>(), <span style="color:#48b685">&#34;pbmc4k&#34;</span>))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">library</span>(DropletUtils)
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">library</span>(Matrix)
</span></span><span style="display:flex;"><span>fname <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">file.path</span>(<span style="color:#06b6ef">tempdir</span>(), <span style="color:#48b685">&#34;pbmc4k/raw_gene_bc_matrices/GRCh38&#34;</span>)
</span></span><span style="display:flex;"><span>sce.pbmc <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">read10xCounts</span>(fname, col.names<span style="color:#5bc4bf">=</span><span style="color:#815ba4">TRUE</span>)
</span></span><span style="display:flex;"><span>sce.pbmc
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#776e71">## class: SingleCellExperiment </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## dim: 33694 737280 </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## metadata(1): Samples</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## assays(1): counts</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## rownames(33694): ENSG00000243485 ENSG00000237613 ... ENSG00000277475</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##   ENSG00000268674</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## rowData names(2): ID Symbol</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## colnames(737280): AAACCTGAGAAACCAT-1 AAACCTGAGAAACCGC-1 ...</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##   TTTGTCATCTTTAGTC-1 TTTGTCATCTTTCCTC-1</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## colData names(2): Sample Barcode</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## reducedDimNames(0):</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## altExpNames(0):</span>
</span></span></code></pre></div><p><strong>æ•´ä½“è§‚å¯Ÿä¸åŒçš„barcodesï¼ˆä¸ä¸€å®šéƒ½æ˜¯çœŸçš„ç»†èƒï¼‰çš„æ–‡åº“å¤§å°åˆ†å¸ƒï¼š</strong></p>
<div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>bcrank <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">barcodeRanks</span>(<span style="color:#06b6ef">counts</span>(sce.pbmc))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#776e71"># Only showing unique points for plotting speed.</span>
</span></span><span style="display:flex;"><span>uniq <span style="color:#5bc4bf">&lt;-</span> <span style="color:#5bc4bf">!</span><span style="color:#06b6ef">duplicated</span>(bcrank<span style="color:#5bc4bf">$</span>rank)
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">plot</span>(bcrank<span style="color:#5bc4bf">$</span>rank[uniq], bcrank<span style="color:#5bc4bf">$</span>total[uniq], log<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;xy&#34;</span>,
</span></span><span style="display:flex;"><span>    xlab<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;Rank&#34;</span>, ylab<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;Total UMI count&#34;</span>, cex.lab<span style="color:#5bc4bf">=</span><span style="color:#f99b15">1.2</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">abline</span>(h<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">metadata</span>(bcrank)<span style="color:#5bc4bf">$</span>inflection, col<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;darkgreen&#34;</span>, lty<span style="color:#5bc4bf">=</span><span style="color:#f99b15">2</span>)
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">abline</span>(h<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">metadata</span>(bcrank)<span style="color:#5bc4bf">$</span>knee, col<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;dodgerblue&#34;</span>, lty<span style="color:#5bc4bf">=</span><span style="color:#f99b15">2</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">legend</span>(<span style="color:#48b685">&#34;bottomleft&#34;</span>, legend<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">c</span>(<span style="color:#48b685">&#34;Inflection&#34;</span>, <span style="color:#48b685">&#34;Knee&#34;</span>), 
</span></span><span style="display:flex;"><span>        col<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">c</span>(<span style="color:#48b685">&#34;darkgreen&#34;</span>, <span style="color:#48b685">&#34;dodgerblue&#34;</span>), lty<span style="color:#5bc4bf">=</span><span style="color:#f99b15">2</span>, cex<span style="color:#5bc4bf">=</span><span style="color:#f99b15">1.2</span>)
</span></span></code></pre></div><p><img loading="lazy" src="https://cdn.jsdelivr.net/gh/Moonerss/CDN/paper/quality-control/rankplot-1.png" alt=""  />
</p>
<blockquote>
<p>çœ‹åˆ°å„ä¸ªbarcodesçš„æ–‡åº“å¤§å°æœ‰é«˜æœ‰ä½ï¼Œå¹¶ä¸”ç›¸å·®è¾ƒå¤§ï¼Œå› æ­¤å¯èƒ½å¯¹åº”ç€çœŸå®å­˜åœ¨çš„ç»†èƒå’Œç©ºæ¶²æ»´ã€‚å½“ç„¶æœ€ç®€å•çš„åŠæ³•è¿˜æ˜¯â€ä¸€åˆ€åˆ‡â€œï¼Œ<strong>ç•™ä¸‹é‚£äº›æ–‡åº“è¾ƒå¤§çš„ç»†èƒ</strong>ï¼Œä¸è¿‡è¿˜æ˜¯æœ‰æŸå¤±çœŸå®ç»†èƒç±»å‹çš„é£é™©</p>
</blockquote>
<h3 id="æ£€æµ‹ç©ºçš„æ¶²æ»´">æ£€æµ‹ç©ºçš„æ¶²æ»´<a hidden class="anchor" aria-hidden="true" href="#æ£€æµ‹ç©ºçš„æ¶²æ»´">#</a></h3>
<p>æˆ‘ä»¬ä½¿ç”¨<code>emptyDrops()</code>å‡½æ•°ï¼Œæ£€æŸ¥æ¯ä¸ªbarcodeçš„è¡¨è¾¾é‡æ˜¯ä¸æ˜¯å’Œå¤–æºRNAçš„è¡¨è¾¾é‡æœ‰æ˜¾è‘—å·®å¼‚(Lun et al. 2019)ã€‚å¦‚æœå­˜åœ¨æ˜¾è‘—å·®å¼‚ï¼Œå°±è¯´æ˜barcodeä¸­æ›´æœ‰å¯èƒ½æ˜¯ä¸€ä¸ªç»†èƒã€‚è¿™ç§æ–¹æ³•å¯<strong>ä»¥å¸®åŠ©åŒºåˆ†ï¼šæµ‹åºè´¨é‡å¥½çš„ç©ºæ¶²æ»´ å’Œ åŒ…å«ç»†èƒä½†RNAå«é‡è¾ƒå°‘çš„æ¶²æ»´</strong>ã€‚å°½ç®¡å®ƒä»¬æ€»ä½“çš„è¡¨è¾¾é‡å¯èƒ½å¾ˆç›¸ä¼¼ï¼Œä½†æœ¬è´¨ä¸åŒï¼Œè¿˜æ˜¯è¦åŒºåˆ†çš„ã€‚</p>
<p><code>emptyDrops()</code> ä½¿ç”¨Monte Carloç»Ÿè®¡æ¨¡æ‹Ÿè®¡ç®—på€¼ï¼Œå¦‚æœè¦é‡å¤ç»“æœï¼Œéœ€è¦è®¾ç½®éšæœºç§å­ã€‚å¦å¤–è®¾ç½® false discovery rate (FDR)ä¸º0.1%ï¼Œæ„å‘³ç€ä¸è¶…è¿‡0.1%çš„barcodesæ˜¯ç©ºçš„ã€‚</p>


        </div>

        

        <footer class="post-footer">
            
<nav class="paginav">
  <a class="prev" href="http://localhost:1313/en/posts/tech/ggplot2/">
    <span class="title">Â« ä¸Šä¸€é¡µ</span>
    <br>
    <span>ggplot2 | å°éƒ¨ä»¶å‚è€ƒ</span>
  </a>
  <a class="next" href="http://localhost:1313/en/posts/read/r_stat/">
    <span class="title">ä¸‹ä¸€é¡µ Â»</span>
    <br>
    <span>R è¯­è¨€ç»Ÿè®¡æ£€éªŒå‡½æ•°æ±‡æ€»</span>
  </a>
</nav>

        </footer>
    </div>



<div>
    <div class="pagination__title">
        <span class="pagination__title-h" style="font-size: 20px;">ğŸ’¬è¯„è®º</span>
        <hr />
    </div>

    <div id="tcomment"></div>

    <script src="https://cdn.staticfile.org/twikoo/1.6.15/twikoo.all.min.js">
    </script>
    

    

    <script>
        twikoo.init({
            envId: "https://twikoo.lvbibir.cn/", 
            el: "#tcomment",
            lang: 'zh-CN',
            
            path: window.TWIKOO_MAGIC_PATH||window.location.pathname,
            
            
            
            
            
            
            
        });
    </script>

</div>
</article>
</main>


<footer class="footer">

    <a href="https://gohugo.io/" target="_blank">
        
        <img style="display: unset;" src="https://image.lvbibir.cn/blog/frame-hugo-blue.svg">
    </a>
    <a href="https://github.com/adityatelange/hugo-PaperMod" target="_blank">
        
        <img style="display: unset;" src="https://image.lvbibir.cn/blog/theme-papermod-lightgrey.svg">
    </a>

    <br>

    <span id="runtime_span"></span>
    <script
        type="text/javascript">function show_runtime() { window.setTimeout("show_runtime()", 1000); X = new Date("7/13/2021 1:00:00"); Y = new Date(); T = (Y.getTime() - X.getTime()); M = 24 * 60 * 60 * 1000; a = T / M; A = Math.floor(a); b = (a - A) * 24; B = Math.floor(b); c = (b - B) * 60; C = Math.floor((b - B) * 60); D = Math.floor((c - C) * 60); runtime_span.innerHTML = "ç½‘ç«™å·²è¿è¡Œ" + A + "å¤©" + B + "å°æ—¶"} show_runtime();</script>
    |
    <script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
    <span id="busuanzi_container">
        
        æ€»è®¿å®¢æ•°:  <span id="busuanzi_value_site_uv"></span>
        |
        æ€»è®¿é—®é‡:  <span id="busuanzi_value_site_pv"></span>
    </span>

    <br>
    <span>
        Copyright
        &copy;
        2021-2024
        <a href="http://localhost:1313/en/" style="border-bottom:1px solid">Jeason&#39;s Blog</a>
    </span>

    
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <span class="topInner">
        <svg class="topSvg" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
            <path d="M12 6H0l6-6z" />
        </svg>
        <span id="read_progress"></span>
    </span>
</a>

<script>
    document.addEventListener('scroll', function (e) {
        const readProgress = document.getElementById("read_progress");
        const scrollHeight = document.documentElement.scrollHeight;
        const clientHeight = document.documentElement.clientHeight;
        const scrollTop = document.documentElement.scrollTop || document.body.scrollTop;
        
        readProgress.innerText = ((scrollTop / (scrollHeight - clientHeight)).toFixed(2) * 100).toFixed(0);
    })
</script>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    let mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 400 || document.documentElement.scrollTop > 400) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };
</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>


<script>
    document.querySelectorAll('pre > code').forEach((codeblock) => {
        const container = codeblock.parentNode.parentNode;

        const copybutton = document.createElement('button');
        copybutton.classList.add('copy-code');
        copybutton.innerText = 'ğŸ“„å¤åˆ¶';

        function copyingDone() {
            copybutton.innerText = 'ğŸ‘ŒğŸ»å·²å¤åˆ¶!';
            setTimeout(() => {
                copybutton.innerText = 'ğŸ“„å¤åˆ¶';
            }, 2000);
        }

        copybutton.addEventListener('click', (cb) => {
            
            
            
            
            
            
            
            
            
            
            
            
            
            

            const range = document.createRange();
            range.selectNodeContents(codeblock);
            const selection = window.getSelection();
            selection.removeAllRanges();
            selection.addRange(range);
            try {
                document.execCommand('copy');
                copyingDone();
            } catch (e) {
            }
            ;
            selection.removeRange(range);
        });

        let language = codeblock.className.replaceAll("language-", "")
        let macTool = document.createElement("div")
        let macTool1 = document.createElement("div")
        let macTool2 = document.createElement("div")
        let macTool3 = document.createElement("div")
        let languageType = document.createElement("div")
        languageType.innerText = language
        macTool.setAttribute('class', 'mac-tool')
        macTool1.setAttribute('class', 'mac bb1')
        macTool2.setAttribute('class', 'mac bb2')
        macTool3.setAttribute('class', 'mac bb3')
        languageType.setAttribute('class', 'language-type')
        macTool.appendChild(macTool1)
        macTool.appendChild(macTool2)
        macTool.appendChild(macTool3)
        macTool.appendChild(languageType)

        if (container.classList.contains("highlight")) {
            container.appendChild(copybutton);
            container.appendChild(macTool)
        } else if (container.parentNode.firstChild === container) {
            
        } else if (codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName === "TABLE") {
            
            codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(copybutton);
            container.appendChild(macTool)
        } else {
            
            codeblock.parentNode.appendChild(copybutton);
            container.appendChild(macTool)
        }
    });
</script>

</body>

</html>
