<!DOCTYPE html>
<html lang="en" dir="auto">

<head><script src="/livereload.js?mindelay=10&amp;v=2&amp;port=1313&amp;path=livereload" data-no-instant defer></script><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">
<title>scRNA的细胞质控 | Jeason&#39;s Blog</title>
<meta name="keywords" content="SingleCell">
<meta name="description" content="动机 scRNA-seq数据的低质量文库可能来自于：细胞分选环节的破坏、文库制备失误（没有足够的反转录或PCR次数）… 表现在：细胞总表达量低、基本没有表达的基因、高线粒体或spike-in占比。 这些低质量的库是有问题的，因为它们可能在下游分析中导致误导的结果 。 聚类问题： 低质量的细胞">
<meta name="author" content="作者:Jeason">
<link rel="canonical" href="http://localhost:1313/en/posts/tech/sc_qc/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.e029fa55ed0a02bb684da89d2a53a4e1ba9e26e3719fa48460be52776460b523.css" integrity="sha256-4Cn6Ve0KArtoTaidKlOk4bqeJuNxn6SEYL5Sd2RgtSM=" rel="preload stylesheet" as="style">
<link rel="icon" href="http://localhost:1313/img/avatar.jpeg">
<link rel="icon" type="image/png" sizes="16x16" href="http://localhost:1313/img/avatar.jpeg">
<link rel="icon" type="image/png" sizes="32x32" href="http://localhost:1313/img/avatar.jpeg">
<link rel="apple-touch-icon" href="http://localhost:1313/img/avatar.jpeg">
<link rel="mask-icon" href="http://localhost:1313/img/avatar.jpeg">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="http://localhost:1313/en/posts/tech/sc_qc/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --code-block-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><script type="text/javascript"
        async
        src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
  tex2jax: {
    inlineMath: [['$','$'], ['\\(','\\)']],
    displayMath: [['$$','$$'], ['\[\[','\]\]']],
    processEscapes: true,
    processEnvironments: true,
    skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
    TeX: { equationNumbers: { autoNumber: "AMS" },
         extensions: ["AMSmath.js", "AMSsymbols.js"] }
  }
});

MathJax.Hub.Queue(function() {
    
    
    
    var all = MathJax.Hub.getAllJax(), i;
    for(i = 0; i < all.length; i += 1) {
        all[i].SourceElement().parentNode.className += ' has-jax';
    }
});
</script>

<style>
code.has-jax {
    font: inherit;
    font-size: 100%;
    background: inherit;
    border: inherit;
    color: #515151;
}
</style>


<script>
var _hmt = _hmt || [];
(function() {
  var hm = document.createElement("script");
  hm.src = ""; 
  var s = document.getElementsByTagName("script")[0]; 
  s.parentNode.insertBefore(hm, s);
})();
</script>



<meta property="og:title" content="scRNA的细胞质控" />
<meta property="og:description" content="动机 scRNA-seq数据的低质量文库可能来自于：细胞分选环节的破坏、文库制备失误（没有足够的反转录或PCR次数）… 表现在：细胞总表达量低、基本没有表达的基因、高线粒体或spike-in占比。 这些低质量的库是有问题的，因为它们可能在下游分析中导致误导的结果 。 聚类问题： 低质量的细胞" />
<meta property="og:type" content="article" />
<meta property="og:url" content="http://localhost:1313/en/posts/tech/sc_qc/" />
<meta property="og:image" content="https://source.unsplash.com/random/400x200?code" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2020-08-12T16:50:44+08:00" />
<meta property="article:modified_time" content="2024-02-29T16:50:44+08:00" />

<meta name="twitter:card" content="summary_large_image" />
<meta name="twitter:image" content="https://source.unsplash.com/random/400x200?code" />
<meta name="twitter:title" content="scRNA的细胞质控"/>
<meta name="twitter:description" content="动机 scRNA-seq数据的低质量文库可能来自于：细胞分选环节的破坏、文库制备失误（没有足够的反转录或PCR次数）… 表现在：细胞总表达量低、基本没有表达的基因、高线粒体或spike-in占比。 这些低质量的库是有问题的，因为它们可能在下游分析中导致误导的结果 。 聚类问题： 低质量的细胞"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "📚 文章",
      "item": "http://localhost:1313/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "👨🏻‍💻 技术",
      "item": "http://localhost:1313/en/posts/tech/"
    }, 
    {
      "@type": "ListItem",
      "position":  3 ,
      "name": "scRNA的细胞质控",
      "item": "http://localhost:1313/en/posts/tech/sc_qc/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "scRNA的细胞质控",
  "name": "scRNA的细胞质控",
  "description": "动机 scRNA-seq数据的低质量文库可能来自于：细胞分选环节的破坏、文库制备失误（没有足够的反转录或PCR次数）… 表现在：细胞总表达量低、基本没有表达的基因、高线粒体或spike-in占比。 这些低质量的库是有问题的，因为它们可能在下游分析中导致误导的结果 。 聚类问题： 低质量的细胞",
  "keywords": [
    "SingleCell"
  ],
  "articleBody": "动机 scRNA-seq数据的低质量文库可能来自于：细胞分选环节的破坏、文库制备失误（没有足够的反转录或PCR次数）… 表现在：细胞总表达量低、基本没有表达的基因、高线粒体或spike-in占比。\n这些低质量的库是有问题的，因为它们可能在下游分析中导致误导的结果 。\n聚类问题： 低质量的细胞会聚集成一群，对结果的解释造成干扰，因为从这群细胞中得不到什么有用的信息，但是它的确也是一群。这种现象产生的原因有可能是：细胞破坏以后，线粒体或核RNAs占比升高。 最差的情况就是：不同类型的低质量细胞，也能聚在一起，因为相比于固有的生物差异，更相似的低质量让它们相依相偎。除此以外，本来非常小的细胞文库也能聚成一群，因为log转换后它们的平均表达量会发生很大的变化(A. Lun 2018). 方差估计或主成分分析:首先在PCA分析时，低质量和高质量之间的差异相比于生物学差异会体现更明显，占据主要的成分，减少降维结果的可靠性。另外，某个基因可能在两个细胞之间没什么表达差异，但是如果所处环境差异很大（一个细胞质量很低，另一个细胞质量正常），那么在差异估算过程中，就会把这个差异也会被当成差异表达基因。例如：一个低质量细胞文库的总表达量非常低（接近0），但恰巧还存在一个基因有表达量，那么这个基因的表达量在后续的文库归一化过程中就会尤为突出 奇怪的转录本上调：实验难免会混入外源的污染转录本，但这个量很少并且在所有细胞中都应该是差不多水平的。但如果某个细胞质量低，文库小，那么在校正文库差异过程中，其中的污染转录本表达量就会“突飞猛进”，看起来是一些明显上调的“奇怪基因”。实际上，这些奇怪的基因依然在其他细胞中存在，并且真实的表达量差不多，并且是不应该占据主体地位的。 为了最大程度避免上面一种或多种情况的发生，应该在归一化之前去掉这些低质量的细胞，这个过程就是**细胞的质控 **\n使用A. T. L. Lun et al. (2017)的小型scRNA数据（未进行QC）进行测试\n#--- loading ---# library(scRNAseq) sce.416b \u003c- LunSpikeInData(which=\"416b\") table(sce.416b$block) 20160113 20160325 96 96 sce.416b$block \u003c- factor(sce.416b$block) sce.416b ## class: SingleCellExperiment ## dim: 46604 192 ## metadata(0): ## assays(1): counts ## rownames(46604): ENSMUSG00000102693 ENSMUSG00000064842 ... ## ENSMUSG00000095742 CBFB-MYH11-mcherry ## rowData names(1): Length ## colnames(192): SLX-9555.N701_S502.C89V9ANXX.s_1.r_1 ## SLX-9555.N701_S503.C89V9ANXX.s_1.r_1 ... ## SLX-11312.N712_S508.H5H5YBBXX.s_8.r_1 ## SLX-11312.N712_S517.H5H5YBBXX.s_8.r_1 ## colData names(9): Source Name cell line ... spike-in addition block ## reducedDimNames(0): ## altExpNames(2): ERCC SIRV QC指标的选择 鉴定细胞是否是低质量的，需要用到几个指标。虽然下面这些指标是使用SMART-seq2数据进行展示的，但依然适用于UMI数据（比如MARS-seq、droplet-based技术）\n文库大小（library size） ：指的是每个细胞中所有基因的表达量之和。细胞的文库如果很小，说明在文库制备过程中存在RNA的损失，要么是由于细胞裂解，要么是cDNA捕获不足导致后续的扩增量少 每个细胞中表达基因的数目（number of expressed features in each cell）：指的是细胞中非0表达量的基因数量。如果细胞中基本没有基因表达，很可能是转录本捕获失败 比对到spike-in转录本的reads比例（proportion of reads mapped to spike-in transcripts）：计算方法是：spike-in counts / all features (including spike-ins) for each cell。每个细胞都应该加入等量的外源转录本（spike-in），如果哪个细胞的spike-in比例提高了，说明它的内源RNA含量减少（比如在细胞分选阶段出现的细胞裂解或者RNA降解） 比对到线粒体基因组的reads比例（proportion of reads mapped to genes in the mitochondrial genome） ：如果没有spike-in，那么使用线粒体指标也是能说明问题的(Islam et al. 2014; Ilicic et al. 2016)。比对到线粒体的reads增多，说明细胞质中的RNA减少，可能存在细胞穿孔的情况，而这个孔的大小，可能只是将细胞质中存在的mRNA流出去，但线粒体的体积超过了孔的大小，所以还留在了细胞中，造成一定程度的富集，导致线粒体RNA占比升高。 对于每个细胞，可以用scater包的perCellQCMetrics()函数进行计算，其中sum这一列表示每个细胞的文库大小；detected这一列表示检测到的基因数量；subsets_Mito_percent这一列表示比对到线粒体基因组的reads占比；altexps_ERCC_percent表示比对到ERCC spike-in的reads占比\n# Retrieving the mitochondrial transcripts using genomic locations included in # the row-level annotation for the SingleCellExperiment. location \u003c- rowRanges(sce.416b) is.mito \u003c- any(seqnames(location)==\"MT\") # ALTERNATIVELY: using resources in AnnotationHub to retrieve chromosomal # locations given the Ensembl IDs; this should yield the same result. library(AnnotationHub) ens.mm.v97 \u003c- AnnotationHub()[[\"AH73905\"]] chr.loc \u003c- mapIds(ens.mm.v97, keys=rownames(sce.416b), keytype=\"GENEID\", column=\"SEQNAME\") is.mito.alt \u003c- which(chr.loc==\"MT\") library(scater) df \u003c- perCellQCMetrics(sce.416b, subsets=list(Mito=is.mito)) df ## DataFrame with 192 rows and 16 columns ## sum detected percent_top_50 percent_top_100 percent_top_200 ## ## 1 865936 7618 26.7218 32.2773 39.7208 ## 2 1076277 7521 29.4043 35.0354 42.2581 ## 3 1180138 8306 27.3454 32.4770 39.3296 ## 4 1342593 8143 35.8092 40.2666 46.2460 ## 5 1668311 7154 34.1198 39.0901 45.6660 ## ... ... ... ... ... ... ## 188 776622 8174 45.9362 49.7010 54.6101 ## 189 1299950 8956 38.0829 42.8930 49.0622 ## 190 1800696 9530 30.6675 35.5839 41.8550 ## 191 46731 6649 32.2998 37.9149 44.5999 ## 192 1866692 10964 26.6632 31.2584 37.5608 ## percent_top_500 subsets_Mito_sum subsets_Mito_detected subsets_Mito_percent ## ## 1 52.9038 78790 20 9.09882 ## 2 55.7454 98613 20 9.16242 ## 3 51.9337 100341 19 8.50248 ## 4 57.1210 104882 20 7.81190 ## 5 58.2004 129559 22 7.76588 ## ... ... ... ... ... ## 188 64.4249 48126 20 6.19684 ## 189 60.6675 112225 25 8.63302 ## 190 53.6781 135693 23 7.53559 ## 191 56.5235 3505 16 7.50037 ## 192 48.9489 150375 29 8.05569 ## altexps_ERCC_sum altexps_ERCC_detected altexps_ERCC_percent ## ## 1 65278 39 6.80658 ## 2 74748 40 6.28030 ## 3 60878 42 4.78949 ## 4 60073 42 4.18567 ## 5 136810 44 7.28887 ## ... ... ... ... ## 188 61575 39 7.17620 ## 189 94982 41 6.65764 ## 190 113707 40 5.81467 ## 191 7580 44 13.48898 ## 192 48664 39 2.51930 ## altexps_SIRV_sum altexps_SIRV_detected altexps_SIRV_percent total ## ## 1 27828 7 2.90165 959042 ## 2 39173 7 3.29130 1190198 ## 3 30058 7 2.36477 1271074 ## 4 32542 7 2.26741 1435208 ## 5 71850 7 3.82798 1876971 ## ... ... ... ... ... ## 188 19848 7 2.313165 858045 ## 189 31729 7 2.224004 1426661 ## 190 41116 7 2.102562 1955519 ## 191 1883 7 3.350892 56194 ## 192 16289 7 0.843271 1931645 另外，还可以使用addPerCellQC()，它会把每个细胞的QC指标加到SingleCellExperiment对象的colData中，方便后面调取\nsce.416b \u003c- addPerCellQC(sce.416b, subsets=list(Mito=is.mito)) colnames(colData(sce.416b)) ## [1] \"Source Name\" \"cell line\" ## [3] \"cell type\" \"single cell well quality\" ## [5] \"genotype\" \"phenotype\" ## [7] \"strain\" \"spike-in addition\" ## [9] \"block\" \"sum\" ## [11] \"detected\" \"percent_top_50\" ## [13] \"percent_top_100\" \"percent_top_200\" ## [15] \"percent_top_500\" \"subsets_Mito_sum\" ## [17] \"subsets_Mito_detected\" \"subsets_Mito_percent\" ## [19] \"altexps_ERCC_sum\" \"altexps_ERCC_detected\" ## [21] \"altexps_ERCC_percent\" \"altexps_SIRV_sum\" ## [23] \"altexps_SIRV_detected\" \"altexps_SIRV_percent\" ## [25] \"total\" 当然，这里做QC统计的前提假设是：每个细胞的qc指标都是独立于生物学状态的。也就是说，qc指标不会那么智能地识别细胞类型然后进行判断。它会认为（如文库太小、高线粒体占比）都是由技术误差引起的，而非生物因素。但是有一个问题，如果某些细胞类型本身的RNA含量就很低，或者它们本来就含有很多的线粒体转录本呢？再根据这个指标进行过滤，会不会损失一些细胞类型呢？\n识别低质量细胞 1. 使用固定的阈值 识别低质量细胞最简单方法是在QC度量上应用阈值。例如设定文库低于10万reads的细胞是低质量的，或者表达基因数量少于5000个，spike-in或线粒体占比高于10%。\nqc.lib \u003c- df$sum \u003c 1e5 qc.nexprs \u003c- df$detected \u003c 5e3 qc.spike \u003c- df$altexps_ERCC_percent \u003e 10 qc.mito \u003c- df$subsets_Mito_percent \u003e 10 discard \u003c- qc.lib | qc.nexprs | qc.spike | qc.mito # Summarize the number of cells removed for each reason. DataFrame(LibSize=sum(qc.lib), NExprs=sum(qc.nexprs), SpikeProp=sum(qc.spike), MitoProp=sum(qc.mito), Total=sum(discard)) ## DataFrame with 1 row and 5 columns ## LibSize NExprs SpikeProp MitoProp Total ## ## 1 3 0 19 14 33 虽然看起来简单，但使用这种方法需要丰富的经验，了解实验设计和细胞状态；另外即使使用同一种文库制备方法，但由于细胞捕获效率和测序深度的不同，这个阈值依然需要适时调整。因此对于研究人员要求很高。\n2. 使用相对阈值 鉴定离群点 为了获得相对阈值，先假设大部分细胞都是高质量的，然后去找离群点作为低质量。那么按照什么方法找离群点呢？常用的一个函数isOutlier使用的是MAD指标（绝对中位差来估计方差,先计算出数据与它们的中位数之间的偏差，然后这些偏差的绝对值的中位数就是MAD，median absolute deviation）。如果超过中位数3倍MAD的值就是离群值。\n使用isOutlier时，如果要相减（例如：df$sum - 3* MAD），就用type=\"lower\"，此时一般还要做个log转换log=TRUE ，保证得到的结果不是负数\nqc.lib2 \u003c- isOutlier(df$sum, log=TRUE, type=\"lower\") qc.nexprs2 \u003c- isOutlier(df$detected, log=TRUE, type=\"lower\") qc.spike2 \u003c- isOutlier(df$altexps_ERCC_percent, type=\"higher\") qc.mito2 \u003c- isOutlier(df$subsets_Mito_percent, type=\"higher\") attr(qc.lib2, \"thresholds\") lower higher 434082.9 Inf attr(qc.nexprs2, \"thresholds\") lower higher 5231.468 Inf attr(qc.spike2, \"thresholds\") lower higher -Inf 14.15371 attr(qc.mito2, \"thresholds\") lower higher -Inf 11.91734 用相对阈值过滤的细胞数量统计：\ndiscard2 \u003c- qc.lib2 | qc.nexprs2 | qc.spike2 | qc.mito2 # Summarize the number of cells removed for each reason. DataFrame(LibSize=sum(qc.lib2), NExprs=sum(qc.nexprs2), SpikeProp=sum(qc.spike2), MitoProp=sum(qc.mito2), Total=sum(discard2)) ## DataFrame with 1 row and 5 columns ## LibSize NExprs SpikeProp MitoProp Total ## ## 1 4 0 1 2 6 除此以外，还有一种更快的计算方法，一步整合了上面的操作：\nreasons \u003c- quickPerCellQC(df, percent_subsets=c(\"subsets_Mito_percent\", \"altexps_ERCC_percent\")) colSums(as.matrix(reasons)) ## low_lib_size low_n_features high_subsets_Mito_percent ## 4 0 2 ## high_altexps_ERCC_percent discard ## 1 6 使用”相对“的阈值一个好处就是可以根据测序深度、cDNA捕获效率、线粒体含量等等进行阈值的调整，在经验不是足够丰富的时候，可以辅助判断。但仍需要注意的是，使用相对阈值是有前提的，那就是：认为大部分细胞都是高质量的\n离群点检测的假设 离群点的检测的假设是大部分细胞的质量都不错。这一点假设可以通过实验去验证（比如肉眼检查细胞完整性）。但当大部分细胞质量都很低，使用相对阈值结果就对大量的低质量细胞无计可施。因为它使用了MAD值，和中位数有关系，那么可以试想：如果一堆数都不合格，那么它们的中位数也不合格，因此原来正确的值，其实在这群不合格的数值中，就是“离群”的。 另一个假设就是：每个细胞的qc指标都是独立于生物学状态的。也就是说，qc指标不会那么智能地识别细胞类型然后进行判断。在异质性很高的组织中， 就是存在内源RNA含量低，而线粒体基因占比高的细胞。如果使用”一刀切“的固定阈值，它们就很可能会被当成离群点被过滤。而是用MAD计算方法检测的结果可能就是：虽然一堆细胞的某个qc指标差异很大，但中位数也在变，随之变化的还有MAD值，这样最后的结果不至于和真实生物学情况差太多 考虑实验的因素 很多大型的实验都包含多个细胞系，而且可能采用的实验方法不同（比如选用不同的测序深度），这就产生了实验的不同批次。这种情况下， 应该对每个批次分别进行检测。\n如果每个批次都是一个SingleCellExperiment对象，那么isOutlier()函数可以按上面的方法直接使用；但是如果不同批次的细胞已经混合成一整个SingleCellExperiment对象，那么batch=这个参数就派上用场了。\n同样以这个416B数据集为例，他包含了两种不同的实验类型。然后我们就可以使用batch=参数去进行质控。\nbatch \u003c- paste0(sce.416b$phenotype, \"-\", sce.416b$Plate) batch.reasons \u003c- quickPerCellQC(df, percent_subsets=c(\"subsets_Mito_percent\", \"altexps_ERCC_percent\"), batch=batch) colSums(as.matrix(batch.reasons)) ## low_lib_size low_n_features high_subsets_Mito_percent ## 4 2 2 ## high_altexps_ERCC_percent discard ## 4 7 但是，batch参数不是万能的，之前也说过，这种相对阈值需要一个假设：每个批次的大部分细胞都是高质量的。当某个批次的细胞整体质量偏低，离群点检测很有可能失败\n例如，在Grun et al. (2016)的数据集中有两个donor的实验是失败的。它们的ERCC含量相对其他批次高，增加了中位数和MAD值，导致过滤低质量细胞失败。因此这种情况下，可以先算其他几个批次的中位数和MAD值，然后参考这些值去对有问题的批次进行过滤。\nlibrary(scRNAseq) sce.grun \u003c- GrunPancreasData() sce.grun \u003c- addPerCellQC(sce.grun) # First attempt with batch-specific thresholds. discard.ercc \u003c- isOutlier(sce.grun$altexps_ERCC_percent, type=\"higher\", batch=sce.grun$donor) with.blocking \u003c- plotColData(sce.grun, x=\"donor\", y=\"altexps_ERCC_percent\", colour_by=I(discard.ercc)) # Second attempt, sharing information across batches # to avoid dramatically different thresholds for unusual batches. discard.ercc2 \u003c- isOutlier(sce.grun$altexps_ERCC_percent, type=\"higher\", batch=sce.grun$donor, subset=sce.grun$donor %in% c(\"D17\", \"D2\", \"D7\")) without.blocking \u003c- plotColData(sce.grun, x=\"donor\", y=\"altexps_ERCC_percent\", colour_by=I(discard.ercc2)) gridExtra::grid.arrange(with.blocking, without.blocking, ncol=2) 注：可以看到，左图是按照每个批次分别鉴定的离群点；右图是用质量好的批次计算的阈值，然后运用到差的批次上的结果\n为了鉴别有问题的批次，可以先将每个批次分别计算结果，然后比较它们的阈值，如果比同类批次超出太多，就要警觉。\nercc.thresholds \u003c- attr(discard.ercc, \"thresholds\")[\"higher\",] ercc.thresholds ## D10 D17 D2 D3 D7 ## 73.610696 7.599947 6.010975 113.105828 15.216956 names(ercc.thresholds)[isOutlier(ercc.thresholds, type=\"higher\")] ## [1] \"D10\" \"D3\" 可以看到D10、D3的阈值就超过其他批次很多\n但是这么做的前提都是：我们认为批次中的细胞质量整体还不错。如果我们保证不了细胞质量，那么这种计算相对阈值的方法就不成立，还是要使用绝对阈值，手动去除。\n3. 其他方法 另一个策略是根据每个细胞的QC指标来在高维空间中识别异常值。利用robustbase 包，将细胞放在高维空间，根据他们的QC指标（文库大小、表达基因数、spike-in比例、线粒体比例），然后使用isOutlier()来识别表现出异常高outlylier水平的低质量细胞\nstats \u003c- cbind(log10(df$sum), log10(df$detected), df$subsets_Mito_percent, df$altexps_ERCC_percent) library(robustbase) outlying \u003c- adjOutlyingness(stats, only.outlyingness = TRUE) multi.outlier \u003c- isOutlier(outlying, type = \"higher\") summary(multi.outlier) ## Mode FALSE TRUE ## logical 182 10 除此以外，有时还可以根据基因表达量进行过滤，不过在bulk转录组中用的比较多，但是在scRNA中这样操作可能会损失一些本身数量就比较少的高质量细胞群体（比如一些罕见细胞，本身基因表达量就不是很高）\n画图检查 检查QC度量的分布以确定可能的问题是一个很好的实践。在最理想的情况下，我们会看到正态分布，可以证明在离群值检测中使用的3 MAD阈值是合理的。另一种模式下的大量细胞表明QC指标可能与某些生物状态相关，可能导致过滤过程中不同细胞类型的丢失;或者有不一致的库准备为一个子集的细胞，一个非罕见的现象在板的协议。\n# 把QC指标和原来的sce.416b细胞信息整合起来 colData(sce.416b) \u003c- cbind(colData(sce.416b), df) # 修改一下整合后的信息 sce.416b$block \u003c- factor(sce.416b$block) sce.416b$phenotype \u003c- ifelse(grepl(\"induced\", sce.416b$phenotype), \"induced\", \"wild type\") sce.416b$discard \u003c- reasons$discard # 绘图 gridExtra::grid.arrange( plotColData(sce.416b, x=\"block\", y=\"sum\", colour_by=\"discard\", other_fields=\"phenotype\") + facet_wrap(~phenotype) + scale_y_log10() + ggtitle(\"Total count\"), plotColData(sce.416b, x=\"block\", y=\"detected\", colour_by=\"discard\", other_fields=\"phenotype\") + facet_wrap(~phenotype) + scale_y_log10() + ggtitle(\"Detected features\"), plotColData(sce.416b, x=\"block\", y=\"subsets_Mito_percent\", colour_by=\"discard\", other_fields=\"phenotype\") + facet_wrap(~phenotype) + ggtitle(\"Mito percent\"), plotColData(sce.416b, x=\"block\", y=\"altexps_ERCC_percent\", colour_by=\"discard\", other_fields=\"phenotype\") + facet_wrap(~phenotype) + ggtitle(\"ERCC percent\"), ncol=1 ) 展示的是不同批次的QC指标\n另一种有用的诊断方法是绘制相对于其他QC指标的线粒体计数比例图。\n为了确保不存在这样的细胞：虽然细胞文库大，但它的线粒体占比也高。另外也是为了避免意外过滤掉本来是高质量但同时具有高代谢活性（即高线粒体占比）的细胞（如肝脏细胞）\n======未完\n针对Droplet数据的细胞判断 背景 由于这种建库方法的独特性，我们没办法事先知道某一个细胞文库（比如一个cell barcode）是真正包含一个细胞还是只是一个空的液滴（droplet）。因此，第一步是需要根据得到的表达量信息，来推断液滴是不是空的。但仅仅根据表达量判断还是不太靠谱，因为还有可能在空的液滴中依然包含外源RNA，最后的细胞文库依旧不为0，但确实不包含细胞。\n这里为了说明这个问题，使用了一个未过滤的10X PBMC数据\n# 数据下载 library(BiocFileCache) bfc \u003c- BiocFileCache(\"raw_data\", ask = FALSE) raw.path \u003c- bfcrpath(bfc, file.path(\"http://cf.10xgenomics.com/samples\", \"cell-exp/2.1.0/pbmc4k/pbmc4k_raw_gene_bc_matrices.tar.gz\")) # 解压数据 untar(raw.path, exdir=file.path(tempdir(), \"pbmc4k\")) library(DropletUtils) library(Matrix) fname \u003c- file.path(tempdir(), \"pbmc4k/raw_gene_bc_matrices/GRCh38\") sce.pbmc \u003c- read10xCounts(fname, col.names=TRUE) sce.pbmc ## class: SingleCellExperiment ## dim: 33694 737280 ## metadata(1): Samples ## assays(1): counts ## rownames(33694): ENSG00000243485 ENSG00000237613 ... ENSG00000277475 ## ENSG00000268674 ## rowData names(2): ID Symbol ## colnames(737280): AAACCTGAGAAACCAT-1 AAACCTGAGAAACCGC-1 ... ## TTTGTCATCTTTAGTC-1 TTTGTCATCTTTCCTC-1 ## colData names(2): Sample Barcode ## reducedDimNames(0): ## altExpNames(0): 整体观察不同的barcodes（不一定都是真的细胞）的文库大小分布：\nbcrank \u003c- barcodeRanks(counts(sce.pbmc)) # Only showing unique points for plotting speed. uniq \u003c- !duplicated(bcrank$rank) plot(bcrank$rank[uniq], bcrank$total[uniq], log=\"xy\", xlab=\"Rank\", ylab=\"Total UMI count\", cex.lab=1.2) abline(h=metadata(bcrank)$inflection, col=\"darkgreen\", lty=2) abline(h=metadata(bcrank)$knee, col=\"dodgerblue\", lty=2) legend(\"bottomleft\", legend=c(\"Inflection\", \"Knee\"), col=c(\"darkgreen\", \"dodgerblue\"), lty=2, cex=1.2) 看到各个barcodes的文库大小有高有低，并且相差较大，因此可能对应着真实存在的细胞和空液滴。当然最简单的办法还是”一刀切“，留下那些文库较大的细胞，不过还是有损失真实细胞类型的风险\n检测空的液滴 我们使用emptyDrops()函数，检查每个barcode的表达量是不是和外源RNA的表达量有显著差异(Lun et al. 2019)。如果存在显著差异，就说明barcode中更有可能是一个细胞。这种方法可以帮助区分：测序质量好的空液滴 和 包含细胞但RNA含量较少的液滴。尽管它们总体的表达量可能很相似，但本质不同，还是要区分的。\nemptyDrops() 使用Monte Carlo统计模拟计算p值，如果要重复结果，需要设置随机种子。另外设置 false discovery rate (FDR)为0.1%，意味着不超过0.1%的barcodes是空的。\n",
  "wordCount" : "5677",
  "inLanguage": "en",
  "image":"https://source.unsplash.com/random/400x200?code","datePublished": "2020-08-12T16:50:44+08:00",
  "dateModified": "2024-02-29T16:50:44+08:00",
  "author":{
    "@type": "Person",
    "name": "Jeason"
  },
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "http://localhost:1313/en/posts/tech/sc_qc/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Jeason's Blog",
    "logo": {
      "@type": "ImageObject",
      "url": "http://localhost:1313/img/avatar.jpeg"
    }
  }
}
</script>
</head>

<body class="" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="http://localhost:1313/en/" accesskey="h" title="Jeason&#39;s Blog (Alt + H)">
                <img src="http://localhost:1313/img/avatar.jpeg" alt="" aria-label="logo"
                    height="35">Jeason&#39;s Blog</a>
            <div class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="18" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="18" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                
            </div>
        </div>
        <ul id="menu">
            <li class="menu-item">
                <a href="http://localhost:1313/en/search" title="🔍 搜索 (Alt &#43; /)" accesskey=/>
                    <span>🔍 搜索</span>
                    
                </a>
            
            </li>
            <li class="menu-item">
                <a href="http://localhost:1313/en/" title="🏡 主页">
                    <span>🏡 主页</span>
                    
                </a>
            
            </li>
            <li class="menu-item">
                <a href="http://localhost:1313/en/posts" title="📚 文章">
                    <span>📚 文章</span>
                    
                </a>
            
            </li>
            <li class="menu-item">
                <a href="http://localhost:1313/en/archives" title="📈 归档">
                    <span>📈 归档</span>
                    
                </a>
            
            </li>
        </ul>
    </nav>
</header>
<main class="main page">
<article class="post-single">
    <div id="single-content">
        <header class="post-header">
            <div class="breadcrumbs"><a href="http://localhost:1313/en/">主页</a>&nbsp;»&nbsp;<a href="http://localhost:1313/en/posts/">📚 文章</a>&nbsp;»&nbsp;<a href="http://localhost:1313/en/posts/tech/">👨🏻‍💻 技术</a></div>
            <h1 class="post-title">
                scRNA的细胞质控
            </h1>
            <div class="post-meta">创建: 2020-08-12 | 更新: 2024-02-29 | 作者:Jeason




                
                <span id="busuanzi_container_page_pv">
                    &nbsp;|&nbsp;访问:&nbsp;<span id="busuanzi_value_page_pv"></span>
                </span>

</div>
        </header> <aside id="toc-container" class="toc-container wide">
    <div class="toc">
        <details  open>
        <summary accesskey="c" title="(Alt + C)">
            <span class="details">文章目录</span>
        </summary>

        <div class="inner"><ul>
                <li>
                    <a href="#%e5%8a%a8%e6%9c%ba" aria-label="动机">动机</a></li>
                <li>
                    <a href="#qc%e6%8c%87%e6%a0%87%e7%9a%84%e9%80%89%e6%8b%a9" aria-label="QC指标的选择">QC指标的选择</a></li>
                <li>
                    <a href="#%e8%af%86%e5%88%ab%e4%bd%8e%e8%b4%a8%e9%87%8f%e7%bb%86%e8%83%9e" aria-label="识别低质量细胞">识别低质量细胞</a><ul>
                        
                <li>
                    <a href="#1-%e4%bd%bf%e7%94%a8%e5%9b%ba%e5%ae%9a%e7%9a%84%e9%98%88%e5%80%bc" aria-label="1. 使用固定的阈值">1. 使用固定的阈值</a></li>
                <li>
                    <a href="#2-%e4%bd%bf%e7%94%a8%e7%9b%b8%e5%af%b9%e9%98%88%e5%80%bc" aria-label="2. 使用相对阈值">2. 使用相对阈值</a><ul>
                        
                <li>
                    <a href="#%e9%89%b4%e5%ae%9a%e7%a6%bb%e7%be%a4%e7%82%b9" aria-label="鉴定离群点">鉴定离群点</a></li></ul>
                </li>
                <li>
                    <a href="#%e7%a6%bb%e7%be%a4%e7%82%b9%e6%a3%80%e6%b5%8b%e7%9a%84%e5%81%87%e8%ae%be" aria-label="离群点检测的假设">离群点检测的假设</a></li>
                <li>
                    <a href="#%e8%80%83%e8%99%91%e5%ae%9e%e9%aa%8c%e7%9a%84%e5%9b%a0%e7%b4%a0" aria-label="考虑实验的因素">考虑实验的因素</a></li>
                <li>
                    <a href="#3-%e5%85%b6%e4%bb%96%e6%96%b9%e6%b3%95" aria-label="3. 其他方法">3. 其他方法</a></li></ul>
                </li>
                <li>
                    <a href="#%e7%94%bb%e5%9b%be%e6%a3%80%e6%9f%a5" aria-label="画图检查">画图检查</a></li>
                <li>
                    <a href="#%e9%92%88%e5%af%b9droplet%e6%95%b0%e6%8d%ae%e7%9a%84%e7%bb%86%e8%83%9e%e5%88%a4%e6%96%ad" aria-label="针对Droplet数据的细胞判断">针对Droplet数据的细胞判断</a><ul>
                        
                <li>
                    <a href="#%e8%83%8c%e6%99%af" aria-label="背景">背景</a></li>
                <li>
                    <a href="#%e6%a3%80%e6%b5%8b%e7%a9%ba%e7%9a%84%e6%b6%b2%e6%bb%b4" aria-label="检测空的液滴">检测空的液滴</a>
                </li>
            </ul>
            </li>
            </ul>
        </div>
        </details>
    </div>
</aside>
<script>
    let activeElement;
    let elements;
    window.addEventListener('DOMContentLoaded', function (event) {
        checkTocPosition();

        elements = document.querySelectorAll('h1[id],h2[id],h3[id],h4[id],h5[id],h6[id]');
        
        activeElement = elements[0];
        const id = encodeURI(activeElement.getAttribute('id')).toLowerCase();
        document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
    }, false);

    window.addEventListener('resize', function(event) {
        checkTocPosition();
    }, false);

    window.addEventListener('scroll', () => {
        
        activeElement = Array.from(elements).find((element) => {
            if ((getOffsetTop(element) - window.pageYOffset) > 0 &&
                (getOffsetTop(element) - window.pageYOffset) < window.innerHeight/2) {
                return element;
            }
        }) || activeElement

        elements.forEach(element => {
            const id = encodeURI(element.getAttribute('id')).toLowerCase();
            if (element === activeElement){
                document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
            } else {
                document.querySelector(`.inner ul li a[href="#${id}"]`).classList.remove('active');
            }
        })
    }, false);

    const main = parseInt(getComputedStyle(document.body).getPropertyValue('--article-width'), 10);
    const toc = parseInt(getComputedStyle(document.body).getPropertyValue('--toc-width'), 10);
    const gap = parseInt(getComputedStyle(document.body).getPropertyValue('--gap'), 10);

    function checkTocPosition() {
        const width = document.body.scrollWidth;
        if (width - main - (toc * 2) - (gap * 4) > 0) {
            document.getElementById("toc-container").classList.add("wide");
        } else {
            document.getElementById("toc-container").classList.remove("wide");
        }
    }

    function getOffsetTop(element) {
        if (!element.getClientRects().length) {
            return 0;
        }
        let rect = element.getBoundingClientRect();
        let win = element.ownerDocument.defaultView;
        return rect.top + win.pageYOffset;
    }
</script>
        <div class="post-content"><h2 id="动机">动机<a hidden class="anchor" aria-hidden="true" href="#动机">#</a></h2>
<p>scRNA-seq数据的低质量文库可能来自于：细胞分选环节的破坏、文库制备失误（没有足够的反转录或PCR次数）… 表现在：细胞总表达量低、基本没有表达的基因、高线粒体或spike-in占比。</p>
<p>这些低质量的库是有问题的，因为它们可能在下游分析中导致误导的结果 。</p>
<ul>
<li><strong>聚类问题：</strong> 低质量的细胞会聚集成一群，对结果的解释造成干扰，因为从这群细胞中得不到什么有用的信息，但是它的确也是一群。这种现象产生的原因有可能是：细胞破坏以后，线粒体或核RNAs占比升高。  最差的情况就是：不同类型的低质量细胞，也能聚在一起，因为相比于固有的生物差异，更相似的低质量让它们相依相偎。除此以外，本来非常小的细胞文库也能聚成一群，因为log转换后它们的平均表达量会发生很大的变化(A. Lun <a href="https://osca.bioconductor.org/quality-control.html#ref-lun2018overcoming" target="_blank" rel="noopener" style="color:#42b983";>2018</a>).</li>
<li><strong>方差估计或主成分分析</strong>:首先在PCA分析时，低质量和高质量之间的差异相比于生物学差异会体现更明显，占据主要的成分，减少降维结果的可靠性。另外，某个基因可能在两个细胞之间没什么表达差异，但是如果所处环境差异很大（一个细胞质量很低，另一个细胞质量正常），那么在差异估算过程中，就会把这个差异也会被当成差异表达基因。例如：一个低质量细胞文库的总表达量非常低（接近0），但恰巧还存在一个基因有表达量，那么这个基因的表达量在后续的文库归一化过程中就会尤为突出</li>
<li><strong>奇怪的转录本上调</strong>：实验难免会混入外源的污染转录本，但这个量很少并且在所有细胞中都应该是差不多水平的。但如果某个细胞质量低，文库小，那么在校正文库差异过程中，其中的污染转录本表达量就会“突飞猛进”，看起来是一些明显上调的“奇怪基因”。实际上，这些奇怪的基因依然在其他细胞中存在，并且真实的表达量差不多，并且是不应该占据主体地位的。</li>
</ul>
<p>为了最大程度避免上面一种或多种情况的发生，应该在归一化之前去掉这些低质量的细胞，这个过程就是**细胞的质控 **</p>
<p>使用A. T. L. Lun et al. (2017)的小型scRNA数据（未进行QC）进行测试</p>
<div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#776e71">#--- loading ---#</span>
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">library</span>(scRNAseq)
</span></span><span style="display:flex;"><span>sce.416b <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">LunSpikeInData</span>(which<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;416b&#34;</span>) 
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">table</span>(sce.416b<span style="color:#5bc4bf">$</span>block)
</span></span><span style="display:flex;"><span><span style="color:#f99b15">20160113</span> <span style="color:#f99b15">20160325</span> 
</span></span><span style="display:flex;"><span>      <span style="color:#f99b15">96</span>       <span style="color:#f99b15">96</span> 
</span></span><span style="display:flex;"><span>sce.416b<span style="color:#5bc4bf">$</span>block <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">factor</span>(sce.416b<span style="color:#5bc4bf">$</span>block)
</span></span><span style="display:flex;"><span>sce.416b
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#776e71">## class: SingleCellExperiment </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## dim: 46604 192 </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## metadata(0):</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## assays(1): counts</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## rownames(46604): ENSMUSG00000102693 ENSMUSG00000064842 ...</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##   ENSMUSG00000095742 CBFB-MYH11-mcherry</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## rowData names(1): Length</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## colnames(192): SLX-9555.N701_S502.C89V9ANXX.s_1.r_1</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##   SLX-9555.N701_S503.C89V9ANXX.s_1.r_1 ...</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##   SLX-11312.N712_S508.H5H5YBBXX.s_8.r_1</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##   SLX-11312.N712_S517.H5H5YBBXX.s_8.r_1</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## colData names(9): Source Name cell line ... spike-in addition block</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## reducedDimNames(0):</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## altExpNames(2): ERCC SIRV</span>
</span></span></code></pre></div><h2 id="qc指标的选择">QC指标的选择<a hidden class="anchor" aria-hidden="true" href="#qc指标的选择">#</a></h2>
<p>鉴定细胞是否是低质量的，需要用到几个指标。虽然下面这些指标是使用SMART-seq2数据进行展示的，但依然适用于UMI数据（比如MARS-seq、droplet-based技术）</p>
<ul>
<li><strong>文库大小（library size）</strong> ：指的是每个细胞中所有基因的表达量之和。细胞的文库如果很小，说明在文库制备过程中存在RNA的损失，要么是由于细胞裂解，要么是cDNA捕获不足导致后续的扩增量少</li>
<li><strong>每个细胞中表达基因的数目（number of expressed features in each cell）</strong>：指的是细胞中非0表达量的基因数量。如果细胞中基本没有基因表达，很可能是转录本捕获失败</li>
<li><strong>比对到spike-in转录本的reads比例（proportion of reads mapped to spike-in transcripts）</strong>：计算方法是：<code>spike-in counts / all features (including spike-ins) for each cell</code>。每个细胞都应该加入等量的外源转录本（spike-in），如果哪个细胞的spike-in比例提高了，说明它的内源RNA含量减少（比如在细胞分选阶段出现的细胞裂解或者RNA降解）</li>
<li><strong>比对到线粒体基因组的reads比例（proportion of reads mapped to genes in the mitochondrial genome）</strong> ：如果没有spike-in，那么使用线粒体指标也是能说明问题的(Islam et al. 2014; Ilicic et al. 2016)。比对到线粒体的reads增多，说明细胞质中的RNA减少，可能存在细胞穿孔的情况，而这个孔的大小，可能只是将细胞质中存在的mRNA流出去，但线粒体的体积超过了孔的大小，所以还留在了细胞中，造成一定程度的富集，导致线粒体RNA占比升高。</li>
</ul>
<p>对于每个细胞，可以用scater包的<code>perCellQCMetrics()</code>函数进行计算，其中<code>sum</code>这一列表示每个细胞的文库大小；<code>detected</code>这一列表示检测到的基因数量；<code>subsets_Mito_percent</code>这一列表示比对到线粒体基因组的reads占比；<code>altexps_ERCC_percent</code>表示比对到ERCC spike-in的reads占比</p>
<div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#776e71"># Retrieving the mitochondrial transcripts using genomic locations included in</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71"># the row-level annotation for the SingleCellExperiment.</span>
</span></span><span style="display:flex;"><span>location <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">rowRanges</span>(sce.416b)
</span></span><span style="display:flex;"><span>is.mito <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">any</span>(<span style="color:#06b6ef">seqnames</span>(location)<span style="color:#5bc4bf">==</span><span style="color:#48b685">&#34;MT&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#776e71"># ALTERNATIVELY: using resources in AnnotationHub to retrieve chromosomal</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71"># locations given the Ensembl IDs; this should yield the same result.</span>
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">library</span>(AnnotationHub)
</span></span><span style="display:flex;"><span>ens.mm.v97 <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">AnnotationHub</span>()[[<span style="color:#48b685">&#34;AH73905&#34;</span>]]
</span></span><span style="display:flex;"><span>chr.loc <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">mapIds</span>(ens.mm.v97, keys<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">rownames</span>(sce.416b),
</span></span><span style="display:flex;"><span>    keytype<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;GENEID&#34;</span>, column<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;SEQNAME&#34;</span>)
</span></span><span style="display:flex;"><span>is.mito.alt <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">which</span>(chr.loc<span style="color:#5bc4bf">==</span><span style="color:#48b685">&#34;MT&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">library</span>(scater)
</span></span><span style="display:flex;"><span>df <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">perCellQCMetrics</span>(sce.416b, subsets<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">list</span>(Mito<span style="color:#5bc4bf">=</span>is.mito))
</span></span><span style="display:flex;"><span>df
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#776e71">## DataFrame with 192 rows and 16 columns</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##           sum  detected percent_top_50 percent_top_100 percent_top_200</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##     &lt;integer&gt; &lt;integer&gt;      &lt;numeric&gt;       &lt;numeric&gt;       &lt;numeric&gt;</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 1      865936      7618        26.7218         32.2773         39.7208</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 2     1076277      7521        29.4043         35.0354         42.2581</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 3     1180138      8306        27.3454         32.4770         39.3296</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 4     1342593      8143        35.8092         40.2666         46.2460</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 5     1668311      7154        34.1198         39.0901         45.6660</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## ...       ...       ...            ...             ...             ...</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 188    776622      8174        45.9362         49.7010         54.6101</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 189   1299950      8956        38.0829         42.8930         49.0622</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 190   1800696      9530        30.6675         35.5839         41.8550</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 191     46731      6649        32.2998         37.9149         44.5999</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 192   1866692     10964        26.6632         31.2584         37.5608</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##     percent_top_500 subsets_Mito_sum subsets_Mito_detected subsets_Mito_percent</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##           &lt;numeric&gt;        &lt;integer&gt;             &lt;integer&gt;            &lt;numeric&gt;</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 1           52.9038            78790                    20              9.09882</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 2           55.7454            98613                    20              9.16242</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 3           51.9337           100341                    19              8.50248</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 4           57.1210           104882                    20              7.81190</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 5           58.2004           129559                    22              7.76588</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## ...             ...              ...                   ...                  ...</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 188         64.4249            48126                    20              6.19684</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 189         60.6675           112225                    25              8.63302</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 190         53.6781           135693                    23              7.53559</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 191         56.5235             3505                    16              7.50037</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 192         48.9489           150375                    29              8.05569</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##     altexps_ERCC_sum altexps_ERCC_detected altexps_ERCC_percent</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##            &lt;integer&gt;             &lt;integer&gt;            &lt;numeric&gt;</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 1              65278                    39              6.80658</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 2              74748                    40              6.28030</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 3              60878                    42              4.78949</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 4              60073                    42              4.18567</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 5             136810                    44              7.28887</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## ...              ...                   ...                  ...</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 188            61575                    39              7.17620</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 189            94982                    41              6.65764</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 190           113707                    40              5.81467</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 191             7580                    44             13.48898</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 192            48664                    39              2.51930</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##     altexps_SIRV_sum altexps_SIRV_detected altexps_SIRV_percent     total</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##            &lt;integer&gt;             &lt;integer&gt;            &lt;numeric&gt; &lt;integer&gt;</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 1              27828                     7              2.90165    959042</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 2              39173                     7              3.29130   1190198</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 3              30058                     7              2.36477   1271074</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 4              32542                     7              2.26741   1435208</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 5              71850                     7              3.82798   1876971</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## ...              ...                   ...                  ...       ...</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 188            19848                     7             2.313165    858045</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 189            31729                     7             2.224004   1426661</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 190            41116                     7             2.102562   1955519</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 191             1883                     7             3.350892     56194</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 192            16289                     7             0.843271   1931645</span>
</span></span></code></pre></div><p>另外，还可以使用<code>addPerCellQC()</code>，它会把每个细胞的QC指标加到<code>SingleCellExperiment</code>对象的<code>colData</code>中，方便后面调取</p>
<div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>sce.416b <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">addPerCellQC</span>(sce.416b, subsets<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">list</span>(Mito<span style="color:#5bc4bf">=</span>is.mito))
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">colnames</span>(<span style="color:#06b6ef">colData</span>(sce.416b))
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#776e71">##  [1] &#34;Source Name&#34;              &#34;cell line&#34;               </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##  [3] &#34;cell type&#34;                &#34;single cell well quality&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##  [5] &#34;genotype&#34;                 &#34;phenotype&#34;               </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##  [7] &#34;strain&#34;                   &#34;spike-in addition&#34;       </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##  [9] &#34;block&#34;                    &#34;sum&#34;                     </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## [11] &#34;detected&#34;                 &#34;percent_top_50&#34;          </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## [13] &#34;percent_top_100&#34;          &#34;percent_top_200&#34;         </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## [15] &#34;percent_top_500&#34;          &#34;subsets_Mito_sum&#34;        </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## [17] &#34;subsets_Mito_detected&#34;    &#34;subsets_Mito_percent&#34;    </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## [19] &#34;altexps_ERCC_sum&#34;         &#34;altexps_ERCC_detected&#34;   </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## [21] &#34;altexps_ERCC_percent&#34;     &#34;altexps_SIRV_sum&#34;        </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## [23] &#34;altexps_SIRV_detected&#34;    &#34;altexps_SIRV_percent&#34;    </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## [25] &#34;total&#34;</span>
</span></span></code></pre></div><p>当然，这里做QC统计的前提假设是：每个细胞的qc指标都是独立于生物学状态的。也就是说，qc指标不会那么智能地识别细胞类型然后进行判断。它会认为（如文库太小、高线粒体占比）都是由技术误差引起的，而非生物因素。但是有一个问题，<strong>如果某些细胞类型本身的RNA含量就很低，或者它们本来就含有很多的线粒体转录本呢？再根据这个指标进行过滤，会不会损失一些细胞类型呢？</strong></p>
<h2 id="识别低质量细胞">识别低质量细胞<a hidden class="anchor" aria-hidden="true" href="#识别低质量细胞">#</a></h2>
<h3 id="1-使用固定的阈值">1. 使用固定的阈值<a hidden class="anchor" aria-hidden="true" href="#1-使用固定的阈值">#</a></h3>
<p>识别低质量细胞最简单方法是在QC度量上应用阈值。例如设定文库低于10万reads的细胞是低质量的，或者表达基因数量少于5000个，spike-in或线粒体占比高于10%。</p>
<div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>qc.lib <span style="color:#5bc4bf">&lt;-</span> df<span style="color:#5bc4bf">$</span>sum <span style="color:#5bc4bf">&lt;</span> <span style="color:#f99b15">1e5</span>
</span></span><span style="display:flex;"><span>qc.nexprs <span style="color:#5bc4bf">&lt;-</span> df<span style="color:#5bc4bf">$</span>detected <span style="color:#5bc4bf">&lt;</span> <span style="color:#f99b15">5e3</span>
</span></span><span style="display:flex;"><span>qc.spike <span style="color:#5bc4bf">&lt;-</span> df<span style="color:#5bc4bf">$</span>altexps_ERCC_percent <span style="color:#5bc4bf">&gt;</span> <span style="color:#f99b15">10</span>
</span></span><span style="display:flex;"><span>qc.mito <span style="color:#5bc4bf">&lt;-</span> df<span style="color:#5bc4bf">$</span>subsets_Mito_percent <span style="color:#5bc4bf">&gt;</span> <span style="color:#f99b15">10</span>
</span></span><span style="display:flex;"><span>discard <span style="color:#5bc4bf">&lt;-</span> qc.lib <span style="color:#5bc4bf">|</span> qc.nexprs <span style="color:#5bc4bf">|</span> qc.spike <span style="color:#5bc4bf">|</span> qc.mito
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#776e71"># Summarize the number of cells removed for each reason.</span>
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">DataFrame</span>(LibSize<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">sum</span>(qc.lib), NExprs<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">sum</span>(qc.nexprs),
</span></span><span style="display:flex;"><span>    SpikeProp<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">sum</span>(qc.spike), MitoProp<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">sum</span>(qc.mito), Total<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">sum</span>(discard))
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#776e71">## DataFrame with 1 row and 5 columns</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##     LibSize    NExprs SpikeProp  MitoProp     Total</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##   &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 1         3         0        19        14        33</span>
</span></span></code></pre></div><p>虽然看起来简单，但使用这种方法需要丰富的经验，了解实验设计和细胞状态；另外即使使用同一种文库制备方法，但由于细胞捕获效率和测序深度的不同，这个阈值依然需要适时调整。因此对于研究人员要求很高。</p>
<h3 id="2-使用相对阈值">2. 使用相对阈值<a hidden class="anchor" aria-hidden="true" href="#2-使用相对阈值">#</a></h3>
<h4 id="鉴定离群点">鉴定离群点<a hidden class="anchor" aria-hidden="true" href="#鉴定离群点">#</a></h4>
<p>为了获得相对阈值，先假设大部分细胞都是高质量的，然后去找离群点作为低质量。那么按照什么方法找离群点呢？常用的一个函数<code>isOutlier</code>使用的是MAD指标（绝对中位差来估计方差,先计算出数据与它们的中位数之间的偏差，然后这些偏差的绝对值的中位数就是MAD，median absolute deviation）。如果<strong>超过中位数3倍MAD的值就是离群值</strong>。</p>
<p>使用<code>isOutlier</code>时，如果要相减（例如：<code>df$sum - 3* MAD</code>），就用<code>type=&quot;lower&quot;</code>，此时一般还要做个log转换<code>log=TRUE</code> ，保证得到的结果不是负数</p>
<div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>qc.lib2 <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">isOutlier</span>(df<span style="color:#5bc4bf">$</span>sum, log<span style="color:#5bc4bf">=</span><span style="color:#815ba4">TRUE</span>, type<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;lower&#34;</span>)
</span></span><span style="display:flex;"><span>qc.nexprs2 <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">isOutlier</span>(df<span style="color:#5bc4bf">$</span>detected, log<span style="color:#5bc4bf">=</span><span style="color:#815ba4">TRUE</span>, type<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;lower&#34;</span>)
</span></span><span style="display:flex;"><span>qc.spike2 <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">isOutlier</span>(df<span style="color:#5bc4bf">$</span>altexps_ERCC_percent, type<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;higher&#34;</span>)
</span></span><span style="display:flex;"><span>qc.mito2 <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">isOutlier</span>(df<span style="color:#5bc4bf">$</span>subsets_Mito_percent, type<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;higher&#34;</span>)
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#06b6ef">attr</span>(qc.lib2, <span style="color:#48b685">&#34;thresholds&#34;</span>)
</span></span><span style="display:flex;"><span>   lower   higher 
</span></span><span style="display:flex;"><span><span style="color:#f99b15">434082.9</span>      <span style="color:#815ba4">Inf</span> 
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">attr</span>(qc.nexprs2, <span style="color:#48b685">&#34;thresholds&#34;</span>)
</span></span><span style="display:flex;"><span>   lower   higher 
</span></span><span style="display:flex;"><span><span style="color:#f99b15">5231.468</span>      <span style="color:#815ba4">Inf</span> 
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">attr</span>(qc.spike2, <span style="color:#48b685">&#34;thresholds&#34;</span>)
</span></span><span style="display:flex;"><span>   lower   higher 
</span></span><span style="display:flex;"><span>    <span style="color:#5bc4bf">-</span><span style="color:#815ba4">Inf</span> <span style="color:#f99b15">14.15371</span> 
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">attr</span>(qc.mito2, <span style="color:#48b685">&#34;thresholds&#34;</span>)
</span></span><span style="display:flex;"><span>   lower   higher 
</span></span><span style="display:flex;"><span>    <span style="color:#5bc4bf">-</span><span style="color:#815ba4">Inf</span> <span style="color:#f99b15">11.91734</span> 
</span></span></code></pre></div><p>用相对阈值过滤的细胞数量统计：</p>
<div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>discard2 <span style="color:#5bc4bf">&lt;-</span> qc.lib2 <span style="color:#5bc4bf">|</span> qc.nexprs2 <span style="color:#5bc4bf">|</span> qc.spike2 <span style="color:#5bc4bf">|</span> qc.mito2
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#776e71"># Summarize the number of cells removed for each reason.</span>
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">DataFrame</span>(LibSize<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">sum</span>(qc.lib2), NExprs<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">sum</span>(qc.nexprs2),
</span></span><span style="display:flex;"><span>    SpikeProp<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">sum</span>(qc.spike2), MitoProp<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">sum</span>(qc.mito2), Total<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">sum</span>(discard2))
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#776e71">## DataFrame with 1 row and 5 columns</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##     LibSize    NExprs SpikeProp  MitoProp     Total</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##   &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## 1         4         0         1         2         6</span>
</span></span></code></pre></div><p><strong>除此以外，还有一种更快的计算方法，一步整合了上面的操作：</strong></p>
<div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>reasons <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">quickPerCellQC</span>(df, percent_subsets<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">c</span>(<span style="color:#48b685">&#34;subsets_Mito_percent&#34;</span>,
</span></span><span style="display:flex;"><span>    <span style="color:#48b685">&#34;altexps_ERCC_percent&#34;</span>))
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">colSums</span>(<span style="color:#06b6ef">as.matrix</span>(reasons))
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#776e71">##              low_lib_size            low_n_features high_subsets_Mito_percent </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##                         4                         0                         2 </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## high_altexps_ERCC_percent                   discard </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##                         1                         6</span>
</span></span></code></pre></div><p>使用”相对“的阈值一个好处就是可以根据测序深度、cDNA捕获效率、线粒体含量等等进行阈值的调整，在经验不是足够丰富的时候，可以辅助判断。但仍需要注意的是，<strong>使用相对阈值是有前提的，那就是：认为大部分细胞都是高质量的</strong></p>
<h3 id="离群点检测的假设">离群点检测的假设<a hidden class="anchor" aria-hidden="true" href="#离群点检测的假设">#</a></h3>
<ul>
<li>离群点的检测的假设是<strong>大部分细胞的质量都不错</strong>。这一点假设可以通过实验去验证（比如肉眼检查细胞完整性）。<strong>但当大部分细胞质量都很低，使用相对阈值结果就对大量的低质量细胞无计可施</strong>。因为它使用了MAD值，和中位数有关系，那么可以试想：如果一堆数都不合格，那么它们的中位数也不合格，因此原来正确的值，其实在这群不合格的数值中，就是“离群”的。</li>
<li>另一个假设就是：<strong>每个细胞的qc指标都是独立于生物学状态的</strong>。也就是说，qc指标不会那么智能地识别细胞类型然后进行判断。在异质性很高的组织中， 就是存在内源RNA含量低，而线粒体基因占比高的细胞。如果使用”一刀切“的固定阈值，它们就很可能会被当成离群点被过滤。而是用MAD计算方法检测的结果可能就是：虽然一堆细胞的某个qc指标差异很大，但中位数也在变，随之变化的还有MAD值，这样最后的结果不至于和真实生物学情况差太多</li>
</ul>
<h3 id="考虑实验的因素">考虑实验的因素<a hidden class="anchor" aria-hidden="true" href="#考虑实验的因素">#</a></h3>
<p>很多大型的实验都包含多个细胞系，而且可能采用的实验方法不同（比如选用不同的测序深度），这就产生了实验的不同批次。<strong>这种情况下， 应该对每个批次分别进行检测。</strong></p>
<p>如果每个批次都是一个<code>SingleCellExperiment</code>对象，那么<code>isOutlier()</code>函数可以按上面的方法直接使用；但是如果不同批次的细胞已经混合成一整个<code>SingleCellExperiment</code>对象，那么<code>batch=</code>这个参数就派上用场了。</p>
<p>同样以这个416B数据集为例，他包含了两种不同的实验类型。然后我们就可以使用<code>batch=</code>参数去进行质控。</p>
<div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>batch <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">paste0</span>(sce.416b<span style="color:#5bc4bf">$</span>phenotype, <span style="color:#48b685">&#34;-&#34;</span>, sce.416b<span style="color:#5bc4bf">$</span>Plate)
</span></span><span style="display:flex;"><span>batch.reasons <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">quickPerCellQC</span>(df, percent_subsets<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">c</span>(<span style="color:#48b685">&#34;subsets_Mito_percent&#34;</span>,
</span></span><span style="display:flex;"><span>    <span style="color:#48b685">&#34;altexps_ERCC_percent&#34;</span>), batch<span style="color:#5bc4bf">=</span>batch)
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">colSums</span>(<span style="color:#06b6ef">as.matrix</span>(batch.reasons))
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#776e71">##              low_lib_size            low_n_features high_subsets_Mito_percent </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##                         4                         2                         2 </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## high_altexps_ERCC_percent                   discard </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##                         4                         7</span>
</span></span></code></pre></div><p><strong>但是，<code>batch</code>参数不是万能的</strong>，之前也说过，这种相对阈值需要一个<strong>假设：每个批次的大部分细胞都是高质量的</strong>。当某个批次的细胞整体质量偏低，离群点检测很有可能失败</p>
<p>例如，在Grun et al. (2016)的数据集中有两个donor的实验是失败的。它们的ERCC含量相对其他批次高，增加了中位数和MAD值，导致过滤低质量细胞失败。因此这种情况下，可以先算其他几个批次的中位数和MAD值，然后参考这些值去对有问题的批次进行过滤。</p>
<div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#06b6ef">library</span>(scRNAseq)
</span></span><span style="display:flex;"><span>sce.grun <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">GrunPancreasData</span>()
</span></span><span style="display:flex;"><span>sce.grun <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">addPerCellQC</span>(sce.grun)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#776e71"># First attempt with batch-specific thresholds.</span>
</span></span><span style="display:flex;"><span>discard.ercc <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">isOutlier</span>(sce.grun<span style="color:#5bc4bf">$</span>altexps_ERCC_percent,
</span></span><span style="display:flex;"><span>    type<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;higher&#34;</span>, batch<span style="color:#5bc4bf">=</span>sce.grun<span style="color:#5bc4bf">$</span>donor)
</span></span><span style="display:flex;"><span>with.blocking <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">plotColData</span>(sce.grun, x<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;donor&#34;</span>, y<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;altexps_ERCC_percent&#34;</span>,
</span></span><span style="display:flex;"><span>    colour_by<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">I</span>(discard.ercc))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#776e71"># Second attempt, sharing information across batches</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71"># to avoid dramatically different thresholds for unusual batches.</span>
</span></span><span style="display:flex;"><span>discard.ercc2 <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">isOutlier</span>(sce.grun<span style="color:#5bc4bf">$</span>altexps_ERCC_percent,
</span></span><span style="display:flex;"><span>    type<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;higher&#34;</span>, batch<span style="color:#5bc4bf">=</span>sce.grun<span style="color:#5bc4bf">$</span>donor,
</span></span><span style="display:flex;"><span>    subset<span style="color:#5bc4bf">=</span>sce.grun<span style="color:#5bc4bf">$</span>donor <span style="color:#5bc4bf">%in%</span> <span style="color:#06b6ef">c</span>(<span style="color:#48b685">&#34;D17&#34;</span>, <span style="color:#48b685">&#34;D2&#34;</span>, <span style="color:#48b685">&#34;D7&#34;</span>))
</span></span><span style="display:flex;"><span>without.blocking <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">plotColData</span>(sce.grun, x<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;donor&#34;</span>, y<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;altexps_ERCC_percent&#34;</span>,
</span></span><span style="display:flex;"><span>    colour_by<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">I</span>(discard.ercc2))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>gridExtra<span style="color:#5bc4bf">::</span><span style="color:#06b6ef">grid.arrange</span>(with.blocking, without.blocking, ncol<span style="color:#5bc4bf">=</span><span style="color:#f99b15">2</span>) 
</span></span></code></pre></div><p><img loading="lazy" src="https://cdn.jsdelivr.net/gh/Moonerss/CDN/paper/quality-control/qc-plot-pancreas-1.png" alt=""  />
</p>
<blockquote>
<p>注：可以看到，左图是按照每个批次分别鉴定的离群点；右图是用质量好的批次计算的阈值，然后运用到差的批次上的结果</p>
</blockquote>
<p>为了鉴别有问题的批次，可以先将每个批次分别计算结果，然后比较它们的阈值，如果比同类批次超出太多，就要警觉。</p>
<div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>ercc.thresholds <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">attr</span>(discard.ercc, <span style="color:#48b685">&#34;thresholds&#34;</span>)[<span style="color:#48b685">&#34;higher&#34;</span>,]
</span></span><span style="display:flex;"><span>ercc.thresholds
</span></span><span style="display:flex;"><span><span style="color:#776e71">##        D10        D17         D2         D3         D7 </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##  73.610696   7.599947   6.010975 113.105828  15.216956</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">names</span>(ercc.thresholds)<span style="color:#06b6ef">[isOutlier</span>(ercc.thresholds, type<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;higher&#34;</span>)]
</span></span><span style="display:flex;"><span><span style="color:#776e71">## [1] &#34;D10&#34; &#34;D3&#34;</span>
</span></span></code></pre></div><blockquote>
<p>可以看到D10、D3的阈值就超过其他批次很多</p>
</blockquote>
<p>但是这么做的前提都是：我们认为批次中的细胞质量整体还不错。如果我们保证不了细胞质量，那么这种计算相对阈值的方法就不成立，还是要使用绝对阈值，手动去除。</p>
<h3 id="3-其他方法">3. 其他方法<a hidden class="anchor" aria-hidden="true" href="#3-其他方法">#</a></h3>
<p>另一个策略是根据每个细胞的QC指标来在高维空间中识别异常值。利用<code>robustbase</code> 包，将细胞放在高维空间，根据他们的QC指标（文库大小、表达基因数、spike-in比例、线粒体比例），然后使用<code>isOutlier()</code>来识别表现出异常高outlylier水平的低质量细胞</p>
<div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>stats <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">cbind</span>(<span style="color:#06b6ef">log10</span>(df<span style="color:#5bc4bf">$</span>sum), <span style="color:#06b6ef">log10</span>(df<span style="color:#5bc4bf">$</span>detected),
</span></span><span style="display:flex;"><span>    df<span style="color:#5bc4bf">$</span>subsets_Mito_percent, df<span style="color:#5bc4bf">$</span>altexps_ERCC_percent)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">library</span>(robustbase)
</span></span><span style="display:flex;"><span>outlying <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">adjOutlyingness</span>(stats, only.outlyingness <span style="color:#5bc4bf">=</span> <span style="color:#815ba4">TRUE</span>)
</span></span><span style="display:flex;"><span>multi.outlier <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">isOutlier</span>(outlying, type <span style="color:#5bc4bf">=</span> <span style="color:#48b685">&#34;higher&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">summary</span>(multi.outlier)
</span></span><span style="display:flex;"><span><span style="color:#776e71">##    Mode   FALSE    TRUE </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## logical     182      10</span>
</span></span></code></pre></div><p>除此以外，有时<strong>还可以根据基因表达量进行过滤</strong>，不过在bulk转录组中用的比较多，但是在scRNA中这样操作可能会损失一些本身数量就比较少的高质量细胞群体（比如一些罕见细胞，本身基因表达量就不是很高）</p>
<h2 id="画图检查">画图检查<a hidden class="anchor" aria-hidden="true" href="#画图检查">#</a></h2>
<p>检查QC度量的分布以确定可能的问题是一个很好的实践。在最理想的情况下，我们会看到正态分布，可以证明在离群值检测中使用的<code>3 MAD</code>阈值是合理的。另一种模式下的大量细胞表明QC指标可能与某些生物状态相关，可能导致过滤过程中不同细胞类型的丢失;或者有不一致的库准备为一个子集的细胞，一个非罕见的现象在板的协议。</p>
<div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#776e71"># 把QC指标和原来的sce.416b细胞信息整合起来</span>
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">colData</span>(sce.416b) <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">cbind</span>(<span style="color:#06b6ef">colData</span>(sce.416b), df)
</span></span><span style="display:flex;"><span><span style="color:#776e71"># 修改一下整合后的信息</span>
</span></span><span style="display:flex;"><span>sce.416b<span style="color:#5bc4bf">$</span>block <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">factor</span>(sce.416b<span style="color:#5bc4bf">$</span>block)
</span></span><span style="display:flex;"><span>sce.416b<span style="color:#5bc4bf">$</span>phenotype <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">ifelse</span>(<span style="color:#06b6ef">grepl</span>(<span style="color:#48b685">&#34;induced&#34;</span>, sce.416b<span style="color:#5bc4bf">$</span>phenotype),
</span></span><span style="display:flex;"><span>    <span style="color:#48b685">&#34;induced&#34;</span>, <span style="color:#48b685">&#34;wild type&#34;</span>)
</span></span><span style="display:flex;"><span>sce.416b<span style="color:#5bc4bf">$</span>discard <span style="color:#5bc4bf">&lt;-</span> reasons<span style="color:#5bc4bf">$</span>discard
</span></span><span style="display:flex;"><span><span style="color:#776e71"># 绘图</span>
</span></span><span style="display:flex;"><span>gridExtra<span style="color:#5bc4bf">::</span><span style="color:#06b6ef">grid.arrange</span>(
</span></span><span style="display:flex;"><span>    <span style="color:#06b6ef">plotColData</span>(sce.416b, x<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;block&#34;</span>, y<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;sum&#34;</span>, colour_by<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;discard&#34;</span>,
</span></span><span style="display:flex;"><span>        other_fields<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;phenotype&#34;</span>) <span style="color:#5bc4bf">+</span> <span style="color:#06b6ef">facet_wrap</span>(<span style="color:#5bc4bf">~</span>phenotype) <span style="color:#5bc4bf">+</span> 
</span></span><span style="display:flex;"><span>        <span style="color:#06b6ef">scale_y_log10</span>() <span style="color:#5bc4bf">+</span> <span style="color:#06b6ef">ggtitle</span>(<span style="color:#48b685">&#34;Total count&#34;</span>),
</span></span><span style="display:flex;"><span>    <span style="color:#06b6ef">plotColData</span>(sce.416b, x<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;block&#34;</span>, y<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;detected&#34;</span>, colour_by<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;discard&#34;</span>, 
</span></span><span style="display:flex;"><span>        other_fields<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;phenotype&#34;</span>) <span style="color:#5bc4bf">+</span> <span style="color:#06b6ef">facet_wrap</span>(<span style="color:#5bc4bf">~</span>phenotype) <span style="color:#5bc4bf">+</span> 
</span></span><span style="display:flex;"><span>        <span style="color:#06b6ef">scale_y_log10</span>() <span style="color:#5bc4bf">+</span> <span style="color:#06b6ef">ggtitle</span>(<span style="color:#48b685">&#34;Detected features&#34;</span>),
</span></span><span style="display:flex;"><span>    <span style="color:#06b6ef">plotColData</span>(sce.416b, x<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;block&#34;</span>, y<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;subsets_Mito_percent&#34;</span>, 
</span></span><span style="display:flex;"><span>        colour_by<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;discard&#34;</span>, other_fields<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;phenotype&#34;</span>) <span style="color:#5bc4bf">+</span> 
</span></span><span style="display:flex;"><span>        <span style="color:#06b6ef">facet_wrap</span>(<span style="color:#5bc4bf">~</span>phenotype) <span style="color:#5bc4bf">+</span> <span style="color:#06b6ef">ggtitle</span>(<span style="color:#48b685">&#34;Mito percent&#34;</span>),
</span></span><span style="display:flex;"><span>    <span style="color:#06b6ef">plotColData</span>(sce.416b, x<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;block&#34;</span>, y<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;altexps_ERCC_percent&#34;</span>, 
</span></span><span style="display:flex;"><span>        colour_by<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;discard&#34;</span>, other_fields<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;phenotype&#34;</span>) <span style="color:#5bc4bf">+</span> 
</span></span><span style="display:flex;"><span>        <span style="color:#06b6ef">facet_wrap</span>(<span style="color:#5bc4bf">~</span>phenotype) <span style="color:#5bc4bf">+</span> <span style="color:#06b6ef">ggtitle</span>(<span style="color:#48b685">&#34;ERCC percent&#34;</span>),
</span></span><span style="display:flex;"><span>    ncol<span style="color:#5bc4bf">=</span><span style="color:#f99b15">1</span>
</span></span><span style="display:flex;"><span>)
</span></span></code></pre></div><p><img loading="lazy" src="https://cdn.jsdelivr.net/gh/Moonerss/CDN/paper/quality-control/qc-dist-416b-1.png" alt=""  />
</p>
<blockquote>
<p>展示的是不同批次的QC指标</p>
</blockquote>
<p>另一种有用的诊断方法是绘制相对于其他QC指标的线粒体计数比例图。</p>
<p>为了确保不存在这样的细胞：虽然细胞文库大，但它的线粒体占比也高。另外也是为了避免意外过滤掉本来是高质量但同时具有高代谢活性（即高线粒体占比）的细胞（如肝脏细胞）</p>
<p>======未完</p>
<h2 id="针对droplet数据的细胞判断">针对Droplet数据的细胞判断<a hidden class="anchor" aria-hidden="true" href="#针对droplet数据的细胞判断">#</a></h2>
<h3 id="背景">背景<a hidden class="anchor" aria-hidden="true" href="#背景">#</a></h3>
<p>由于这种建库方法的独特性，我们没办法事先知道某一个细胞文库（比如一个cell barcode）是真正包含一个细胞还是只是一个空的液滴（droplet）。因此，<strong>第一步是需要根据得到的表达量信息，来推断液滴是不是空的</strong>。但仅仅根据表达量判断还是不太靠谱，因为还有可能在空的液滴中依然包含外源RNA，最后的细胞文库依旧不为0，但确实不包含细胞。</p>
<p>这里为了说明这个问题，<strong>使用了一个未过滤的10X PBMC数据</strong></p>
<div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#776e71"># 数据下载</span>
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">library</span>(BiocFileCache)
</span></span><span style="display:flex;"><span>bfc <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">BiocFileCache</span>(<span style="color:#48b685">&#34;raw_data&#34;</span>, ask <span style="color:#5bc4bf">=</span> <span style="color:#815ba4">FALSE</span>)
</span></span><span style="display:flex;"><span>raw.path <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">bfcrpath</span>(bfc, <span style="color:#06b6ef">file.path</span>(<span style="color:#48b685">&#34;http://cf.10xgenomics.com/samples&#34;</span>,
</span></span><span style="display:flex;"><span>    <span style="color:#48b685">&#34;cell-exp/2.1.0/pbmc4k/pbmc4k_raw_gene_bc_matrices.tar.gz&#34;</span>))
</span></span><span style="display:flex;"><span><span style="color:#776e71"># 解压数据</span>
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">untar</span>(raw.path, exdir<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">file.path</span>(<span style="color:#06b6ef">tempdir</span>(), <span style="color:#48b685">&#34;pbmc4k&#34;</span>))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">library</span>(DropletUtils)
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">library</span>(Matrix)
</span></span><span style="display:flex;"><span>fname <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">file.path</span>(<span style="color:#06b6ef">tempdir</span>(), <span style="color:#48b685">&#34;pbmc4k/raw_gene_bc_matrices/GRCh38&#34;</span>)
</span></span><span style="display:flex;"><span>sce.pbmc <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">read10xCounts</span>(fname, col.names<span style="color:#5bc4bf">=</span><span style="color:#815ba4">TRUE</span>)
</span></span><span style="display:flex;"><span>sce.pbmc
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#776e71">## class: SingleCellExperiment </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## dim: 33694 737280 </span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## metadata(1): Samples</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## assays(1): counts</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## rownames(33694): ENSG00000243485 ENSG00000237613 ... ENSG00000277475</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##   ENSG00000268674</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## rowData names(2): ID Symbol</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## colnames(737280): AAACCTGAGAAACCAT-1 AAACCTGAGAAACCGC-1 ...</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">##   TTTGTCATCTTTAGTC-1 TTTGTCATCTTTCCTC-1</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## colData names(2): Sample Barcode</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## reducedDimNames(0):</span>
</span></span><span style="display:flex;"><span><span style="color:#776e71">## altExpNames(0):</span>
</span></span></code></pre></div><p><strong>整体观察不同的barcodes（不一定都是真的细胞）的文库大小分布：</strong></p>
<div class="highlight"><pre tabindex="0" style="color:#e7e9db;background-color:#2f1e2e;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>bcrank <span style="color:#5bc4bf">&lt;-</span> <span style="color:#06b6ef">barcodeRanks</span>(<span style="color:#06b6ef">counts</span>(sce.pbmc))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#776e71"># Only showing unique points for plotting speed.</span>
</span></span><span style="display:flex;"><span>uniq <span style="color:#5bc4bf">&lt;-</span> <span style="color:#5bc4bf">!</span><span style="color:#06b6ef">duplicated</span>(bcrank<span style="color:#5bc4bf">$</span>rank)
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">plot</span>(bcrank<span style="color:#5bc4bf">$</span>rank[uniq], bcrank<span style="color:#5bc4bf">$</span>total[uniq], log<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;xy&#34;</span>,
</span></span><span style="display:flex;"><span>    xlab<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;Rank&#34;</span>, ylab<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;Total UMI count&#34;</span>, cex.lab<span style="color:#5bc4bf">=</span><span style="color:#f99b15">1.2</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">abline</span>(h<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">metadata</span>(bcrank)<span style="color:#5bc4bf">$</span>inflection, col<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;darkgreen&#34;</span>, lty<span style="color:#5bc4bf">=</span><span style="color:#f99b15">2</span>)
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">abline</span>(h<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">metadata</span>(bcrank)<span style="color:#5bc4bf">$</span>knee, col<span style="color:#5bc4bf">=</span><span style="color:#48b685">&#34;dodgerblue&#34;</span>, lty<span style="color:#5bc4bf">=</span><span style="color:#f99b15">2</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#06b6ef">legend</span>(<span style="color:#48b685">&#34;bottomleft&#34;</span>, legend<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">c</span>(<span style="color:#48b685">&#34;Inflection&#34;</span>, <span style="color:#48b685">&#34;Knee&#34;</span>), 
</span></span><span style="display:flex;"><span>        col<span style="color:#5bc4bf">=</span><span style="color:#06b6ef">c</span>(<span style="color:#48b685">&#34;darkgreen&#34;</span>, <span style="color:#48b685">&#34;dodgerblue&#34;</span>), lty<span style="color:#5bc4bf">=</span><span style="color:#f99b15">2</span>, cex<span style="color:#5bc4bf">=</span><span style="color:#f99b15">1.2</span>)
</span></span></code></pre></div><p><img loading="lazy" src="https://cdn.jsdelivr.net/gh/Moonerss/CDN/paper/quality-control/rankplot-1.png" alt=""  />
</p>
<blockquote>
<p>看到各个barcodes的文库大小有高有低，并且相差较大，因此可能对应着真实存在的细胞和空液滴。当然最简单的办法还是”一刀切“，<strong>留下那些文库较大的细胞</strong>，不过还是有损失真实细胞类型的风险</p>
</blockquote>
<h3 id="检测空的液滴">检测空的液滴<a hidden class="anchor" aria-hidden="true" href="#检测空的液滴">#</a></h3>
<p>我们使用<code>emptyDrops()</code>函数，检查每个barcode的表达量是不是和外源RNA的表达量有显著差异(Lun et al. 2019)。如果存在显著差异，就说明barcode中更有可能是一个细胞。这种方法可<strong>以帮助区分：测序质量好的空液滴 和 包含细胞但RNA含量较少的液滴</strong>。尽管它们总体的表达量可能很相似，但本质不同，还是要区分的。</p>
<p><code>emptyDrops()</code> 使用Monte Carlo统计模拟计算p值，如果要重复结果，需要设置随机种子。另外设置 false discovery rate (FDR)为0.1%，意味着不超过0.1%的barcodes是空的。</p>


        </div>

        

        <footer class="post-footer">
            
<nav class="paginav">
  <a class="prev" href="http://localhost:1313/en/posts/tech/ggplot2/">
    <span class="title">« 上一页</span>
    <br>
    <span>ggplot2 | 小部件参考</span>
  </a>
  <a class="next" href="http://localhost:1313/en/posts/read/r_stat/">
    <span class="title">下一页 »</span>
    <br>
    <span>R 语言统计检验函数汇总</span>
  </a>
</nav>

        </footer>
    </div>



<div>
    <div class="pagination__title">
        <span class="pagination__title-h" style="font-size: 20px;">💬评论</span>
        <hr />
    </div>

    <div id="tcomment"></div>

    <script src="https://cdn.staticfile.org/twikoo/1.6.15/twikoo.all.min.js">
    </script>
    

    

    <script>
        twikoo.init({
            envId: "https://twikoo.lvbibir.cn/", 
            el: "#tcomment",
            lang: 'zh-CN',
            
            path: window.TWIKOO_MAGIC_PATH||window.location.pathname,
            
            
            
            
            
            
            
        });
    </script>

</div>
</article>
</main>


<footer class="footer">

    <a href="https://gohugo.io/" target="_blank">
        
        <img style="display: unset;" src="https://image.lvbibir.cn/blog/frame-hugo-blue.svg">
    </a>
    <a href="https://github.com/adityatelange/hugo-PaperMod" target="_blank">
        
        <img style="display: unset;" src="https://image.lvbibir.cn/blog/theme-papermod-lightgrey.svg">
    </a>

    <br>

    <span id="runtime_span"></span>
    <script
        type="text/javascript">function show_runtime() { window.setTimeout("show_runtime()", 1000); X = new Date("7/13/2021 1:00:00"); Y = new Date(); T = (Y.getTime() - X.getTime()); M = 24 * 60 * 60 * 1000; a = T / M; A = Math.floor(a); b = (a - A) * 24; B = Math.floor(b); c = (b - B) * 60; C = Math.floor((b - B) * 60); D = Math.floor((c - C) * 60); runtime_span.innerHTML = "网站已运行" + A + "天" + B + "小时"} show_runtime();</script>
    |
    <script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
    <span id="busuanzi_container">
        
        总访客数:  <span id="busuanzi_value_site_uv"></span>
        |
        总访问量:  <span id="busuanzi_value_site_pv"></span>
    </span>

    <br>
    <span>
        Copyright
        &copy;
        2021-2024
        <a href="http://localhost:1313/en/" style="border-bottom:1px solid">Jeason&#39;s Blog</a>
    </span>

    
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <span class="topInner">
        <svg class="topSvg" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
            <path d="M12 6H0l6-6z" />
        </svg>
        <span id="read_progress"></span>
    </span>
</a>

<script>
    document.addEventListener('scroll', function (e) {
        const readProgress = document.getElementById("read_progress");
        const scrollHeight = document.documentElement.scrollHeight;
        const clientHeight = document.documentElement.clientHeight;
        const scrollTop = document.documentElement.scrollTop || document.body.scrollTop;
        
        readProgress.innerText = ((scrollTop / (scrollHeight - clientHeight)).toFixed(2) * 100).toFixed(0);
    })
</script>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    let mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 400 || document.documentElement.scrollTop > 400) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };
</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>


<script>
    document.querySelectorAll('pre > code').forEach((codeblock) => {
        const container = codeblock.parentNode.parentNode;

        const copybutton = document.createElement('button');
        copybutton.classList.add('copy-code');
        copybutton.innerText = '📄复制';

        function copyingDone() {
            copybutton.innerText = '👌🏻已复制!';
            setTimeout(() => {
                copybutton.innerText = '📄复制';
            }, 2000);
        }

        copybutton.addEventListener('click', (cb) => {
            
            
            
            
            
            
            
            
            
            
            
            
            
            

            const range = document.createRange();
            range.selectNodeContents(codeblock);
            const selection = window.getSelection();
            selection.removeAllRanges();
            selection.addRange(range);
            try {
                document.execCommand('copy');
                copyingDone();
            } catch (e) {
            }
            ;
            selection.removeRange(range);
        });

        let language = codeblock.className.replaceAll("language-", "")
        let macTool = document.createElement("div")
        let macTool1 = document.createElement("div")
        let macTool2 = document.createElement("div")
        let macTool3 = document.createElement("div")
        let languageType = document.createElement("div")
        languageType.innerText = language
        macTool.setAttribute('class', 'mac-tool')
        macTool1.setAttribute('class', 'mac bb1')
        macTool2.setAttribute('class', 'mac bb2')
        macTool3.setAttribute('class', 'mac bb3')
        languageType.setAttribute('class', 'language-type')
        macTool.appendChild(macTool1)
        macTool.appendChild(macTool2)
        macTool.appendChild(macTool3)
        macTool.appendChild(languageType)

        if (container.classList.contains("highlight")) {
            container.appendChild(copybutton);
            container.appendChild(macTool)
        } else if (container.parentNode.firstChild === container) {
            
        } else if (codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName === "TABLE") {
            
            codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(copybutton);
            container.appendChild(macTool)
        } else {
            
            codeblock.parentNode.appendChild(copybutton);
            container.appendChild(macTool)
        }
    });
</script>

</body>

</html>
